{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7563604a-2ba5-4e5d-b9e0-309615df0b5b",
   "metadata": {},
   "source": [
    "## **Notebook to make predictions with reprocessed Yoneda data and EFLUX2** \n",
    "\n",
    "By Christina Schenk and Garrett Roell\n",
    "\n",
    "Tested on biodesign_3.7 kernel on jprime"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8edc3c57-de26-4407-a568-88c6e6d99f02",
   "metadata": {
    "tags": []
   },
   "source": [
    "### EFLUX2 predictions and evaluations\n",
    "This notebook predicts fluxes for R. opacus cultures growing with glucose. The data was published in [Yoneda (2016)](https://academic.oup.com/nar/article/44/5/2240/2465306).\n",
    "\n",
    "\n",
    "\n",
    "#### **The data uses the following mapping as introduced in Notebook E:**\n",
    "\n",
    "#### Yoneda transcriptomics data: \n",
    "* WT 1.0 g/L Glucose, 0.05g/L ammonium sulfate (**WT-LN-G**) (3 trials)\n",
    "\n",
    "#### Combined with Rhiannon 2018 metabolomics and OD data:\n",
    "* Metabolomics and OD data for WT Glucose (before **WT-G** but now **WT-LN-G**)\n",
    "                                               \n",
    "\n",
    "### Method: \n",
    "<ol>\n",
    "<li>Predict fluxes with EFLUX2</li>\n",
    "<li>Compare predictions with 13CMFA: Scatter plots and flux maps</li>\n",
    "<li>Load file with observed growth rates (Notebook E)</li>\n",
    "<li>Compare growth rate predictions with growth rate observations</li>\n",
    "</ol>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a6edbe7-d3d0-49ac-9d13-8f43882945f0",
   "metadata": {},
   "source": [
    "##### **Import python packages**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5e37f811-1a95-45e0-932f-962edb144307",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cobra\n",
    "import scipy.stats\n",
    "#import cplex\n",
    "%matplotlib inline\n",
    "\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.offsetbox import (TextArea, DrawingArea, OffsetImage,\n",
    "                                  AnnotationBbox)\n",
    "from matplotlib.cbook import get_sample_data\n",
    "import matplotlib.image as mpimg\n",
    "import matplotlib.cm as cm\n",
    "\n",
    "from statistics import stdev\n",
    "\n",
    "from edd_utils import login, export_study, export_metadata\n",
    "\n",
    "from sklearn.metrics import r2_score\n",
    "\n",
    "from optlang.symbolics import add"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "244d4ba7-9d90-4f21-8de8-18a77f1a87dd",
   "metadata": {},
   "source": [
    "##### **Load data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c5da98ca-295a-449d-b007-fc96acf2e518",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define blank dictionary and fill with dataframes mapping transcript meausurements to GSM reaction ids\n",
    "reaction_transcripts = {}\n",
    "\n",
    "# add the dataframes that map reaction ids to transcript values to the object \n",
    "reaction_transcripts['glucose_cpm'] = pd.read_csv('../transcript_mapping/glucose_cpm.csv', index_col=0)\n",
    "reaction_transcripts['glucose_fpkm'] = pd.read_csv('../transcript_mapping/glucose_fpkm.csv', index_col=0)\n",
    "reaction_transcripts['glucose_mr'] = pd.read_csv('../transcript_mapping/glucose_mr.csv', index_col=0)\n",
    "reaction_transcripts['glucose_tmm'] = pd.read_csv('../transcript_mapping/glucose_tmm.csv', index_col=0)\n",
    "\n",
    "reaction_transcripts['phenol_cpm'] = pd.read_csv('../transcript_mapping/phenol_cpm.csv', index_col=0)\n",
    "reaction_transcripts['phenol_fpkm'] = pd.read_csv('../transcript_mapping/phenol_fpkm.csv', index_col=0)\n",
    "reaction_transcripts['phenol_mr'] = pd.read_csv('../transcript_mapping/phenol_mr.csv', index_col=0)\n",
    "reaction_transcripts['phenol_tmm'] = pd.read_csv('../transcript_mapping/phenol_tmm.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcf2477e-6117-4ff5-9532-7e43560a4400",
   "metadata": {},
   "source": [
    "#### Calculate the percentage of reactions with zero transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3e56b90e-2031-4710-9c10-9f9d1841961f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "glucose_cpm: 4.97% of reactions have zero transcripts\n",
      "glucose_fpkm: 4.97% of reactions have zero transcripts\n",
      "glucose_mr: 4.97% of reactions have zero transcripts\n",
      "glucose_tmm: 7.51% of reactions have zero transcripts\n",
      "phenol_cpm: 32.11% of reactions have zero transcripts\n",
      "phenol_fpkm: 32.11% of reactions have zero transcripts\n",
      "phenol_mr: 32.11% of reactions have zero transcripts\n",
      "phenol_tmm: 32.58% of reactions have zero transcripts\n"
     ]
    }
   ],
   "source": [
    "# loop over the dictionary containing the reaction to transcript mapping dataframes\n",
    "for df_name, reaction_transcript_df in reaction_transcripts.items():\n",
    "\n",
    "    # get a list of all the values in the dataframe\n",
    "    all_values = list(reaction_transcript_df.values.flatten())\n",
    "    \n",
    "    # find the total number of reactions with zero transcripts in the dataframe\n",
    "    number_zeros = all_values.count(0)\n",
    "\n",
    "    print(f'{df_name}: {100 * number_zeros/len(all_values):.2f}% of reactions have zero transcripts')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7252a93-7cea-4b7c-8375-c03a194a9524",
   "metadata": {},
   "source": [
    "##### **Load Genome Scale Model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f6b96da-3aaf-49fd-9dec-b674a678d243",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name =  '../GSMs/Ropacus_annotated_curated.xml'\n",
    "model = cobra.io.read_sbml_model(file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "240c3766-0bbf-41c8-80e2-9b70b175ff4c",
   "metadata": {},
   "source": [
    "#### Implementation of eflux with reaction and transcripts already mapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f1289cea-3d2e-4489-9195-da121e2cdc8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def eflux2(reaction_transcripts_df, carbon_source, sub_uptake_rate):\n",
    "    \n",
    "    # assign important reaction names based on carbon source\n",
    "    if carbon_source == 'glucose':\n",
    "        uptake_reaction_id = 'EX_glc__D_e'\n",
    "        objective_reaction_id = 'Growth_Glucose'\n",
    "    elif carbon_source == 'phenol':\n",
    "        uptake_reaction_id = 'EX_phenol_e'\n",
    "        objective_reaction_id = 'Growth_Phenol'\n",
    "    else:\n",
    "        print('carbon source is not compatible with this E-Flux2 implementation')\n",
    "        return\n",
    "        \n",
    "    # This is based on previous code, I'm not sure it's needed -Garrett\n",
    "    eflux2_model = model.copy()\n",
    "    \n",
    "    # open context manager to run FBA with eflux2 constraints\n",
    "    with eflux2_model:\n",
    "        \n",
    "        # ensure that a 1000 default reaction constraint will not come into limitiing. This is important\n",
    "        # becuase it makes the function robust to data with high transcript counts\n",
    "        for reaction in eflux2_model.reactions:\n",
    "            if reaction.lower_bound <= -1000:\n",
    "                reaction.lower_bound = -np.Inf\n",
    "            if reaction.upper_bound >= 1000:\n",
    "                reaction.upper_bound = np.Inf \n",
    "        \n",
    "        # set the upper and lower limits of reactions with transcript data\n",
    "        for reaction_id in reaction_transcripts_df.index:\n",
    "            \n",
    "            # get the reaction object using the id\n",
    "            reaction = eflux2_model.reactions.get_by_id(reaction_id)\n",
    "            \n",
    "            # get transcript level for reaction\n",
    "            transcript_level = reaction_transcripts_df[reaction_id]\n",
    "            \n",
    "            # if a reaction is capable of having negative flux, it's flux is capped at the transcript level\n",
    "            # Note: reactions with 0 transcripts are not bounded because when reactions with zero transcripts \n",
    "            # were set to zero flux, the model was not able to grow\n",
    "            if reaction.lower_bound < 0.0 and transcript_level != 0:\n",
    "                reaction.lower_bound = -1 * transcript_level\n",
    "                \n",
    "            \n",
    "            # if a reaction is capable of having negative flux, it's flux is capped at the transcript level\n",
    "            # Note: reactions with 0 transcripts are not bounded because when reactions with zero transcripts \n",
    "            # were set to zero flux, the model was not able to grow\n",
    "            if reaction.upper_bound > 0.0 and transcript_level != 0:\n",
    "                reaction.upper_bound = reaction_transcripts_df[reaction_id]\n",
    "                \n",
    "        # set up the medium so that it never limits the reaction network\n",
    "        medium = {k: np.inf for k,v in eflux2_model.medium.items()}\n",
    "        \n",
    "        # EFlux2 interestingly does not have set a bound for its carbon source\n",
    "        # The carbon source is allowed to have infinite uptake flux\n",
    "        # The uptake is eventually limited by a transcript measurement somewhere else in the network\n",
    "        \n",
    "        # All unused carbon source uptake fluxes are set to zero\n",
    "        \n",
    "        # for glucose, remove phenol uptake flux\n",
    "        if carbon_source == 'glucose':\n",
    "            medium['EX_phenol_e'] = 0\n",
    "        \n",
    "        # for phenol, remove glucose uptake flux\n",
    "        if carbon_source == 'phenol':\n",
    "            medium['EX_glc__D_e'] = 0\n",
    "        \n",
    "        # for phenol and glucose, remove all other carbon sources\n",
    "        medium['EX_guaiacol_e'] = 0\n",
    "        medium['EX_vanlt_e'] = 0\n",
    "        medium['EX_tag'] = 0\n",
    "\n",
    "        # set the model medium to the customized medium dictionary\n",
    "        eflux2_model.medium = medium\n",
    "\n",
    "        # check medium\n",
    "        # print(medium)\n",
    "        \n",
    "        # set the objective function with constant defined above\n",
    "        eflux2_model.objective = objective_reaction_id\n",
    "\n",
    "        # fix the unused biomass reactions to zero\n",
    "        if carbon_source == 'glucose':\n",
    "            eflux2_model.reactions.get_by_id('Growth_Phenol').upper_bound = 0\n",
    "            eflux2_model.reactions.get_by_id('Growth_Phenol').lower_bound = 0\n",
    "        if carbon_source == 'phenol':\n",
    "            eflux2_model.reactions.get_by_id('Growth_Glucose').upper_bound = 0\n",
    "            eflux2_model.reactions.get_by_id('Growth_Glucose').lower_bound = 0\n",
    "            \n",
    "        # always set the b.subtilis flux to zero\n",
    "        eflux2_model.reactions.get_by_id('Growth').upper_bound = 0\n",
    "        eflux2_model.reactions.get_by_id('Growth').lower_bound = 0\n",
    "                \n",
    "        # this prevents the solver from failing\n",
    "        eflux2_model.tolerance = 1e-9\n",
    "        \n",
    "        # solve the network\n",
    "        fba_sol = eflux2_model.optimize()\n",
    "    \n",
    "        # determine the growth and uptake rate of the transcript derived solution\n",
    "        transcript_derived_sub_uptake_rate = fba_sol.fluxes[uptake_reaction_id]\n",
    "        transcript_derived_growth_rate = fba_sol.fluxes[objective_reaction_id]\n",
    "        \n",
    "        print(f'The transcript derived substrate uptake rate is {transcript_derived_sub_uptake_rate:.4f}')\n",
    "        print(f'The transcript derived growth rate is {transcript_derived_growth_rate:.4f}')\n",
    "\n",
    "        # Constrain the biomass to the optimal value\n",
    "        eflux2_model.reactions.get_by_id(objective_reaction_id).upper_bound = fba_sol.objective_value\n",
    "        eflux2_model.reactions.get_by_id(objective_reaction_id).lower_bound = fba_sol.objective_value\n",
    "\n",
    "        # Minimize the sum of squared flux values\n",
    "        \"\"\"Note: Because of quadratic objective still have to use cplex objective formulation.\n",
    "        Optlang does not support quadratic type of constraints and objectives yet.\"\"\"\n",
    "\n",
    "        # check medium\n",
    "#         print(eflux2_model.medium)\n",
    "        \n",
    "        eflux2_model.objective = eflux2_model.problem.Objective(add([rxn.flux_expression**2 for rxn in eflux2_model.reactions]), direction='min')\n",
    "        eflux2_sol = eflux2_model.optimize()\n",
    "        # print('EFlux2 status', eflux2_sol.status)\n",
    "        # print('EFlux2 solution', eflux2_sol.objective_value)\n",
    "        \n",
    "        # Scale the solution by ratio of substrate uptake rates\n",
    "        scale_factor = -1 * sub_uptake_rate / transcript_derived_sub_uptake_rate\n",
    "        \n",
    "        # need to figure out how to scale this solution object\n",
    "        scaled_eflux2_sol = pd.Series(\n",
    "            [flux * scale_factor for flux in eflux2_sol.fluxes], # values\n",
    "            index=eflux2_sol.fluxes.index # index\n",
    "        )\n",
    "        \n",
    "        print(f'The scaled substrate uptake rate is {scaled_eflux2_sol[uptake_reaction_id]:.4f}')\n",
    "        print(f'The scaled growth rate is {scaled_eflux2_sol[objective_reaction_id]:.4f}')\n",
    "        \n",
    "        # spacer\n",
    "        print()\n",
    "\n",
    "        # return the fluxes as a series with reaction ids as the index and fluxes as values\n",
    "        return scaled_eflux2_sol "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bda32ef9-7b0e-4a92-8b70-7edbe52634a0",
   "metadata": {},
   "source": [
    "#### Calculate E-Flux2 fluxes for WT phenol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d89e93c4-a4c9-46a9-947f-eb61e0e4dec6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The transcript derived substrate uptake rate is -6.3347\n",
      "The transcript derived growth rate is 0.5249\n",
      "The scaled substrate uptake rate is -1.3521\n",
      "The scaled growth rate is 0.1120\n",
      "\n",
      "The transcript derived substrate uptake rate is -5.0643\n",
      "The transcript derived growth rate is 0.4244\n",
      "The scaled substrate uptake rate is -1.3521\n",
      "The scaled growth rate is 0.1133\n",
      "\n",
      "The transcript derived substrate uptake rate is -7.3496\n",
      "The transcript derived growth rate is 0.5481\n",
      "The scaled substrate uptake rate is -1.3521\n",
      "The scaled growth rate is 0.1008\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WT-P-R1</th>\n",
       "      <th>WT-P-R2</th>\n",
       "      <th>WT-P-R3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GUADEM</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tag_production</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_tag</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Phenol</th>\n",
       "      <td>0.11204</td>\n",
       "      <td>0.113316</td>\n",
       "      <td>0.100822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Glucose</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                WT-P-R1   WT-P-R2   WT-P-R3\n",
       "GUADEM          0.00000  0.000000  0.000000\n",
       "tag_production  0.00000  0.000000  0.000000\n",
       "EX_tag          0.00000  0.000000  0.000000\n",
       "Growth_Phenol   0.11204  0.113316  0.100822\n",
       "Growth_Glucose  0.00000  0.000000  0.000000"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculated in notebook E (units mmol glucose / g dry cell weight/ hr)\n",
    "wt_phenol_uptake_rate = 1.352072\n",
    "\n",
    "# this could be made into a loop for other normalization methods\n",
    "data_source = 'phenol_cpm'\n",
    "\n",
    "# define dataframe to hold WT phenol E-Flux2 flux values\n",
    "wt_phenol_eflux2_df = pd.DataFrame()\n",
    "\n",
    "# loop over WT glucose trials\n",
    "for trial in ['WT-P-R1', 'WT-P-R2', 'WT-P-R3']:\n",
    "    \n",
    "    # for each trial, calculate the fluxes\n",
    "    trial_fluxes = eflux2(reaction_transcripts[data_source][trial], 'phenol', wt_phenol_uptake_rate)\n",
    "    \n",
    "    # add flux series to results dataframe\n",
    "    wt_phenol_eflux2_df[trial] = trial_fluxes\n",
    "    \n",
    "# show shape of dataframe\n",
    "wt_phenol_eflux2_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a60f47-3aaa-4e79-af82-58b471c90b0a",
   "metadata": {},
   "source": [
    "#### Calculate the average and standard deviation of WT phenol E-Flux2 fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5791f076-3397-497a-9232-e15c36bbaafc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WT-P-R1</th>\n",
       "      <th>WT-P-R2</th>\n",
       "      <th>WT-P-R3</th>\n",
       "      <th>WT-P-Average</th>\n",
       "      <th>WT-P-Std-Dev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GUADEM</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tag_production</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_tag</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Phenol</th>\n",
       "      <td>0.11204</td>\n",
       "      <td>0.113316</td>\n",
       "      <td>0.100822</td>\n",
       "      <td>0.108726</td>\n",
       "      <td>0.006875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Glucose</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                WT-P-R1   WT-P-R2   WT-P-R3  WT-P-Average  WT-P-Std-Dev\n",
       "GUADEM          0.00000  0.000000  0.000000      0.000000      0.000000\n",
       "tag_production  0.00000  0.000000  0.000000      0.000000      0.000000\n",
       "EX_tag          0.00000  0.000000  0.000000      0.000000      0.000000\n",
       "Growth_Phenol   0.11204  0.113316  0.100822      0.108726      0.006875\n",
       "Growth_Glucose  0.00000  0.000000  0.000000      0.000000      0.000000"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define blank lists to hold average and standard deviation data\n",
    "average_col = []\n",
    "std_dev_col = []\n",
    "\n",
    "# loop over the rows of dataframe to fill the lists\n",
    "for _, row in wt_phenol_eflux2_df.iterrows():\n",
    "    \n",
    "    # add average and standard deviation of each row the the lists\n",
    "    average_col.append((row['WT-P-R1'] + row['WT-P-R2'] + row['WT-P-R3']) / 3)\n",
    "    std_dev_col.append(stdev([row['WT-P-R1'], row['WT-P-R2'], row['WT-P-R3']]))\n",
    "    \n",
    "# add the lists as columns to the dataframe\n",
    "wt_phenol_eflux2_df['WT-P-Average'] = average_col\n",
    "wt_phenol_eflux2_df['WT-P-Std-Dev'] = std_dev_col\n",
    "\n",
    "wt_phenol_eflux2_df.tail(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c757e5e8-6e06-4f8a-97c5-d32d92447ae7",
   "metadata": {},
   "source": [
    "#### Calculate E-Flux2 fluxes for PVHG phenol\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4f9b8b07-61cd-4163-9e1a-7efd5c884d17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The transcript derived substrate uptake rate is -6.3685\n",
      "The transcript derived growth rate is 0.5250\n",
      "The scaled substrate uptake rate is -1.9675\n",
      "The scaled growth rate is 0.1622\n",
      "\n",
      "The transcript derived substrate uptake rate is -3.2695\n",
      "The transcript derived growth rate is 0.2813\n",
      "The scaled substrate uptake rate is -1.9675\n",
      "The scaled growth rate is 0.1693\n",
      "\n",
      "The transcript derived substrate uptake rate is -5.7536\n",
      "The transcript derived growth rate is 0.4773\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The scaled substrate uptake rate is -1.9675\n",
      "The scaled growth rate is 0.1632\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PVHG-P-R1</th>\n",
       "      <th>PVHG-P-R2</th>\n",
       "      <th>PVHG-P-R3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GUADEM</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tag_production</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_tag</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Phenol</th>\n",
       "      <td>0.1622</td>\n",
       "      <td>0.169269</td>\n",
       "      <td>0.163222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Glucose</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                PVHG-P-R1  PVHG-P-R2  PVHG-P-R3\n",
       "GUADEM             0.0000   0.000000   0.000000\n",
       "tag_production     0.0000   0.000000   0.000000\n",
       "EX_tag             0.0000   0.000000   0.000000\n",
       "Growth_Phenol      0.1622   0.169269   0.163222\n",
       "Growth_Glucose     0.0000   0.000000   0.000000"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculated in notebook E (units mmol glucose / g dry cell weight/ hr)\n",
    "pvhg_phenol_uptake_rate = 1.967485\n",
    "\n",
    "# this could be made into a loop for other normalization methods\n",
    "data_source = 'phenol_cpm'\n",
    "\n",
    "# define dataframe to hold PVHG phenol E-Flux2 flux values\n",
    "pvhg_phenol_eflux2_df = pd.DataFrame()\n",
    "\n",
    "# loop over PVHG glucose trials\n",
    "for trial in ['PVHG-P-R1', 'PVHG-P-R2', 'PVHG-P-R3']:\n",
    "    \n",
    "    # for each trial, calculate the fluxes\n",
    "    trial_fluxes = eflux2(reaction_transcripts[data_source][trial], 'phenol', pvhg_phenol_uptake_rate)\n",
    "    \n",
    "    # add flux series to results dataframe\n",
    "    pvhg_phenol_eflux2_df[trial] = trial_fluxes\n",
    "    \n",
    "# show shape of dataframe\n",
    "pvhg_phenol_eflux2_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e407092b-7a93-48c8-84cc-b08fca9b3dd7",
   "metadata": {},
   "source": [
    "#### Calculate the average and standard deviation of PVHG phenol E-Flux2 fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c81b537e-843f-40d4-b352-cfd8097a01d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PVHG-P-R1</th>\n",
       "      <th>PVHG-P-R2</th>\n",
       "      <th>PVHG-P-R3</th>\n",
       "      <th>PVHG-P-Average</th>\n",
       "      <th>PVHG-P-Std-Dev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GUADEM</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tag_production</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_tag</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Phenol</th>\n",
       "      <td>0.1622</td>\n",
       "      <td>0.169269</td>\n",
       "      <td>0.163222</td>\n",
       "      <td>0.164897</td>\n",
       "      <td>0.00382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Glucose</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                PVHG-P-R1  PVHG-P-R2  PVHG-P-R3  PVHG-P-Average  \\\n",
       "GUADEM             0.0000   0.000000   0.000000        0.000000   \n",
       "tag_production     0.0000   0.000000   0.000000        0.000000   \n",
       "EX_tag             0.0000   0.000000   0.000000        0.000000   \n",
       "Growth_Phenol      0.1622   0.169269   0.163222        0.164897   \n",
       "Growth_Glucose     0.0000   0.000000   0.000000        0.000000   \n",
       "\n",
       "                PVHG-P-Std-Dev  \n",
       "GUADEM                 0.00000  \n",
       "tag_production         0.00000  \n",
       "EX_tag                 0.00000  \n",
       "Growth_Phenol          0.00382  \n",
       "Growth_Glucose         0.00000  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define blank lists to hold average and standard deviation data\n",
    "average_col = []\n",
    "std_dev_col = []\n",
    "\n",
    "# loop over the rows of dataframe to fill the lists\n",
    "for _, row in pvhg_phenol_eflux2_df.iterrows():\n",
    "    \n",
    "    # add average and standard deviation of each row the the lists\n",
    "    average_col.append((row['PVHG-P-R1'] + row['PVHG-P-R2'] + row['PVHG-P-R3']) / 3)\n",
    "    std_dev_col.append(stdev([row['PVHG-P-R1'], row['PVHG-P-R2'], row['PVHG-P-R3']]))\n",
    "    \n",
    "# add the lists as columns to the dataframe\n",
    "pvhg_phenol_eflux2_df['PVHG-P-Average'] = average_col\n",
    "pvhg_phenol_eflux2_df['PVHG-P-Std-Dev'] = std_dev_col\n",
    "\n",
    "pvhg_phenol_eflux2_df.tail(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a7c0b1-d01a-49f2-97db-077c7d069dca",
   "metadata": {},
   "source": [
    "#### Calculate E-Flux2 fluxes for WT glucose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed46399c-c6cc-446c-8445-07a8ef8bd87c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The transcript derived substrate uptake rate is -6.0820\n",
      "The transcript derived growth rate is 0.4762\n",
      "The scaled substrate uptake rate is -3.1659\n",
      "The scaled growth rate is 0.2805\n",
      "\n",
      "The transcript derived substrate uptake rate is -1.5991\n",
      "The transcript derived growth rate is 0.1619\n"
     ]
    }
   ],
   "source": [
    "# calculated in notebook E (units mmol glucose / g dry cell weight/ hr)\n",
    "wt_glucose_uptake_rate = 3.582471\n",
    "\n",
    "# this could be made into a loop for other normalization methods\n",
    "data_source = 'glucose_cpm'\n",
    "\n",
    "# define dataframe to hold WT glucose E-Flux2 flux values\n",
    "glucose_eflux2_df = pd.DataFrame()\n",
    "\n",
    "# loop over WT glucose trials\n",
    "for trial in ['WT-G-R1', 'WT-G-R2', 'WT-G-R3']:\n",
    "    \n",
    "    # for each trial, calculate the fluxes\n",
    "    trial_fluxes = eflux2(reaction_transcripts[data_source][trial], 'glucose', wt_glucose_uptake_rate)\n",
    "    \n",
    "    # add flux series to results dataframe\n",
    "    glucose_eflux2_df[trial] = trial_fluxes\n",
    "    \n",
    "# show shape of dataframe\n",
    "glucose_eflux2_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92134b98-7d96-4eb6-a6c9-fb733c787760",
   "metadata": {},
   "source": [
    "#### Calculate the average and standard deviation of WT glucose E-Flux2 fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6984b3fc-ba15-4801-8c09-c3c45ea6278d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define blank lists to hold average and standard deviation data\n",
    "average_col = []\n",
    "std_dev_col = []\n",
    "\n",
    "# loop over the rows of dataframe to fill the lists\n",
    "for _, row in glucose_eflux2_df.iterrows():\n",
    "\n",
    "    average_col.append((row['WT-G-R1'] + row['WT-G-R2'] + row['WT-G-R3']) / 3)\n",
    "    std_dev_col.append(stdev([row['WT-G-R1'], row['WT-G-R2'], row['WT-G-R3']]))\n",
    "    \n",
    "# add the lists as columns to the dataframe\n",
    "glucose_eflux2_df['WT-G-Average'] = average_col\n",
    "glucose_eflux2_df['WT-G-Std-Dev'] = std_dev_col\n",
    "\n",
    "glucose_eflux2_df.tail(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "980ae63b-5759-4d4b-b2aa-7621b00c9373",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biodesign_3.7",
   "language": "python",
   "name": "biod_3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
