{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7563604a-2ba5-4e5d-b9e0-309615df0b5b",
   "metadata": {},
   "source": [
    "# **Notebook to make predictions with reprocessed Yoneda and Henson data and EFLUX2 and SPOT** "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a6edbe7-d3d0-49ac-9d13-8f43882945f0",
   "metadata": {},
   "source": [
    "### **Import python packages**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5e37f811-1a95-45e0-932f-962edb144307",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cobra\n",
    "from statistics import mean, median\n",
    "%matplotlib inline\n",
    "from matplotlib import pyplot as plt\n",
    "from edd_utils import login, export_study, export_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e20d0fe-15c8-4248-b6be-aa65f868fc95",
   "metadata": {},
   "source": [
    "### **Import Joonhoons Code**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "faf7a9d4-095c-42c6-a32b-5b5fd19ffca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OptlangDir = '../EFLUX2_SPOT'\n",
    "# sys.path.append(OptlangDir)\n",
    "# from spot_cplex import SPOT #spot not working yet because quadratic constraints not supported yet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "244d4ba7-9d90-4f21-8de8-18a77f1a87dd",
   "metadata": {},
   "source": [
    "### **Sign into EDD**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e8232d35-b2bf-4b51-9625-3a12a76661f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Password for garrettroell:  ················\n"
     ]
    }
   ],
   "source": [
    "study_slug = 'biodesign_henson2018_reprocessed'\n",
    "edd_server = 'public-edd.jbei.org'\n",
    "user = 'garrettroell' # 'schenkch'\n",
    "session = login(edd_server=edd_server, user=user)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8bfde81-2fb8-4b07-9273-4070bc5fea43",
   "metadata": {},
   "source": [
    "### **Get transcriptomic data from [EDD](https://public-edd.jbei.org/s/biodesign_henson2018_reprocessed/)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f177a0d5-a877-4bd0-af30-304b0badbc93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3d7a4ca1e90a4d558d786cc5da15d63a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=440493.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study ID</th>\n",
       "      <th>Study Name</th>\n",
       "      <th>Line ID</th>\n",
       "      <th>Line Name</th>\n",
       "      <th>Line Description</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Assay ID</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Formal Type</th>\n",
       "      <th>Measurement Type</th>\n",
       "      <th>Compartment</th>\n",
       "      <th>Units</th>\n",
       "      <th>Value</th>\n",
       "      <th>Hours</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14905</td>\n",
       "      <td>WT-M-R1</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14948</td>\n",
       "      <td>WT-M-R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14905</td>\n",
       "      <td>WT-M-R1</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14948</td>\n",
       "      <td>WT-M-R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14906</td>\n",
       "      <td>WT-M-R2</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14949</td>\n",
       "      <td>WT-M-R2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14906</td>\n",
       "      <td>WT-M-R2</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14949</td>\n",
       "      <td>WT-M-R2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14907</td>\n",
       "      <td>WT-M-R3</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14950</td>\n",
       "      <td>WT-M-R3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Study ID                        Study Name  Line ID Line Name  \\\n",
       "0     14409  Biodesign_Henson2018_reprocessed    14905   WT-M-R1   \n",
       "1     14409  Biodesign_Henson2018_reprocessed    14905   WT-M-R1   \n",
       "2     14409  Biodesign_Henson2018_reprocessed    14906   WT-M-R2   \n",
       "3     14409  Biodesign_Henson2018_reprocessed    14906   WT-M-R2   \n",
       "4     14409  Biodesign_Henson2018_reprocessed    14907   WT-M-R3   \n",
       "\n",
       "                                    Line Description         Protocol  \\\n",
       "0  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "1  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "2  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "3  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "4  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "\n",
       "   Assay ID Assay Name Formal Type Measurement Type  Compartment Units  Value  \\\n",
       "0     14948    WT-M-R1         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "1     14948    WT-M-R1         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "2     14949    WT-M-R2         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "3     14949    WT-M-R2         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "4     14950    WT-M-R3         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "\n",
       "   Hours  \n",
       "0   20.0  \n",
       "1   32.0  \n",
       "2   20.0  \n",
       "3   32.0  \n",
       "4   20.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = export_study(session, study_slug, edd_server=edd_server)\n",
    "trans_df = df[df['Protocol'].str.contains('Transcriptomics')]\n",
    "trans_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7252a93-7cea-4b7c-8375-c03a194a9524",
   "metadata": {},
   "source": [
    "### **Load Genome Scale Model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f6b96da-3aaf-49fd-9dec-b674a678d243",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name =  '../GSMs/Ropacus_annotated_curated_with_phenol_custom_biomass.xml'\n",
    "model = cobra.io.read_sbml_model(file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c050bfe4-28bd-42d3-a42d-7eb33402d132",
   "metadata": {},
   "source": [
    "### **Define functions used in main EFlux function**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f157a170-aa60-403a-9d41-903bb8b454d4",
   "metadata": {},
   "source": [
    "Define function to plot histogram from list of numbers. Will be used for transcriptomic measurements and flux bounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b39e769d-8315-4adb-b033-7b5dbc6a8cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_histogram(data, title, maximum=400, bin_size=5):\n",
    "\n",
    "    data = [value for value in data if value != np.Inf]\n",
    "    bins = np.arange(0, maximum, bin_size) # x axis goes from 0 to 40 with bars of width 5\n",
    "    plt.hist(data, bins=bins, alpha=0.5) \n",
    "    plt.title(f'{title}\\nmean = {mean(data):.3f}, median={median(data):.3f}, max={max(data):.3f}')\n",
    "    plt.xlabel('Value')\n",
    "    plt.ylabel('count')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a779be49-3fc2-4cf7-a1e2-00df3b5ddb89",
   "metadata": {},
   "source": [
    "Define a function to find the transcriptomic value for a given reaction id. The function takes into account genes that are required in parallel and in series."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e82e91bd-a148-4bdd-a4d3-fb4a1fc5bc60",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rxn_id_to_transcript_level(model, trans_df, rxn_id):\n",
    "    rxn = model.reactions.get_by_id(rxn_id)\n",
    "    gpr = rxn.gene_reaction_rule\n",
    "    parallel_transcript_values = []\n",
    "    \n",
    "    # A reaction without transcript measurements should not be constrained\n",
    "    if gpr == '' or gpr == 'spontaneous':\n",
    "        return np.Inf\n",
    "    else:\n",
    "        for parallel_gene_string in gpr.split(' or '):\n",
    "            parallel_gene_string = parallel_gene_string.strip('() ')\n",
    "            \n",
    "            series_transcript_values = []\n",
    "            for gene_name in parallel_gene_string.split(' and '):\n",
    "                gene_name = gene_name.strip('() ')\n",
    "                if gene_name == 'spontaneous':\n",
    "                    series_transcript_values.append(np.Inf) # A spontaneous reaction should not require transcripts\n",
    "                else:\n",
    "                    transcript_value = float(trans_df[trans_df['Measurement Type'] == gene_name].Value)\n",
    "                    series_transcript_values.append(transcript_value)\n",
    "            \n",
    "            parallel_transcript_values.append(min(series_transcript_values))\n",
    "            \n",
    "        return sum(parallel_transcript_values)\n",
    "#     else:\n",
    "#         return np.Inf # Reactions without transcript measurements shouldn't be constrained"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32f10d13-e30e-4d78-bc3c-3e386c47a181",
   "metadata": {},
   "source": [
    "### **Main EFlux function**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "f38895b6-054a-404e-a293-56847f42a1f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from optlang.symbolics import add\n",
    "\n",
    "def EFlux2test(\n",
    "    model, \n",
    "    edd_df, \n",
    "    line_name, \n",
    "    min_bound=10, \n",
    "    transcript_scaling_factor = 1, \n",
    "    infinite_bound_reactions=[], \n",
    "    zero_bound_reactions=[]\n",
    "    ):\n",
    "    \n",
    "    # get transcriptomics measurement for a single experiment and plot values\n",
    "    trans_df = edd_df[edd_df['Line Name'] == line_name]\n",
    "    plot_histogram(trans_df['Value'], 'Transcriptomic values by genes', maximum=400, bin_size=5)\n",
    "    \n",
    "    # make a dataframe mapping reaction to transcript levels using cobrapy's gene_reaction_rule\n",
    "    reaction_df = pd.DataFrame()\n",
    "    reaction_df['reaction_id'] = [rxn.id for rxn in model.reactions]\n",
    "    reaction_df['transcipt_level'] = [rxn_id_to_transcript_level(model, trans_df, rxn_id) for rxn_id in reaction_df.reaction_id]\n",
    "    plot_histogram(reaction_df['transcipt_level'], 'Transcriptomic values by reactions', maximum=400, bin_size=5)\n",
    "    \n",
    "    # calculate maximum biomass yield without constraints\n",
    "    with model:      \n",
    "        # Constrain model according to parameter lists\n",
    "        for rxn_id in infinite_bound_reactions:\n",
    "            model.reactions.get_by_id(rxn_id).bounds = (-np.Inf, np.Inf)\n",
    "        for rxn_id in zero_bound_reactions:\n",
    "            model.reactions.get_by_id(rxn_id).bounds = (-1e-9, 1e-9)\n",
    "        \n",
    "        # Ensure that medium is correct, and the carbon source is the only limiting substrate\n",
    "        medium = {key:1000 for (key,value) in model.medium.items()}\n",
    "        medium[\"EX_glc__D_e\"] = 0.0\n",
    "        medium[\"EX_phenol_e\"] = 100.0 # 1.76124\n",
    "        model.medium = medium\n",
    "\n",
    "        # Run unbounded model\n",
    "        model.tolerance = 1e-9\n",
    "        unbounded_sol = model.optimize()\n",
    "        print(f'''\\nWithout constraints, flux balance analysis found \n",
    "            that {unbounded_sol.objective_value:.3f} grams of biomass \n",
    "            could be made.\\n''')\n",
    "        \n",
    "    # calculate maximum biomass yield with constraints\n",
    "    bounded_model = model.copy()\n",
    "    with bounded_model:\n",
    "\n",
    "        # Add transcriptomic bounds to model\n",
    "        for rxn in model.reactions:\n",
    "            if rxn.id in infinite_bound_reactions:\n",
    "                bounded_model.reactions.get_by_id(rxn.id).bounds = (-np.Inf, np.Inf)\n",
    "            elif rxn.id in zero_bound_reactions:\n",
    "                bounded_model.reactions.get_by_id(rxn.id).bounds = (-1e-9, 1e-9)\n",
    "            elif rxn in bounded_model.boundary:\n",
    "                pass\n",
    "            else:\n",
    "                # can replace this line with a reaction_df dataframe method\n",
    "                bound = transcript_scaling_factor * rxn_id_to_transcript_level(bounded_model, trans_df, rxn.id) \n",
    "                bound = max(bound, min_bound) # min_bound is a parameter of this function\n",
    "                if rxn.upper_bound > 0:\n",
    "                    bounded_model.reactions.get_by_id(rxn.id).upper_bound = bound\n",
    "                if rxn.lower_bound < 0:\n",
    "                    bounded_model.reactions.get_by_id(rxn.id).lower_bound = -1*bound\n",
    "\n",
    "        # Ensure that medium is correct, and the carbon source is the only limiting substrate\n",
    "        medium = {key:np.Inf for (key,value) in bounded_model.medium.items()}\n",
    "        medium[\"EX_glc__D_e\"] = 0.0\n",
    "        medium[\"EX_phenol_e\"] = 100.0 # 1.76124\n",
    "        bounded_model.medium = medium\n",
    "        \n",
    "        # Add constraints from bounded model to the reaction dataframe\n",
    "        reaction_df['upper_bound'] = [bounded_model.reactions.get_by_id(rxn_id).upper_bound for rxn_id in reaction_df.reaction_id]\n",
    "        reaction_df['lower_bound'] = [bounded_model.reactions.get_by_id(rxn_id).lower_bound for rxn_id in reaction_df.reaction_id]\n",
    "\n",
    "        # Run bounded model\n",
    "        model.tolerance = 1e-9\n",
    "        bounded_sol = bounded_model.optimize()\n",
    "        print(f'''\\n\n",
    "            Given the bounds from transcriptmics and a minimum absolute bound of {min_bound} and \n",
    "            a transcript scaling factor of {transcipt_scaling_factor} flux balance analysis found \n",
    "            that {bounded_sol.objective_value:.3f} grams of biomass could be made.\\n''')\n",
    "            \n",
    "    # Add fluxes from unbound and bound solutions\n",
    "    reaction_df['unbound_flux'] = [unbounded_sol[rxn_id] for rxn_id in reaction_df.reaction_id]\n",
    "    reaction_df['EFlux_flux'] = [bounded_sol[rxn_id] for rxn_id in reaction_df.reaction_id]\n",
    "    reaction_df.sort_values('unbound_flux', axis=0, ascending=False, inplace=True)\n",
    "    display(reaction_df.head(50))\n",
    "    \n",
    "    \n",
    "#         # Constrain the biomass to the optimal value\n",
    "#             print('FBA status', bounded_sol.status)\n",
    "#             print('FBA solution', bounded_sol.objective_value)\n",
    "#             display(bounded_model.objective)\n",
    "#             # Constrain the biomass to the optimal value\n",
    "#             for rxn_id in bounded_model.reactions:\n",
    "#                 if rxn_id.objective_coefficient:\n",
    "#                     rxn_id.lower_bound = bounded_sol.objective_value\n",
    "#             # minimize the sum of squared flux values\n",
    "#             bounded_model.objective = bounded_model.problem.Objective(add([rxn_id.flux_expression**2 for rxn_id in bounded_model.reactions]), direction='min')\n",
    "#             display(bounded_model.problem.Objective)\n",
    "#             eflux2_sol = bounded_model.optimize()\n",
    "#             print('EFlux2 status', eflux2_sol.status)\n",
    "#             print('EFlux2 solution', eflux2_sol.objective_value)\n",
    "#     return eflux2_sol"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f441bd3-e7cf-469c-a0a2-db836dd8818f",
   "metadata": {},
   "source": [
    "### **Main EFlux function test**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "1ca8506a-4510-4ddc-b388-b1e3c1ec68b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAElCAYAAAAV9s4VAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3debwcVZn/8c+XhDVsCYkBksAFCWJExRhZVDQjyiqCIz8niBIRxGFgxF3EhaiojDOIggiiIouyiQuLMBhRVMZBCAqBsEYWSQwkEIEEHQR8fn+c06TS6b6nb3J7ucn3/Xr161adU8vT1bf7qTpVdUoRgZmZWX/W6nYAZmbW+5wszMysyMnCzMyKnCzMzKzIycLMzIqcLMzMrMjJwoYMSZ+WdGa34wCQ9DNJh3RwfUdIuq5T68vrvF7Suzu5Tutdw7sdgHWepKWV0Q2Ap4Hn8vj7IuL7nY+qLCI+3+q0kk4ExkfEu9sUy57tWK5Zr3KyWANFxIa1YUkPAEdExM+bTS9peEQ824nYejkGszWZm6FsBZJOlHSxpAslLQHeKWk3STdIelzSAkmnSlo7Tz9cUkh6n6S5kv4i6dTK8raX9GtJT0h6VNIFlbqXSvq5pMWSHpb0sX5iOFHSObl+u7zO90r6c359MNe9GfgYcIikpZJuzuXjJV2Z13WvpPfUveeL8vqWSrpV0gslfUrSIkl/kvTGyvTLNdHk936XpCWSbpf08gbb9VuSTqor+6mk9+fhT0m6Ly9jjqS3NPl8tpMUdWX18RyR4/mLpKslTcjla+XPbmH+PGZLmtTwHyGZKGlWnvbHkkbm5Vwj6ai6GO6QtH+TmA/L2/BRScdLmidpaiWm4yX9MddfVFlP7XM+NM+zSNJxleX2N+8Gki6Q9Fj+v71R0uh+3qv1JyL8WoNfwAPAG+vKTgT+DuxP2qFYH3gVsAvpaHRb4B7gmDz9cCCAy4BNgD5gcW25wA+Aj+dlrQe8JpdvAjwCHAusC2wM7NxPDCcC5+T67fI6zyc1pb0ceAyYWpn/nLr39T/AaTmGycCjwOsr0/8NeGN+PxcA9wPH5fGjgHsry7oeeHcePhh4CHglIGB7YEKDbf2GvL2VxzfL6xybx98ObJHf7zuApZW6I4Drqu+9btnVeN4G3A28KMc+A/hNrtsPuDFv+7WAScDmTf43rs/vaxIwAvhJZfu/A/ifyrSvBBYCwxss56XAEuDV+XM+BXi28ll9OH824/Jn823g/LrP+czK5/Y0MLGFeY/OMa8PDAOmABt2+zs3VF9dD8CvLv8DNE8WvyjM9xHgB3m4lix2rdT/CPhIHr4AOAMYV7eMdwE3NVn+CjHQOFlsV6n/CvDN+mnz+DbAM8CIStl/At+uTH91pe6twBPAWnl8ZF7fhnm8+uN8LXB0C9t6LWA+8Oo8fhTws36mvx3YLw8PJFnMBKZX6obnH9hxwJ7AXaTEv1Yh3uuBEyvjLwP+j5QQ1wceB7bJdV8FTm2ynM/VfsDz+AiWTxb3kpN2Hp+Q17NW5XPevFL/e+CgFuY9Mr+Hl3b7e7Y6vNwMZc08VB2RtENuMnlY0pOkH4D6Q/qHK8N/BWrnRj4MrA3MknSbpOm5fALwx1ZjaGGaB4Etm0y3JfBoRDxVN/24yvgjleG/AYsi4h+VcVj2nqpK7wOAvKyLSUcikPbOn7+YQNK7c/PX45IeB3ZgxW3ciq2B0yvLeRT4B+mE/89Ie+lnAI9IOlPSRv0sq377rguMioi/AZeSmgeHAdNIR3mNbFldTv4M/lKp3wq4ohLvbbn8BZV5mv1v9TfvOcDPgUskzZd0kiSfp11JThbWTH13xN8k7eluFxEbA58h7WGWFxSxICKOiIgtSE0DZ0nahvQD8sIBxNDIhMrwVsCfm8z7Z2C0pBF1089vYR0lpfdRdSHw//L7n0w6AkPStqQf8KOAzSJiU9IRQKNt/FSeZ4NK2eZ18RweEZtWXutHxO8AIuKrETEZ2JHUxPShfuKt375Pk5oYAc4FDiEdrfwlIm5qsowFwPjaSP4MRlbq5wFvqot3vboE0UzTeSPi7xExIyJeDLyWdLTYscudVzdOFtaqjUjNMk9JejHwvlZnlPR2SbU9+MdJP+TPAZcDW0k6RtK6kjaWtPMA4/q0pPUlvRSYTtpzh3SU0CdJABFxPzAL+GJe107AYcD3Bri+Rr4NfEzSK5RMrJ1Qrpd/UJ8EzgKuiogluWpD0nZZBEjSe0lHFo08nF/vlDRM0pGko4maM4FP5s8JSZtKOigP75xfw0lJ5++ko45mDs1HlSOAzwKXRG7vITXxrAP8B82PKiCdszpQ0q6S1iEdlVadSfpctsoxvqDZyf0Gms4r6Q2SdpS0FmmbP1N4r9YPJwtr1YdJP8ZLSEcZF/c/+XJ2AW6S9BRpT/roiPhTRDwBvIl0QvYR0knz1w8wruuB+4CfAV+KiF/k8otJP2SLJd2Yy/4FmEj6ob0UOD4irhvg+lYQEReSfjAvJv0o/Yjl95zrXUg6kf78VWERMZt08v1G0p74i4DfNVlfAO8Fjic1MW1XnTYifkA6f/OD3GQ4G9grV28KfIeUtB/I6/pKP7GeT0qoC0gniT9QF8d5pCOUpvfm5Pf2QVLS+DPpQoTHSEcp5PX/N3Ct0pVvvyVdUNGK/ubdkvRZPAnMITVJXdBoIVamZTsJZkOHpO1IVye11BRm7aF0+fGhETF1APNsTEpWW0dEK+elrAf4yMLMVkpumvo3UpNaadq35PseNgROBn7vRDG0OFmY2YBJ2o90X8WfaK1J8q2kJqh5pPtwDu53aus5boYyM7MiH1mYmVmRk4VZnXxz3PWV8aX5PgizNZaTxWoo37cwS9LTyh3vVep2lTRTqTO9RZJ+IGmLSv26+a7eR/I0V1TukWi0rp0k3Szpr/nvTpW6jyp1qrdE0v2SPtpg/mNz3VOS7pS0/SBthkETERtGxH2dWp+k6XlbPqnUed6Xq3ceSxql1KnfU5IelPSOfpb1QaXOCZ9U6mzxlLpl9Un6Zf787tLynSWemRNl7fV0vjx1yMn/19/J22uJpFsk7VOpL30vZkh6pm57bFup3z//ry+V9Fs16ZxR0rVKHSM2vJO8hTi69p1yslg9/ZnU19HZDepGkq5e6SPdyLUE+G6l/lhgN1I/QFuSumU4rdFKlG6wuox0Hf5I0h29l+VySHcfH5rr9gaOkTStMv8RwOGkzu02BN5Mum9gTbcB6X6G0aR7VPYg9cVVczrpZrqxpDuSz5D0kibLuhyYnO+635HU4eL7K/UXAn8gdWr4SeBSSWMAIuJfc6LcMFK39heS7pUYioaT7mx/PakTxU+RugHpy/Wl7wXAxdXtUduBkDSRdJ/Jv5LuY7kCuLw+ISg9LGvtQpylOLr3nep251S98CLdnPRR0s1LT5FuWhoLXJ0/rJ8DIyvT70q6+edx4FZyh2i57jDgzjzffaSHCdXqppKuBvkw6UqSBcBhbXxfK/S82mCaycCSyvgZwJcr4/sBdzeZd09SdxmqlP0J2LvJ9KcCp+XhtUhf3j1W8r1dl9/fb0m9s15B+sH7PukmrJuAvsr0O5A62FtM6pH17ZW6zUg/qk+Sbor7PHB9pf75Dgvz9vhDnvYhYEZlur487fS8HR4FPjkIn+OHgCvy8AhSoti+Un8+cFILy9ks/y9/I49vT7oxbqPKNL8B/rXBvCPy//TrW4y59r/+scr/+oHAvqSbLxeTboqsTb8z8L/5O7UA+DqwTq57dd6WE/L4y0k7MTus4nadDbytxe/FDOB7TaY9BvhpZXwtUl9ie1TKNsnve9f8P7JC77ytxNHO71Tp5SOLZd5Gupt4e1K32FeT7pAdQ/oQas8cGAf8lPRDNYq0x/fD2t4Y6YvxZlJ324cBp0iaXFnP5qR/nHGkPYDTlfvfryfpG8odpDV4zR6k9/060t2tNd8BXiNpS6W+hw4hbYtGXgLMjvyfms3O5fXvRcDulXWNz68dJT2UD5s/q9Q1Q6umkXquHUfqm+l/SXtho0gJ+4S87hGkRHEBqYO5acA3Kk0Fp5N6Kt0CeE9+NfMUac9uU1LiOErSgXXTvJZ0B/YewGe0rNuNd/TzeT6u3GVFA9XPaHvg2Yi4p1J/Kw22eU1e75OkH9yXk+7AJ89zXyzrcqS/Zb2N1BXJr5utp4HNSd2GjyP1JfYt4J2k7sx3J3XVsk2e9jnSXd6jSUe2e5Du4SAifptjPlfS+qQj2U9HxF35/V3Zzza9ssk2GUvalnMa1bPi9wJg/9w8NEd1z/Jg+T68lF87Vsq+SNoRa6W/q1IcaSXt+U41144MNNRepCOLQyrjPwTOqIz/O/CTPPxxKt0t57JrqHQJXVf3E+DYPDyVtMcxvFK/kErX3oP8vvo9siA1NS0Gdq+UbQJcRNr7eZa0Fz2qyfyfBi6qK/s+lb3tSvlnST9E6+bxV+d1/JT0w9tH2vN6b4vv7Toqe+2kG72qXYzvD9ySh/+F/DyHSv03SclkGKnPoB0qdV+kyZFFgzi+CpySh/vytOMr9TcC01bhM3wPaQ99dB7fHXi4bpr3krsvLyxrIumoafM8/i7ghrppvtDof4bUDfsKn2s/66r9rw/L4xvlbbNLZZqbgQObzP8B4MeV8bXz9LeRuvdQq7E0WPbapCOsbw7gezGJ1Cw7LP/vLgAOznU7kHYippK6mPk0qQ+qT+T6KcAtpKaw2v9I8ciiURzt/E6VXj6yWKa+e+r68VqXyFuTeg19fu+FtCe5BYCkfZSeKLc41+3L8t1MPxbLPx602t1yxyh1l3E1KZH9plJ1Oqkb6s1ITQ8/ovmRxVLSEVTVxqTmiuq6jiHtje8XEbX+gGpdfn85Ih6PiAdIP+D7DuBtDOQz26XuMzuEtOc7hmXt2TUPNluhpF2UTggvkvQEqZ261a7aByQfsXwJ2Cciau3OLW3zRiLiXtJe6DcGsqx8xDOV1A/UQDwWEbVnu9c+74afkdLTFK/Usi7wv0hlu0bEM6Qux3cETo786zhQeS/7fFJT3jEN6ht+LyLijoj4c0Q8F+lI52vAQbnuLlLT49dJSWQ0cAcwL6/vG3l5LT8WuJ/vZ62+Xd+pppwsBu4h0pFFtUvkERFxkqR1SUcl/0V6wtmmwFW02JV3Pa14NUr11ezwuZXlbk3as/p8RNT3FroTac9ycf4nPA3YWY0fRzkHeFk+HK55GZXDZqW+g44jtaPOq0x3N+kLW/3St+sO0YeAX9V9ZhtGxFGkppVnWbEr7mYuIJ3fmBARm5B6PW3p85V0SD+f59JqM5SkvUnNNvtHxG2VxdwDDM8nVWteTvPmlHrDWdad+hxgWy3/PItGy3oX6al47bwi7AxSl+wTI52MP57Kds3NvyeQmhlPzt+1Wt3V/WzTqyvTiWXnI9+WExCV+v6+F/WiGl9EXBoRO0bEZjnOPtJ5s41JRxYXS3o4l0FKJLs3WnApjq59pwbj8GSov6h7WhypTXRGZfwI4Od5eAJpz3Ev0iHpeqS9rvGkQ+3nSFdcCNiHtGd5Yp53KjCvv3UP0vsZnuP6Emkvaj3yYS+p/fiP5KfYNZj3u6SEtwnpcP14YH6Tadch7YXXHot6TB6vnZg8JG+rFzeZ/zzgyrzdxpN+LA7PdX2kf/S+JvNeBxxRGa9/Mt4bgbl5eKMc17vye1qb1DPpi3P9xaSmtw1IzQ3zaH6CeyG5yZF0UnYh+cQnDZoY6uNs8fN7A6lX1tc1qb+IdGXSCOA1pK7jX9Jk2iOAF+ThSaRE8JVK/Q2knZv1SF1yPA6MqVvG3cB7Giz7HJo0c9b/r7PsaYp9lbLrgXfm4RtZ9oyUHfI6r891Ip1z+o88fA2VizAGsF3PzO93hUerUv5eHEC6Akn5c5/P8k8jfCXp92AMcAlwQSX2zSuvV+XtMI78PRlgHCv9nVrV16D9QA3lFwNIFnl8F+BXpPbERaQ2wq1y3dGkQ+3HST/UF9H5ZDEj/0NWXzNy3Ql5fGn1VZm3dkXRwvweric/FzvXX83yV7G8gtSW/DfS4y5fUam7n3Q+oLquMyv1G+fts4S09/8ZlnVBs3veNms3eY/X0WKyyOMvyp/TItIP8S+AnXLdmPwFa+VqqINIiWdJnufrDH6y+CXpaKe63arnY0aRzoU9Rbrq6h2Vut3rPs/v5v/Hp/L2/E9gvUp9X47xb6Qf6PpH7O6W592oQZzX0qQ9nIEni9eRftiWkq7I+hzLksWxpLb52k7IlvlzbNiW3ySerfP6/69uux7S4vfiwvx/szTH+f665V+f/ycWk5p+RjSJo9H/yJwBxLHS36lVfblvKOtJkj5FeqzpN4sTW8cp3UtzK/CyqGvOsdWTk4WZmRX5BLeZmRU5WZiZWZGThZmZFTXs+XCoGz16dPT19XU7DDOzIeXmm29+NCLGNKpbLZNFX18fs2bN6nYYZmZDiqSmvRe4GcrMzIqcLMzMrMjJwszMipwszMysyMnCzMyKnCzMzKzIycLMzIqcLMzMrMjJwszMilbLO7hX1Skz71lu/INv2r5LkZiZ9QYfWZiZWZGThZmZFTlZmJlZkZOFmZkVOVmYmVmRk4WZmRU5WZiZWZGThZmZFTlZmJlZkZOFmZkVOVmYmVmRk4WZmRU5WZiZWZGThZmZFTlZmJlZUduShaQJkn4p6Q5JcyQdm8tHSZop6d78d2Qul6RTJc2VNFvS5Mqypufp75U0vV0xm5lZY+08sngW+HBETAJ2BY6WNAk4Drg2IiYC1+ZxgH2Aifl1JHAGpOQCnADsAuwMnFBLMGZm1hltSxYRsSAifp+HlwB3AuOAA4Bz82TnAgfm4QOA8yK5AdhU0hbAXsDMiFgcEX8BZgJ7tytuMzNbUUfOWUjqA14B/A4YGxELctXDwNg8PA54qDLbvFzWrNzMzDqk7clC0obAD4EPRMST1bqICCAGaT1HSpoladaiRYsGY5FmZpa1NVlIWpuUKL4fET/KxY/k5iXy34W5fD4woTL7+FzWrHw5EXFWREyJiCljxowZ3DdiZraGa+fVUAK+A9wZEV+pVF0O1K5omg5cVik/NF8VtSvwRG6uugbYU9LIfGJ7z1xmZmYdMryNy34N8C7gNkm35LLjgZOASyQdDjwIvD3XXQXsC8wF/gocBhARiyV9HrgpT/e5iFjcxrjNzKxO25JFRFwPqEn1Hg2mD+DoJss6Gzh78KIzM7OB8B3cZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW5GRhZmZFbUsWks6WtFDS7ZWyGZLmS7olv/at1H1C0lxJd0vaq1K+dy6bK+m4dsVrZmbNtfPI4hxg7wblp0TETvl1FYCkScA04CV5nm9IGiZpGHA6sA8wCTg4T2tmZh00vF0LjohfS+prcfIDgIsi4mngfklzgZ1z3dyIuA9A0kV52jsGOVwzM+tHN85ZHCNpdm6mGpnLxgEPVaaZl8uala9A0pGSZkmatWjRonbEbWa2xup0sjgDeCGwE7AAOHmwFhwRZ0XElIiYMmbMmMFarJmZ0cZmqEYi4pHasKRvAVfm0fnAhMqk43MZ/ZSbmVmHdPTIQtIWldG3ArUrpS4HpklaV9I2wETgRuAmYKKkbSStQzoJfnknYzYzszYeWUi6EJgKjJY0DzgBmCppJyCAB4D3AUTEHEmXkE5cPwscHRHP5eUcA1wDDAPOjog57YrZzMwaa+fVUAc3KP5OP9N/AfhCg/KrgKsGMTQzMxsg38FtZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW5GRhZmZFThZmZlbkZGFmZkUtJQtJ17ZSZmZmq6d+H34kaT1gA9LT7kYCylUbA+PaHJuZmfWI0pPy3gd8ANgSuJllyeJJ4OttjMvMzHpIv8kiIr4GfE3Sv0fEaR2KyczMekxLz+COiNMkvRroq84TEee1KS4zM+shLSULSecDLwRuAZ7LxQE4WZiZrQFaShbAFGBSREQ7gzEzs97U6n0WtwObtzMQMzPrXa0eWYwG7pB0I/B0rTAi3tKWqMzMrKe0mixmtDMIMzPrba1eDfWrdgdiZma9q9WroZaQrn4CWAdYG3gqIjZuV2BmZtY7Wj2y2Kg2LEnAAcCu7QrKzMx6y4B7nY3kJ8BebYjHzMx6UKvNUP9cGV2LdN/F/7UlIjMz6zmtXg21f2X4WeABUlOUmZmtAVo9Z3FYuwMxM7Pe1erDj8ZL+rGkhfn1Q0nj2x2cmZn1hlZPcH8XuJz0XIstgStymZmZrQFaTRZjIuK7EfFsfp0DjGljXGZm1kNaTRaPSXqnpGH59U7gsXYGZmZmvaPVZPEe4O3Aw8AC4CDg3W2KyczMekyrl85+DpgeEX8BkDQK+C9SEjEzs9Vcq0cWL6slCoCIWAy8or8ZJJ2dr5y6vVI2StJMSffmvyNzuSSdKmmupNmSJlfmmZ6nv1fS9IG9PTMzGwytJou1aj/s8PyRRemo5Bxg77qy44BrI2IicG0eB9gHmJhfRwJnVNZzArALsDNwQjUOMzPrjFaTxcnA/0r6vKTPA78FvtzfDBHxa2BxXfEBwLl5+FzgwEr5ebnfqRuATSVtQep/amZELM5HNjNZMQGZmVmbtXoH93mSZgFvyEX/HBF3rMT6xkbEgjz8MDA2D48DHqpMNy+XNStfgaQjSUclbLXVVisRmpmZNdPqCW5ycliZBNFseSEpylO2vLyzgLMApkyZMmjLNTOzleiifBU9kpuXyH8X5vL5wITKdONzWbNyMzProE4ni8uB2hVN04HLKuWH5quidgWeyM1V1wB7ShqZT2zvmcvMzKyDWm6GGihJFwJTgdGS5pGuajoJuETS4cCDpBv9AK4C9gXmAn8FDoN0iW4+oX5Tnu5z+bJdMzProLYli4g4uEnVHg2mDeDoJss5Gzh7EEMzM7MB6nQzlJmZDUFOFmZmVuRkYWZmRU4WZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVtS2m/JWJ6fMvOf54Q++afsuRmJm1h0+sjAzsyInCzMzK3KyMDOzIicLMzMrcrIwM7MiJwszMytysjAzsyInCzMzK3KyMDOzIicLMzMrcrIwM7MiJwszMytysjAzsyInCzMzK3KyMDOzIicLMzMrcrIwM7MiJwszMytysjAzsyInCzMzK3KyMDOzIicLMzMrcrIwM7MiJwszMytysjAzsyInCzMzK3KyMDOzIicLMzMrcrIwM7OiriQLSQ9Iuk3SLZJm5bJRkmZKujf/HZnLJelUSXMlzZY0uRsxm5mtyYZ3cd3/FBGPVsaPA66NiJMkHZfHPw7sA0zMr12AM/Lfrjhl5j3LjX/wTdt3KRIzs87ppWaoA4Bz8/C5wIGV8vMiuQHYVNIW3QjQzGxN1a1kEcDPJN0s6chcNjYiFuThh4GxeXgc8FBl3nm5bDmSjpQ0S9KsRYsWtStuM7M1UreaoV4bEfMlvQCYKemuamVEhKQYyAIj4izgLIApU6YMaF4zM+tfV44sImJ+/rsQ+DGwM/BIrXkp/12YJ58PTKjMPj6XmZlZh3Q8WUgaIWmj2jCwJ3A7cDkwPU82HbgsD18OHJqvitoVeKLSXGVmZh3QjWaoscCPJdXWf0FE/Lekm4BLJB0OPAi8PU9/FbAvMBf4K3BY50M2M1uzdTxZRMR9wMsblD8G7NGgPICjOxCamZk10UuXzpqZWY9ysjAzsyInCzMzK+pmdx+rBXf/YWZrAh9ZmJlZkZOFmZkVOVmYmVmRk4WZmRU5WZiZWZGThZmZFTlZmJlZkZOFmZkV+aa8Qeab9MxsdeQjCzMzK3KyMDOzIjdDtZmbpcxsdeAjCzMzK3KyMDOzIicLMzMrcrIwM7Min+DuMJ/wNrOhyEcWZmZW5GRhZmZFbobqsmqzlJukzKxX+cjCzMyKfGTRQ3zy28x6lZPFEOJkYmbd4mYoMzMr8pFFD6s/kjAz6xYniyGslEzcTGVmg8XNUGZmVuQjizWUT5ab2UA4WazGnBDMbLA4WaxBfMLczFaWk4UBPlluZv1zsrCWDDSZuM8rs9WLk4UNiv6SSencyao2jzkZmbWfk4V1XDvPnbg5zaw9hkyykLQ38DVgGPDtiDipyyFZjxhI8hloouovuQz21WZuurNepojodgxFkoYB9wBvAuYBNwEHR8QdjaafMmVKzJo1a6XX56uGzFq3Ks2KQykprsrOwVC5jF3SzRExpWHdEEkWuwEzImKvPP4JgIj4UqPpnSzMbE21Komov2QxVJqhxgEPVcbnAbtUJ5B0JHBkHl0q6e5VWN9o4NFVmL9dejUu6N3YejUu6N3YejUu6N3YeiauD61YNJDYtm5WMVSSRVFEnAWcNRjLkjSrWXbtpl6NC3o3tl6NC3o3tl6NC3o3tl6NCwYvtqHSkeB8YEJlfHwuMzOzDhgqyeImYKKkbSStA0wDLu9yTGZma4wh0QwVEc9KOga4hnTp7NkRMaeNqxyU5qw26NW4oHdj69W4oHdj69W4oHdj69W4YLCa54fC1VBmZtZdQ6UZyszMusjJwszMipwsKiTtLeluSXMlHdcD8Twg6TZJt0ialctGSZop6d78d2QH4jhb0kJJt1fKGsah5NS8DWdLmtyF2GZImp+32y2S9q3UfSLHdrekvdoY1wRJv5R0h6Q5ko7N5V3fbv3E1tXtJmk9STdKujXH9dlcvo2k3+X1X5wvckHSunl8bq7va0dchdjOkXR/ZZvtlMs7/T0YJukPkq7M44O/zSLCr3TeZhjwR2BbYB3gVmBSl2N6ABhdV/Zl4Lg8fBzwHx2I43XAZOD2UhzAvsDVgIBdgd91IbYZwEcaTJcXmFcAAAUZSURBVDspf67rAtvkz3tYm+LaApichzcidVczqRe2Wz+xdXW75fe+YR5eG/hd3haXANNy+ZnAUXn434Az8/A04OI2brNmsZ0DHNRg+k5/Dz4EXABcmccHfZv5yGKZnYG5EXFfRPwduAg4oMsxNXIAcG4ePhc4sN0rjIhfA4tbjOMA4LxIbgA2lbRFh2Nr5gDgooh4OiLuB+aSPvd2xLUgIn6fh5cAd5J6Iuj6dusntmY6st3ye1+aR9fOrwDeAFyay+u3WW1bXgrsIUmDHVchtmY69nlKGg/sB3w7j4s2bDMni2UadSnS3xeoEwL4maSblbozARgbEQvy8MPA2O6E1jSOXtmOx+TD/7MrTXVdiS0f6r+CtDfaU9utLjbo8nbLzSm3AAuBmaSjmMcj4tkG634+rlz/BLBZO+JqFFtE1LbZF/I2O0XSuvWxNYh7sH0V+Bjwjzy+GW3YZk4Wve21ETEZ2Ac4WtLrqpWRjiW7fu1zr8RRcQbwQmAnYAFwcrcCkbQh8EPgAxHxZLWu29utQWxd324R8VxE7ETqpWFnYIdOx9BMfWySdgQ+QYrxVcAo4OOdjEnSm4GFEXFzu9flZLFMz3UpEhHz89+FwI9JX55Haoez+e/CLoXXLI6ub8eIeCR/sf8BfItlTSYdjU3S2qQf4+9HxI9ycU9st0ax9cp2y7E8DvwS2I3UhFO7gbi67ufjyvWbAI+1M6662PbOTXoREU8D36Xz2+w1wFskPUBqOn8D6bk/g77NnCyW6akuRSSNkLRRbRjYE7g9xzQ9TzYduKw7ETaN43Lg0Hw1yK7AE5Vml46oaxt+K2m71WKblq8I2QaYCNzYphgEfAe4MyK+Uqnq+nZrFlu3t5ukMZI2zcPrk55fcyfph/mgPFn9Nqtty4OAX+SjtUHXJLa7KolfpPMC1W3W9s8zIj4REeMjoo/0m/WLiDiEdmyzdp2dH4ov0hUM95DaST/Z5Vi2JV2BciswpxYPqX3xWuBe4OfAqA7EciGpWeIZUvvn4c3iIF39cXrehrcBU7oQ2/l53bPzl2OLyvSfzLHdDezTxrheS2pimg3ckl/79sJ26ye2rm434GXAH/L6bwc+U/ku3Eg6sf4DYN1cvl4en5vrt23jNmsW2y/yNrsd+B7Lrpjq6Pcgr3Mqy66GGvRt5u4+zMysyM1QZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYbaSlHpu3auu7AOSzuhnnqXN6sx6mZOF2cq7kHQjVNW0XG62WnGyMFt5lwL7VZ4V0AdsCfxB0rWSfq/0PJIVei+WNLX27IE8/nVJ787Dr5T0q9yB5DXt7LXXrFVOFmYrKSIWk+6C3ScXTSM9R+BvwFsjdQL5T8DJrXYDnftsOo30jIRXAmcDXxjs2M0Ganh5EjPrR60p6rL893BSVw9fzL0E/4PULfRYUpfkJS8CdgRm5vwyjNSdiVlXOVmYrZrLgFPyYzM3iIibc3PSGOCVEfFM7hF0vbr5nmX5I/tavYA5EbFbe8M2Gxg3Q5mtgkhPT/slqbmodmJ7E9IzBp6R9E/A1g1mfRCYlHty3RTYI5ffDYyRtBukZilJL2nrmzBrgY8szFbdhaTnjdSujPo+cIWk24BZwF31M0TEQ5IuIfVWej+pR1Mi4u+SDgJOlbQJ6Tv6VVLPw2Zd415nzcysyM1QZmZW5GRhZmZFThZmZlbkZGFmZkVOFmZmVuRkYWZmRU4WZmZW9P8BG0b8AAIWMAUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAElCAYAAAALP/6mAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3debgcZZn38e8vCQRkC8sBQxIIQhCDowEiwoBOBNllIg7yojgEBYMsKoqjgM4rOODgKMLgAoLsyr5I5MWRVREdlgAh7BAhmMSEBEiAAILA/f7xPF0pOt19+iTpJTm/z3X1dWqvu586VXfVU1VPKyIwMzMDGNDpAMzMrHs4KZiZWcFJwczMCk4KZmZWcFIwM7OCk4KZmRWcFKxtJP27pDM7HQeApBskHdDG9R0i6XftWl9e5+2SDmrnOpc1SQMlLZS0Uadj6S8GdToA6ztJC0u97wBeA97M/YdGxC/bH1XvIuI/mp1W0onA8Ig4qEWx7NqK5drSkXQ78POIOB8gIt4EVu9oUP2Mk8JyKCKKnUTSdOCQiLip3vSSBkXEG+2IrZtjsGXD23LF5uqjFZCkEyVdJukSSS8Bn5G0vaQ7JC2QNFvS6ZJWytMPkhSSDpU0TdJ8SaeXlre5pNskvSDpWUkXl8b9g6SbJD0vaY6krzeI4URJ5+fxm+V1fl7SX/PnK3ncx4CvAwfkqoN78vDhkq7L63pC0ueqvvOleX0LJd0vaVNJ35I0T9JfJH20NP3bqlbyd39U0kuSHpT0/hrlerakk6uG/T9JX8rd35L0ZF7GQ5L+uc722UxSVA2rjueQHM98Sb+RNCIPH5C33dy8PaZKGl3zHyEZJWlynvYaSWvn5fxW0mFVMTwsae968Ur6rKS/ADfk4TuU/qemSPpwVfyP5LL4s6RDqpb5iTzPi/l/bldJ3wO2B87M2/C00v/myDzfEEm/yNt0uqRjJam0zt9LOjXH9KSkXUvrPDjP81Iet3+Dcuu/IsKf5fgDTAc+WjXsROB1YG9S4l8V+ADwQdLV4buAx4Ej8/SDgACuBdYCRgLPV5YLXAF8Iy9rFWCHPHwt4Bngy8BgYE1g2wYxnAicn8dvltd5EakK7P3Ac8C40vznV32vPwI/yjFsDTwL/FNp+leBj+bvczHwFHBM7j8MeKK0rNuBg3L3p4AZwDaAgM2BETXKeqdc3sr96+Z1bpD79wOG5u/7aWBhadwhwO/K371q2eV4/gV4DHh3jv144A953F7AXbnsBwCjgXfW+d+4PX+v0cBqwK9K5f9p4I+labcB5gKDaiynsq3Oy9tqVWBE3l675Th2z9tj3TzP3qT/M+VyexV4Xx73j8ACYOc87wjg3dXlUPW/OTL3XwxcDayRlz8NmFAq478DnwMGAl8EZuRxawIvAKNy/1BgdKf33278dDwAf5ZyA9ZPCrf0Mt/XgCtyd2XH2640/mrga7n7YuAMYFjVMv4VuLvO8heLgdpJYbPS+B8CP6ueNvdvknf41UrDvk+qf65M/5vSuH3yQWBA7l87r2/13F8+CN8MHNFEWQ8AZgH/mPsPA25oMP2DwF65uy9J4cbKga60fV4DhgG7Ao+SEvyAXuK9HTix1P8+4G+kA/WqpAPzJnncacDpdZZT2VYblYZ9EzivarqbgQPqLOO6ShkD5wDfbxDzQVXfPUgnKisBbwCbl8YfAdxUKuNHS+PWzPOul7sX5P+LVdqxby6vH1cfrbhmlHskbZGrOuZIehH4DmlnKZtT6n6FRTf4jibtkJMlPSBpQh4+AvhzszE0Mc3TwIZ1ptsQeDYiXq6aflip/5lS96vAvIh4q9QPtW9a9vY9AMjLuox0ZQHpbLu4qS/poFxttUDSAmALFi/jZmwM/KS0nGeBt0g33m8AziQl6WcknSlpjQbLqi7fwcA6EfEqcCWpWm8gsD/pqq2R8rI2Bj5ViTHHuR15+0n6mKQ7c1XfAlIyq5RFU+Vdw/qkK4Cnq75T+X+g+n8Y0onAi6TtdgQwJ1dDbr4EMazwnBRWXNXN3/6MdOa6WUSsCfxf0hlj7wuKmB0Rh0TEUNJOdZakTUgHiU37EEMtI0rdGwF/rTPvX4H1JK1WNf2sJtbRm96+R9klwCfz99+adEWFpHeRDtSHkapQhpDO6GuV8ct5nneUhr2zKp6DI2JI6bNqRNwJEBGnRcTWwHtJVUNfbRBvdfm+RqoaBLgAOIB0wJ4fEXc3+uKRT79LMZ5XFeNqEfF9SauSEs5/kqrPhpDuQ6g0b73ybvQ/M5f0lN3GVd+pqf+BiPhNRHyUVHU0jbRPWBUnhf5jDVJ1ysuS3gMc2uyMkvaTVDkbW0Dacd8EJgEbSTpS0mBJa0rato9x/bukVSX9AzCBdCYO6ax/ZOUmYkQ8BUwGvpvXNQb4LPCLPq6vlp8DX5e0lZJRlRu71fKB80XgLOD6iHgpj1qdVC7zAEn6POlKoZY5+fMZpefwJ/L2A92ZwDfzdqrcXN03d2+bP4NIyeV10lVEPQfmq8TVgBOAy0sH99uBlYHv0ftVQrWLgH0k7ZK/wyqSPiJpQ9LVyMq5LN5UenBg59K85wCH5OkHKD1A8O487hnSvYLFRMTfScnmu5JWz4n5KzTxPyBpqKS9cyJ+nVR2jcqt33JS6D+OJh10XyKdIV3WePK3+SBwt6SXSWfGR0TEXyLiBWAX0o3RZ0g3r/+pj3HdDjxJOpP8z4i4JQ+/jHRgeV7SXXnY/wFGkQ6oVwLHRcTv+ri+xUTEJaQD42WkA/7VpHsQ9VxCuqFdPIUVEVNJN8HvAmaTbhLfWWd9AXweOI5UNbRZedqIuIJ0f+WKXNU3lXRDF2AI6aC6gHQ/aXaetp6LSAfN2aSql6Oq4riQdMXRp3dbImI6qX7+30kH/7+Q/scGRMQC0sH6GtJVyb6kewqVef+Uv//ppBOVW1l0RXMai6qlan2vw0kH9enA70lXOxc2EfJA4N9I5fAc6Wb3Ec1+3/5Eb78iNGsPSZuRngZqqgrLWkPpsd4DI2Jcp2Ox7uArBbN+KlcpHU6qCjMDnBTM+iVJe5Fu3P6FvlUl2grO1UdmZlbwlYKZmRWcFKxfkjRO0sxS/0OSxnUwJLOu4KSwHMrP6Z8j6encuNcUSXuUxo9WagRtfv7cpFKjaZK+khsEe1GpIbpT83Pv9db3Dkk/VWoM7wVJt1XFcqakZ/Lbq78uvdOAUuNls/O6HldVw2jdIiK2XBaPtzYrJ6W3lBp+q3wmlMY3XW6S9pf0WN42cyVdIGnNGtM8IullpQbqPpSHryzpSqWG4mJ5T4z53Ydbc1lMrxq3UVV5L8zf+eg8fqikSXmfKBrhK83/A6WGGF9SaqzwwDoxHJjnb7TN1lFqoPDlvB9/uuo7PJAfy30uT1fep4ZJujbvbzMlfWGJCqueTrez4U/fP6TGzY4ntQczAPgY6f2DkXn8kDxOpOezvwRMLc2/KTAkd68D3AJ8tcH6fgFcCvTk5W1TGvd14H5gA1JDdRcCV5fGbwkMzt1bkN4x2GZZlMNSluE4YGa3rr8v5UZ6xn+93L066Z2D00vjdyE1B7Fd/n8ZRm7HivQuyFHAjqRn+Md1etssZbluS2qTayIwvZdpNyG9hDky929Aehpre0qN8JWmPyFviwGkd3fmk9vBKk2zNulN9gdJTdrXW/clpBv8q+eyfwHYshTHhrl7MPBfwKTSvLeS3udYidSQ5PPAR5ZZGXZ6I7bxn2U66eWVqaS3Gc/Jhf8b0gH1JmDt0vTbAX8ivSR0f3lnIb1J+0ie70nSD9tUxo0DZpJe5Jmbd7TPtuH7TQX+pcbwQaSXdF6pM9+6+bv/tM74LUgvdK1ZZ/wZwH+V+vcCHqsz7btzeezX5Hc6H/hp3kYLSa2kvjPvEPPzzrdVafoNgatIL1M9BXypNG7VvLz5wMP5f2Fmafx0FrUKuy3wv3nbzwZ+DKxcmjaALwBP5Gl+Qn5oow/baxxNJqW+lFs+yFxIetu6MuxPpGYzept3Jn1MCrksDs9l8RLwH6STjj/l/5vLK2VHOmBel7fP/Nw9PI9bJ69/79L3mEZ6h2JJ9oeP0ntS+DZwa519ZrGkUGO6ScDRVcPOzOXxO+okBdJJ3eu8vWG/i4CTa0w7mNRcyMOlcgmgpzTNWcBFS1JONeNbVgvq9k/e6e8gJYJhpAP2vcBWpDPcW4Bv52mHkd563JN0VrBL7u/J4/fK//givcH7CrB1HjeO1JLjd0iZfM88fu06cf00H1hqfaY2+d02ILV+uUXV8AU5lreAb1WN+3TeaStNM7y/zrIPBB4ATiW9ffsApeQDjCUdrDckNat8MXBaje/4Sl7XveSWSpv4XufndW5T2kZP5ZgGklpGvTVPOwC4h9Sm08qkphKeBHbL408G/kA6+IwgncnVSwrbkE4KBpGuuB4BjipNG6QD2hBS2zvzgN3zuB0bbM8FwI6l/5PXSW+CP5XLd7UlLTcWnW0G6aRn1zx8YF7PMaSD7ExSklu1xjKWNClcS2qFdEtS20o35/Jfi5SAJ+Rp1yW9/f4OUrMrVwC/Ki1rV9IV0frA2cCVpXHHNCrXGnE1TAqkfffPlFpkLY3rNSmQTjJmV7Z7HrYtqSmWATROCltRdZJGarX416X+jfJ3e4vUOvBBefgaObb1S9OeDdzXl+3WcJsuqwV1+4e00x9Q6r8KOKPU/8XKPyjptwMuqpr/t5SaM64a9yvgy7l7HKlFzkGl8XMpNUu9jL/XSqQz/Z/VGV95QWmvOuNHkc7u6rXJf1z+JzyedLD9J9JZ+3vy+LVIVUtBSkD3kVrhrF7OQNKB61vASk1+t/OBs6u20SOl/n+oHBBIl/N/qZr/WHLzzqQEUd6BJ1InKdSI4yjgmlJ/kA/uuf9y4Jg+brd3khqzG0Cqxrit1jbsa7mRTmiOJ5+FkpJ1kA5WQ0ktlf4ROKnGvEuaFHYo9d8DfKPUfwpVJwmlcWNIDfGVh/2IdOIxi/zbDEu4X/SWFD6U/48XS7Q0lxQuAP6HRY/1D8xlvF3u/x31k8KHgDlVwz5Pblq9avg6pONRuVn723n774o8T52r8yX59LcbzdVNK1f3V5pV3pjUEma5WeAdSTsVkvZQ+sWpSrPAe/L2JpKfi7f/XGG5GeplRtIA0mXn68CRtaaJ1NT0mcCFktavMf4J4CHSWWktr5LOVE6MiNcj4vekOs3KL1r9hHSJuy4pAV1Nqu6pXs+bEXE7MJzUkmiz+rLNNqzaZseRrqIgHRyrm5GuSemX5q7TombGv0vzzYw3JSLmRMTDEfFWpMb+vk46i66erk/lFhGzSAerS/OgSpPhP4rU2u2zpLaS9uxLvL1oahvlBxZ+lm+svkhKhEOUmu6uOIvUFtP5EfHcMoyx2gTgqohY2OuUVSR9nxTjfpGP0qQTr6kRcUcTi1hIurIqW5NU/fY2EfE8KQFdW3oY5ADSicQMUvXtL0gJfZnob0mhWTNIVwrVzQKfLGkw6SrjByxqFvh6mmyGulp+cqf6iYjK56EG84lF90X+JVILkvUMIF2yD6szfhD1mzKeWmNYlLrHkHbg5yPiNdIZzLaS6v2OQKN1LY0ZwFNV22yNiKgc/GazeDPS9ZxBul8xKlIz48fR5PaV9KEG23Nh5amfGoLG+2Nfyq2YNiLmkw4Y5W0WtWZqg6NJ90c+mMu18vOdlZ/THEhKChcChyu1j0Ued1yjcu1LEEpNe3+SdLDtE0knAHuQqudeLI3amdRq7BxJc0gN7p0i6cc1FvM4MEjSqNKw95NOzmoZRKpSWxMgIp6OiI9FRE9EfJB0wnJXnXn7zEmhtl8Ae0vaTYuaBR4naTipCmUwqR75DaVHQXdttLBGIuILEbF6nc+WDWY9A3gP6cbcq+URSs0Zb5VjX5N0ZjifVDde+S3b9XP3aFI1y8111nMbqSmEY5V+L3cH4COk6jSAu0nNM6+l9JvPhwN/jYhnJa2v9Cjk6jmW3Ug/dFKsaxk+BnkX8JKkbyg1xT1Q0nslfSCPvzx/h7Xzdvxig2WtQbrfslDSFvThyiYi/tBge64eEX+A4rHDjZWMIN3zuDaP67XcyiQdIGmj3L0xcFLVtOcBX8zLXZvUgul1pfkHS1ol966c/98rB+qDVPVo51JYg3TlsEDSOqQbvWWVqsrPkX5V78LKVUREfLdRuZa+y4D8XVZKvVpF0spV69mHtD/cWh1gnndw7i2XC5KOJd2L+2iNq5iDSPvjmPyZTHpa6ZvV68hX71cD35G0Wt6nxpObL1f6/ep35+/SQ9p/78tXDUh6j6Q1lB4n/gzp+NOopdw+cVKoISJmkDbScaSD/wzS0yoDIrWf/yXSQWY+6Z9kUjvjyzv+oaR/vjmlM6YD8iRDSI+8vUC6mbYpqT79b3n8DsADSk1hX58/x5WW/1BlWfkKZDypuuEF0k2tAyPi0Tz510g3uZ8gldWepJ0O0g5+GOlMdT7p6uqoiJiU1zOCdMn8wNKWSUS8SXo0dwzpxu2zpN9JWCtPcgKpyugpUjPdjX4/4Guk7foS6fu2om2grUhP6Lyc/z5A+r+C3sut8rx95WpnNPCnvD3/SPp958+X1vUfpOT9OOnE4D5S4qh4jHSwHkZK9q+y6PcdRuRlLgunkW7QPkt66ON/KiMkbUP6saAD87b8HqkcjunjOj5Miv960tXgq6TtXTaBVBNQ64rpVVL1DqSrxfIJ13fzMqeV9rnjACJiQa4SnBMRc0hVui9Gal6+cqVTrlY9nFQWc0n76mERUblSGEYqm8q+8RaL9ilIzag/Sfrf+AJp357Xe9E0x20fWcfks5wtI+LYTsditUm6gfQQxSOdjsXaw0nBzMwKrj4yM7OCk4KZmRWcFMzMrFC3ZczlwXrrrRcjR47sdBhmZsuVe+6559mI6Kk1brlOCiNHjmTy5MmdDsPMbLkiqe4b/a4+MjOzQsuSQn6T8C5J9+eXoU7Iw8+X9JTSD8NMkTQmD5ek0yVNkzRV0tatis3MzGprZfXRa8BOEbEwN39we+mNvn+LiCurpt+D1GLnKFKLl2fkv2Zm1iYtu1KIpPK6+Er50+hNufHAhXm+O0itJw5tVXxmZra4lt5TyI15TSG173FjRNyZR52Uq4hOVWp1FFJ7H+WmjWdSo1VPSROVfn948rx5y6y5DzMzo8VJIbcFP4bUFvy2kt5LapFzC+ADLPoBib4s86yIGBsRY3t6aj5RZWZmS6gtTx9FxAJSM7W75x/6iNz2/nmkn7CD9EtL5fbuh+dhZmbWJq18+qhH0pDcvSrpd44frdwnyO21f5z0W7mQmp8+MD+FtB3wQkTMblV8Zma2uFY+fTQUuCD/SMYA4PKIuE7SLfmHIwRMIbUHDqn98z1JPy7+CvDZFsZmZmY1tCwpRMRU0g+JVA/fqc70ARzRqniqnXrj42/r/8oum7dr1WZmXctvNJuZWcFJwczMCk4KZmZWcFIwM7OCk4KZmRWcFMzMrOCkYGZmBScFMzMrOCmYmVnBScHMzApOCmZmVnBSMDOzgpOCmZkVnBTMzKzgpGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmZlZwUjAzs4KTgpmZFVqWFCStIukuSfdLekjSCXn4JpLulDRN0mWSVs7DB+f+aXn8yFbFZmZmtbXySuE1YKeIeD8wBthd0nbA94BTI2IzYD5wcJ7+YGB+Hn5qns7MzNqoZUkhkoW5d6X8CWAn4Mo8/ALg47l7fO4nj99ZkloVn5mZLa6l9xQkDZQ0BZgL3Aj8GVgQEW/kSWYCw3L3MGAGQB7/ArBujWVOlDRZ0uR58+a1Mnwzs36npUkhIt6MiDHAcGBbYItlsMyzImJsRIzt6elZ6hjNzGyRtjx9FBELgFuB7YEhkgblUcOBWbl7FjACII9fC3iuHfGZmVnSyqePeiQNyd2rArsAj5CSw755sgnAtbl7Uu4nj78lIqJV8ZmZ2eIG9T7JEhsKXCBpICn5XB4R10l6GLhU0onAfcA5efpzgIskTQOeB/ZvYWxmZlZDy5JCREwFtqox/EnS/YXq4X8DPtmqeMzMrHd+o9nMzApOCmZmVnBSMDOzgpOCmZkVnBTMzKzgpGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmZlZwUjAzs4KTgpmZFZwUzMys4KRgZmYFJwUzMys4KZiZWcFJwczMCk4KZmZWcFIwM7OCk4KZmRWcFMzMrNCypCBphKRbJT0s6SFJX87Dj5c0S9KU/NmzNM+xkqZJekzSbq2KzczMahvUwmW/ARwdEfdKWgO4R9KNedypEfGD8sSSRgP7A1sCGwI3Sdo8It5sYYxmZlbSsiuFiJgdEffm7peAR4BhDWYZD1waEa9FxFPANGDbVsVnZmaLa8s9BUkjga2AO/OgIyVNlXSupLXzsGHAjNJsM6mRRCRNlDRZ0uR58+a1MGozs/6n5UlB0urAVcBREfEicAawKTAGmA2c0pflRcRZETE2Isb29PQs83jNzPqzliYFSSuREsIvI+JqgIh4JiLejIi3gLNZVEU0CxhRmn14HmZmZm3SyqePBJwDPBIRPywNH1qabB/gwdw9Cdhf0mBJmwCjgLtaFZ+ZmS2ulU8f7QD8K/CApCl52HHApySNAQKYDhwKEBEPSboceJj05NIRfvLIzKy9WpYUIuJ2QDVGXd9gnpOAk1oVk5mZNeY3ms3MrOCkYGZmBScFMzMrOCmYmVnBScHMzApOCmZmVnBSMDOzgpOCmZkVnBTMzKzgpGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmZlZwUjAzs4KTgpmZFZwUzMys4KRgZmYFJwUzMys4KZiZWcFJwczMCk4KZmZWaFlSkDRC0q2SHpb0kKQv5+HrSLpR0hP579p5uCSdLmmapKmStm5VbGZmVlsrrxTeAI6OiNHAdsARkkYDxwA3R8Qo4ObcD7AHMCp/JgJntDA2MzOroWVJISJmR8S9ufsl4BFgGDAeuCBPdgHw8dw9HrgwkjuAIZKGtio+MzNbXFvuKUgaCWwF3AlsEBGz86g5wAa5exgwozTbzDyselkTJU2WNHnevHkti9nMrD9qKilIurmZYXXmXR24CjgqIl4sj4uIAKKZ5ZTmOSsixkbE2J6enr7MamZmvRjUaKSkVYB3AOvlG8LKo9akxll8jflXIiWEX0bE1XnwM5KGRsTsXD00Nw+fBYwozT48DzMzszbp7UrhUOAeYIv8t/K5FvhxoxklCTgHeCQiflgaNQmYkLsn5GVVhh+Yn0LaDnihVM1kZmZt0PBKISL+G/hvSV+MiB/1cdk7AP8KPCBpSh52HHAycLmkg4Gngf3yuOuBPYFpwCvAZ/u4PjMzW0oNk0JFRPxI0j8CI8vzRMSFDea5nUXVTdV2rjF9AEc0E4+ZmbVGU0lB0kXApsAU4M08OIC6ScHMzJY/TSUFYCwwOp/Nm5nZCqrZ9xQeBN7ZykDMzKzzmr1SWA94WNJdwGuVgRHxzy2JyszMOqLZpHB8K4MwM7Pu0OzTR79vdSBmZtZ5zT599BKLmqNYGVgJeDki1mxVYGZm1n7NXimsUenObyqPJzWHbWZmK5A+t5Kam7b+FbBbC+IxM7MOarb66BOl3gGk9xb+1pKIzMysY5p9+mjvUvcbwHRSFZKZma1Amr2n4MbpzMz6gWZ/ZGe4pGskzc2fqyQNb3VwZmbWXs3eaD6P9HsHG+bPr/MwMzNbgTSbFHoi4ryIeCN/zgf8W5hmZiuYZpPCc5I+I2lg/nwGeK6VgZmZWfs1mxQ+R/qFtDnAbGBf4KAWxWRmZh3S7COp3wEmRMR8AEnrAD8gJQszM1tBNHul8L5KQgCIiOeBrVoTkpmZdUqzSWGApLUrPflKodmrDDMzW040e2A/BfhfSVfk/k8CJ7UmJDMz65SmrhQi4kLgE8Az+fOJiLio0TySzs0vuj1YGna8pFmSpuTPnqVxx0qaJukxSW5sz8ysA5quAoqIh4GH+7Ds84EfAxdWDT81In5QHiBpNLA/sCXp5bibJG0eEW/2YX1mZraU+tx0drMi4jbg+SYnHw9cGhGvRcRTwDRg21bFZmZmtbUsKTRwpKSpuXqpcvN6GDCjNM3MPGwxkiZKmixp8rx581odq5lZv9LupHAGsCkwhvQS3Cl9XUBEnBURYyNibE+PW9owM1uW2poUIuKZiHgzIt4CzmZRFdEsYERp0uF5mJmZtVFbk4KkoaXefYDKk0mTgP0lDZa0CTAKuKudsZmZWQtfQJN0CTAOWE/STODbwDhJY4Ag/XrboQAR8ZCky0lPN70BHOEnj8zM2q9lSSEiPlVj8DkNpj8JvxBnZtZRnXj6yMzMupSTgpmZFZwUzMys4KRgZmYFJwUzMys4KZiZWcFJwczMCk4KZmZWcFIwM7OCk4KZmRWcFMzMrOCkYGZmBScFMzMrOCmYmVnBScHMzApOCmZmVnBSMDOzgpOCmZkVnBTMzKzgpGBmZgUnBTMzK7QsKUg6V9JcSQ+Whq0j6UZJT+S/a+fhknS6pGmSpkraulVxmZlZfa28Ujgf2L1q2DHAzRExCrg59wPsAYzKn4nAGS2My8zM6mhZUoiI24DnqwaPBy7I3RcAHy8NvzCSO4Ahkoa2KjYzM6ut3fcUNoiI2bl7DrBB7h4GzChNNzMPW4ykiZImS5o8b9681kVqZtYPdexGc0QEEEsw31kRMTYixvb09LQgMjOz/qvdSeGZSrVQ/js3D58FjChNNzwPMzOzNmp3UpgETMjdE4BrS8MPzE8hbQe8UKpmMjOzNhnUqgVLugQYB6wnaSbwbeBk4HJJBwNPA/vlya8H9gSmAa8An21VXGZmVl/LkkJEfKrOqJ1rTBvAEa2KxczMmuM3ms3MrOCkYGZmBScFMzMrOCmYmVmhZTeaVzSn3vj42/q/ssvmHYrEzKx1fKVgZmYFXylkvhIwM/OVgpmZlfhKoY7qKwczs/7AVwpmZlZwUjAzs4KTgpmZFZwUzMys4KRgZmYFJwUzMyv4kdQl5JfdzGxF5CsFMzMrOCmYmVnBScHMzApOCmZmVvCN5hYp34j2TWgzW150JClImg68BLwJvBERYyWtA1wGjASmA/tFxPxOxGdm1l91svroI6KcaN0AAAkESURBVBExJiLG5v5jgJsjYhRwc+43M7M26qZ7CuOBC3L3BcDHOxiLmVm/1KmkEMANku6RNDEP2yAiZufuOcAGnQnNzKz/6tSN5h0jYpak9YEbJT1aHhkRISlqzZiTyESAjTbaqPWRmpn1Ix25UoiIWfnvXOAaYFvgGUlDAfLfuXXmPSsixkbE2J6ennaFbGbWL7Q9KUhaTdIalW5gV+BBYBIwIU82Abi23bGZmfV3nag+2gC4RlJl/RdHxP9Iuhu4XNLBwNPAfh2IzcysX2t7UoiIJ4H31xj+HLBzu+NZHrmFVjNrlW56JNXMzDrMzVwsI9Vn72ZmyyNfKZiZWcFJwczMCq4+6oC+VjX5RrKZtYuvFMzMrOCkYGZmBScFMzMr+J6CvY1fjDPr33ylYGZmBV8ptEGrX2zry+9B+0rAzBpxUlgO9CWp+KBvZkvDSaGfc/McZlbmpGBt05ermG6+4unm2MyWlm80m5lZwVcK1lC33MT22blZezgprOBaec+gv9yP6C/f0wycFMyWOV/V2PLMScE6xgdPs+7jpGBdo53VNI3ulbi6qHftTOhLuy6ffPSNk4KZLbXeEqkPxH3XqWTmpGBN6+Yz6N52oG6JvZNn2L1ZXg7cvnJora5LCpJ2B/4bGAj8PCJO7nBItoSWlyef2tk21dLO201JpD/q68nH8phwuiopSBoI/ATYBZgJ3C1pUkQ83NnIbHnTqfsT1n7L04G4L+/9dIoiotMxFCRtDxwfEbvl/mMBIuI/a00/duzYmDx58hKtyzuyWffoZHVfed3L03FhaZKKpHsiYmytcV11pQAMA2aU+mcCHyxPIGkiMDH3LpT02BKuaz3g2SWct9W6NbZujQsc25Lomri+uvigtsVWY92NrChltnG9Ed2WFHoVEWcBZy3tciRNrpcpO61bY+vWuMCxLYlujQu6N7ZujQuWXWzd1iDeLGBEqX94HmZmZm3QbUnhbmCUpE0krQzsD0zqcExmZv1GV1UfRcQbko4Efkt6JPXciHioRatb6iqoFurW2Lo1LnBsS6Jb44Luja1b44JlFFtXPX1kZmad1W3VR2Zm1kFOCmZmVuiXSUHS7pIekzRN0jEdjmW6pAckTZE0OQ9bR9KNkp7If9duUyznSpor6cHSsJqxKDk9l+FUSVt3ILbjJc3KZTdF0p6lccfm2B6TtFsL4xoh6VZJD0t6SNKX8/COlluDuLqhzFaRdJek+3NsJ+Thm0i6M8dwWX7YBEmDc/+0PH5km+M6X9JTpTIbk4e3dR/I6xwo6T5J1+X+ZV9mEdGvPqQb2H8G3gWsDNwPjO5gPNOB9aqG/RdwTO4+Bvhem2L5MLA18GBvsQB7Ar8BBGwH3NmB2I4HvlZj2tF5uw4GNsnbe2CL4hoKbJ271wAez+vvaLk1iKsbykzA6rl7JeDOXBaXA/vn4WcCh+Xuw4Ezc/f+wGVtjut8YN8a07d1H8jr/CpwMXBd7l/mZdYfrxS2BaZFxJMR8TpwKTC+wzFVGw9ckLsvAD7ejpVGxG3A803GMh64MJI7gCGShrY5tnrGA5dGxGsR8RQwjbTdWxHX7Ii4N3e/BDxCejO/o+XWIK562llmERELc+9K+RPATsCVeXh1mVXK8kpgZ0lqY1z1tHUfkDQc2Av4ee4XLSiz/pgUajWl0WhnabUAbpB0j1ITHgAbRMTs3D0H2KAzoTWMpVvK8ch86X5uqZqtI7HlS/StSGeYXVNuVXFBF5RZrgaZAswFbiRdmSyIiDdqrL+ILY9/AVi3HXFFRKXMTspldqqkwdVx1Yi5FU4Dvg68lfvXpQVl1h+TQrfZMSK2BvYAjpD04fLISNd/XfHccDfFkp0BbAqMAWYDp3QqEEmrA1cBR0XEi+VxnSy3GnF1RZlFxJsRMYbUasG2wBadiKNadVyS3gscS4rvA8A6wDfaHZekjwFzI+KeVq+rPyaFrmpKIyJm5b9zgWtIO8gzlcvQ/Hdup+JrEEvHyzEinsk78VvA2Syq7mhrbJJWIh14fxkRV+fBHS+3WnF1S5lVRMQC4FZge1L1S+WF2vL6i9jy+LWA59oU1+65Ki4i4jXgPDpTZjsA/yxpOqnKeyfS784s8zLrj0mha5rSkLSapDUq3cCuwIM5ngl5sgnAtZ2IL6sXyyTgwPwExnbAC6Xqkraoqr/dh1R2ldj2z09gbAKMAu5qUQwCzgEeiYgflkZ1tNzqxdUlZdYjaUjuXpX0+ymPkA7C++bJqsusUpb7Arfkq692xPVoKbmLVGdfLrO27AMRcWxEDI+IkaRj1i0RcQCtKLNW3SXv5g/pqYHHSfWY3+xgHO8iPfFxP/BQJRZS3d/NwBPATcA6bYrnElKVwt9J9ZMH14uF9MTFT3IZPgCM7UBsF+V1T807wdDS9N/MsT0G7NHCuHYkVQ1NBabkz56dLrcGcXVDmb0PuC/H8CDwf0v7w12km9xXAIPz8FVy/7Q8/l1tjuuWXGYPAr9g0RNKbd0HSnGOY9HTR8u8zNzMhZmZFfpj9ZGZmdXhpGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmDSi1NLpb1bCjJJ3RYJ6F9caZdTsnBbPGLiG9LFS2fx5utsJxUjBr7Epgr1I79SOBDYH7JN0s6V6l38NYrKVdSeMq7d7n/h9LOih3byPp97khxN+2snVNs75wUjBrICKeJ70RukcetD+pDftXgX0iNWb4EeCUZpsmzm0S/YjURv82wLnAScs6drMlMaj3Scz6vUoV0rX578GkJg6+m1u1fYvUVPEGpGaye/Nu4L3AjTmPDCQ14WHWcU4KZr27Fjg1/9ziOyLinlwN1ANsExF/z61XrlI13xu8/Wq8Ml7AQxGxfWvDNus7Vx+Z9SLSr3HdSqrmqdxgXovUvv3fJX0E2LjGrE8Do3PLo0OAnfPwx4AeSdtDqk6StGVLv4RZk3ylYNacS0i/d1F5EumXwK8lPQBMBh6tniEiZki6nNS65lOkFjiJiNcl7QucLmkt0n54GqmlXLOOciupZmZWcPWRmZkVnBTMzKzgpGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmZlb4/3bql05VkQwdAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Without constraints, flux balance analysis found that 7.966 grams of biomass could be made.\n",
      "\n",
      "\n",
      "Given the bounds from transcriptmics and a minimum absolute bound of 25, flux balance analysis found that 5.086 grams of biomass could be made.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>transcipt_level</th>\n",
       "      <th>upper_bound</th>\n",
       "      <th>lower_bound</th>\n",
       "      <th>unbound_flux</th>\n",
       "      <th>EFlux_flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>ATPS4rpp</td>\n",
       "      <td>126.50221</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>521.244896</td>\n",
       "      <td>442.967452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1575</th>\n",
       "      <td>O2t</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>397.269146</td>\n",
       "      <td>506.738766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229</th>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>299.675525</td>\n",
       "      <td>408.274981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153</th>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>205.468755</td>\n",
       "      <td>239.651922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898</th>\n",
       "      <td>FUM</td>\n",
       "      <td>577.03996</td>\n",
       "      <td>5770.3996</td>\n",
       "      <td>-5770.3996</td>\n",
       "      <td>203.462234</td>\n",
       "      <td>171.669954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1381</th>\n",
       "      <td>MDH</td>\n",
       "      <td>177.40199</td>\n",
       "      <td>1774.0199</td>\n",
       "      <td>-1774.0199</td>\n",
       "      <td>203.462234</td>\n",
       "      <td>180.542888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>CBFC2pp</td>\n",
       "      <td>4.57131</td>\n",
       "      <td>45.7131</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>196.554686</td>\n",
       "      <td>45.713100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1501</th>\n",
       "      <td>NDH1_4pp</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>196.554686</td>\n",
       "      <td>45.713100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>CYO1bpp_syn</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>196.554686</td>\n",
       "      <td>45.713100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>SUCDi</td>\n",
       "      <td>14.15348</td>\n",
       "      <td>141.5348</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>195.418357</td>\n",
       "      <td>141.534800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>CYTBDpp</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>25.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>195.418357</td>\n",
       "      <td>25.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>872</th>\n",
       "      <td>FMNRx</td>\n",
       "      <td>34.10900</td>\n",
       "      <td>341.0900</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>129.713929</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>867</th>\n",
       "      <td>FLDO</td>\n",
       "      <td>109.71144</td>\n",
       "      <td>1097.1144</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>129.713929</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>CS</td>\n",
       "      <td>726.39871</td>\n",
       "      <td>7263.9871</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>103.551824</td>\n",
       "      <td>71.727148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>ICDHx</td>\n",
       "      <td>0.52746</td>\n",
       "      <td>25.0000</td>\n",
       "      <td>-25.0000</td>\n",
       "      <td>103.551824</td>\n",
       "      <td>-25.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>ACONT</td>\n",
       "      <td>661.78501</td>\n",
       "      <td>6617.8501</td>\n",
       "      <td>-6617.8501</td>\n",
       "      <td>103.551824</td>\n",
       "      <td>71.727148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2381</th>\n",
       "      <td>phenol_transport</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2382</th>\n",
       "      <td>PHEMOX</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>CATDOX</td>\n",
       "      <td>17124.03884</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>3OXCOAT</td>\n",
       "      <td>388.20970</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>3OADPCOAT</td>\n",
       "      <td>3.95594</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466</th>\n",
       "      <td>MUCLI</td>\n",
       "      <td>1310.03191</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626</th>\n",
       "      <td>OXOAEL</td>\n",
       "      <td>154.45753</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1464</th>\n",
       "      <td>MUCCYCI</td>\n",
       "      <td>2662.52427</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>886</th>\n",
       "      <td>FRDO</td>\n",
       "      <td>456.86726</td>\n",
       "      <td>4568.6726</td>\n",
       "      <td>-4568.6726</td>\n",
       "      <td>95.400739</td>\n",
       "      <td>17.393206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1601</th>\n",
       "      <td>OOR3r</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>95.400739</td>\n",
       "      <td>17.393206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>83.280866</td>\n",
       "      <td>53.165915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1033</th>\n",
       "      <td>GLYBt2pp</td>\n",
       "      <td>61.88850</td>\n",
       "      <td>618.8850</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>74.849305</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1035</th>\n",
       "      <td>GLYBtex</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>74.849305</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155</th>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>-inf</td>\n",
       "      <td>67.732307</td>\n",
       "      <td>43.239825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1292</th>\n",
       "      <td>LEUTA</td>\n",
       "      <td>271.02593</td>\n",
       "      <td>2710.2593</td>\n",
       "      <td>-2710.2593</td>\n",
       "      <td>55.993660</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1776</th>\n",
       "      <td>PPCK</td>\n",
       "      <td>2223.94225</td>\n",
       "      <td>22239.4225</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>43.312735</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1694</th>\n",
       "      <td>PGM</td>\n",
       "      <td>915.14107</td>\n",
       "      <td>9151.4107</td>\n",
       "      <td>-9151.4107</td>\n",
       "      <td>37.342190</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1577</th>\n",
       "      <td>OAADC</td>\n",
       "      <td>11.86782</td>\n",
       "      <td>118.6782</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>33.938780</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1768</th>\n",
       "      <td>PPA</td>\n",
       "      <td>310.40952</td>\n",
       "      <td>3104.0952</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>31.159198</td>\n",
       "      <td>19.891812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1690</th>\n",
       "      <td>PGK</td>\n",
       "      <td>210.01652</td>\n",
       "      <td>2100.1652</td>\n",
       "      <td>-2100.1652</td>\n",
       "      <td>25.889676</td>\n",
       "      <td>16.527419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>ADK1</td>\n",
       "      <td>72.70141</td>\n",
       "      <td>727.0141</td>\n",
       "      <td>-727.0141</td>\n",
       "      <td>23.227850</td>\n",
       "      <td>109.045566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>GLNS</td>\n",
       "      <td>4.39549</td>\n",
       "      <td>43.9549</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>16.115349</td>\n",
       "      <td>10.286824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1826</th>\n",
       "      <td>PSP_L</td>\n",
       "      <td>417.83530</td>\n",
       "      <td>4178.3530</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>11.452514</td>\n",
       "      <td>74.011489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1825</th>\n",
       "      <td>PSERT</td>\n",
       "      <td>31.12007</td>\n",
       "      <td>311.2007</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>11.452514</td>\n",
       "      <td>74.011489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1688</th>\n",
       "      <td>PGCD</td>\n",
       "      <td>135.20528</td>\n",
       "      <td>1352.0528</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>11.452514</td>\n",
       "      <td>74.011489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>PIt2r</td>\n",
       "      <td>30.32888</td>\n",
       "      <td>303.2888</td>\n",
       "      <td>-303.2888</td>\n",
       "      <td>10.859181</td>\n",
       "      <td>6.932424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1659</th>\n",
       "      <td>PDH</td>\n",
       "      <td>356.82592</td>\n",
       "      <td>3568.2592</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>9.211256</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>966</th>\n",
       "      <td>GHMT2r</td>\n",
       "      <td>52.21843</td>\n",
       "      <td>522.1843</td>\n",
       "      <td>-522.1843</td>\n",
       "      <td>8.629515</td>\n",
       "      <td>72.209306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>FBP</td>\n",
       "      <td>632.86269</td>\n",
       "      <td>6328.6269</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>8.509997</td>\n",
       "      <td>5.432355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>ASPK</td>\n",
       "      <td>207.46714</td>\n",
       "      <td>2074.6714</td>\n",
       "      <td>-2074.6714</td>\n",
       "      <td>8.288363</td>\n",
       "      <td>5.291232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>Growth</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7.966152</td>\n",
       "      <td>5.085535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1454</th>\n",
       "      <td>MTHFC</td>\n",
       "      <td>62.24014</td>\n",
       "      <td>622.4014</td>\n",
       "      <td>-622.4014</td>\n",
       "      <td>7.658417</td>\n",
       "      <td>138.289650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1455</th>\n",
       "      <td>MTHFD</td>\n",
       "      <td>61.53686</td>\n",
       "      <td>615.3686</td>\n",
       "      <td>-615.3686</td>\n",
       "      <td>7.658417</td>\n",
       "      <td>138.289650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1816</th>\n",
       "      <td>PRPPS</td>\n",
       "      <td>89.49218</td>\n",
       "      <td>894.9218</td>\n",
       "      <td>-894.9218</td>\n",
       "      <td>7.211148</td>\n",
       "      <td>4.603546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           reaction_id  transcipt_level  upper_bound  lower_bound  \\\n",
       "390           ATPS4rpp        126.50221          inf         -inf   \n",
       "1575               O2t              inf          inf         -inf   \n",
       "2229          EX_co2_e              inf          inf       0.0000   \n",
       "2153          EX_h2o_e              inf          inf         -inf   \n",
       "898                FUM        577.03996    5770.3996   -5770.3996   \n",
       "1381               MDH        177.40199    1774.0199   -1774.0199   \n",
       "449            CBFC2pp          4.57131      45.7131       0.0000   \n",
       "1501          NDH1_4pp              inf          inf       0.0000   \n",
       "541        CYO1bpp_syn              inf          inf       0.0000   \n",
       "1985             SUCDi         14.15348     141.5348       0.0000   \n",
       "552            CYTBDpp          0.00000      25.0000       0.0000   \n",
       "872              FMNRx         34.10900     341.0900       0.0000   \n",
       "867               FLDO        109.71144    1097.1144       0.0000   \n",
       "521                 CS        726.39871    7263.9871       0.0000   \n",
       "1197             ICDHx          0.52746      25.0000     -25.0000   \n",
       "189              ACONT        661.78501    6617.8501   -6617.8501   \n",
       "2381  phenol_transport              inf          inf         -inf   \n",
       "2382            PHEMOX              inf          inf         -inf   \n",
       "448             CATDOX      17124.03884          inf         -inf   \n",
       "96             3OXCOAT        388.20970          inf         -inf   \n",
       "72           3OADPCOAT          3.95594          inf         -inf   \n",
       "1466             MUCLI       1310.03191          inf         -inf   \n",
       "1626            OXOAEL        154.45753          inf         -inf   \n",
       "1464           MUCCYCI       2662.52427          inf         -inf   \n",
       "886               FRDO        456.86726    4568.6726   -4568.6726   \n",
       "1601             OOR3r              inf          inf       0.0000   \n",
       "1513              NH4t              inf          inf         -inf   \n",
       "1033          GLYBt2pp         61.88850     618.8850       0.0000   \n",
       "1035           GLYBtex              inf          inf         -inf   \n",
       "2155            EX_h_e              inf          inf         -inf   \n",
       "1292             LEUTA        271.02593    2710.2593   -2710.2593   \n",
       "1776              PPCK       2223.94225   22239.4225       0.0000   \n",
       "1694               PGM        915.14107    9151.4107   -9151.4107   \n",
       "1577             OAADC         11.86782     118.6782       0.0000   \n",
       "1768               PPA        310.40952    3104.0952       0.0000   \n",
       "1690               PGK        210.01652    2100.1652   -2100.1652   \n",
       "217               ADK1         72.70141     727.0141    -727.0141   \n",
       "996               GLNS          4.39549      43.9549       0.0000   \n",
       "1826             PSP_L        417.83530    4178.3530       0.0000   \n",
       "1825             PSERT         31.12007     311.2007       0.0000   \n",
       "1688              PGCD        135.20528    1352.0528       0.0000   \n",
       "1739             PIt2r         30.32888     303.2888    -303.2888   \n",
       "1659               PDH        356.82592    3568.2592       0.0000   \n",
       "966             GHMT2r         52.21843     522.1843    -522.1843   \n",
       "823                FBP        632.86269    6328.6269       0.0000   \n",
       "377               ASPK        207.46714    2074.6714   -2074.6714   \n",
       "2378            Growth              inf          inf       0.0000   \n",
       "1454             MTHFC         62.24014     622.4014    -622.4014   \n",
       "1455             MTHFD         61.53686     615.3686    -615.3686   \n",
       "1816             PRPPS         89.49218     894.9218    -894.9218   \n",
       "\n",
       "      unbound_flux  EFlux_flux  \n",
       "390     521.244896  442.967452  \n",
       "1575    397.269146  506.738766  \n",
       "2229    299.675525  408.274981  \n",
       "2153    205.468755  239.651922  \n",
       "898     203.462234  171.669954  \n",
       "1381    203.462234  180.542888  \n",
       "449     196.554686   45.713100  \n",
       "1501    196.554686   45.713100  \n",
       "541     196.554686   45.713100  \n",
       "1985    195.418357  141.534800  \n",
       "552     195.418357   25.000000  \n",
       "872     129.713929    0.000000  \n",
       "867     129.713929    0.000000  \n",
       "521     103.551824   71.727148  \n",
       "1197    103.551824  -25.000000  \n",
       "189     103.551824   71.727148  \n",
       "2381    100.000000  100.000000  \n",
       "2382    100.000000  100.000000  \n",
       "448     100.000000  100.000000  \n",
       "96      100.000000  100.000000  \n",
       "72      100.000000  100.000000  \n",
       "1466    100.000000  100.000000  \n",
       "1626    100.000000  100.000000  \n",
       "1464    100.000000  100.000000  \n",
       "886      95.400739   17.393206  \n",
       "1601     95.400739   17.393206  \n",
       "1513     83.280866   53.165915  \n",
       "1033     74.849305    0.000000  \n",
       "1035     74.849305    0.000000  \n",
       "2155     67.732307   43.239825  \n",
       "1292     55.993660    0.000000  \n",
       "1776     43.312735    0.000000  \n",
       "1694     37.342190    0.000000  \n",
       "1577     33.938780    0.000000  \n",
       "1768     31.159198   19.891812  \n",
       "1690     25.889676   16.527419  \n",
       "217      23.227850  109.045566  \n",
       "996      16.115349   10.286824  \n",
       "1826     11.452514   74.011489  \n",
       "1825     11.452514   74.011489  \n",
       "1688     11.452514   74.011489  \n",
       "1739     10.859181    6.932424  \n",
       "1659      9.211256    0.000000  \n",
       "966       8.629515   72.209306  \n",
       "823       8.509997    5.432355  \n",
       "377       8.288363    5.291232  \n",
       "2378      7.966152    5.085535  \n",
       "1454      7.658417  138.289650  \n",
       "1455      7.658417  138.289650  \n",
       "1816      7.211148    4.603546  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sol_WTPR1 = EFlux2test(\n",
    "    model, \n",
    "    trans_df, \n",
    "    'WT-P-R1', \n",
    "    min_bound = 25,\n",
    "    transcript_scaling_factor = 10,\n",
    "    infinite_bound_reactions=['phenol_transport', 'PHEMOX', 'CATDOX', 'MUCCYCI', 'MUCLI', 'OXOAEL', '3OADPCOAT', '3OXCOAT', 'ATPS4rpp'],\n",
    "    zero_bound_reactions=['CAT23DOX', 'HMSH', 'HMSD', '4OT']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "bfd4264b-f30c-4d0e-9df9-d4aa8dee805f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>3OXCOAT</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>3-oxoadipyl-CoA thiolase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f40e810c750</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>coa_c + oxadpcoa_c <=> accoa_c + succoa_c</p>\n",
       "                    <p style='text-align:right'>Coenzyme A + 3-Oxoadipyl-CoA <=> Acetyl-CoA + Succinyl-CoA</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>WP_005250766_1 or WP_005251119_1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-inf</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>inf</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction 3OXCOAT at 0x7f40e810c750>"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('3OXCOAT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "40957e8b-46b6-4852-a213-613253e70bfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>CATDOX</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>CATDOX</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f40dfd072d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>catechol_c + o2_c --> ccmuac_c + 2.0 h_c</p>\n",
       "                    <p style='text-align:right'>Catechol + O2 --> Cis,cis-Muconate + 2.0 H+</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>WP_005251124_1 or WP_005251389_1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction CATDOX at 0x7f40dfd072d0>"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('CATDOX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "d92daefa-abff-413f-bc7a-81e3aca7ba09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BZDIOLDH: bz12diol_c + nad_c --> catechol_c + co2_c + nadh_c\n",
      "PHEMOX: h_c + nadh_c + o2_c + phenol_c --> catechol_c + h2o_c + nad_c\n",
      "CATDOX: catechol_c + o2_c --> ccmuac_c + 2.0 h_c\n",
      "CAT23DOX: catechol_c + o2_c --> 2hmcnsad_c\n"
     ]
    }
   ],
   "source": [
    "for rxn in model.metabolites.get_by_id('catechol_c').reactions:\n",
    "    print(rxn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "2e92001f-2eaa-4dca-8fb4-ca7a698b6c4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT23DOX: catechol_c + o2_c --> 2hmcnsad_c\n",
      "HMSH: 2hmcnsad_c + h2o_c --> for_c + 2.0 h_c + op4en_c\n",
      "HMSD: 2hmcnsad_c + h2o_c + nad_c --> 2hmc_c + 3.0 h_c + nadh_c\n"
     ]
    }
   ],
   "source": [
    "for rxn in model.metabolites.get_by_id('2hmcnsad_c').reactions:\n",
    "    print(rxn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "9ebd8cc0-400b-405d-8ded-2594515f5a17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4OT: 2hmc_c <=> oxalc_c\n",
      "HMSD: 2hmcnsad_c + h2o_c + nad_c --> 2hmc_c + 3.0 h_c + nadh_c\n"
     ]
    }
   ],
   "source": [
    "for rxn in model.metabolites.get_by_id('2hmc_c').reactions:\n",
    "    print(rxn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1df050d-7679-4c9a-8e9c-cd391b80ff2f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "20978510-4343-48aa-bd76-cae4e38a8f01",
   "metadata": {},
   "source": [
    "Unused code used for reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "5c70e3fb-7997-4935-b6af-7a18e443a7fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# boundry reactions and phenol uptake reactions should not have transcriptomic enforced bounds\n",
    "#         infinite_bound_reactions = ['phenol_transport', 'PHEMOX', 'CATDOX', 'MUCCYCI', 'MUCLI', 'OXOAEL', '3OADPCOAT', '3OXCOAT']\n",
    "#         infinite_bound_reactions = [rxn.id for rxn in model.boundary]\n",
    "#         infinite_bound_reactions.extend(['phenol_transport', 'PHEMOX', 'CATDOX', 'MUCCYCI', 'MUCLI', 'OXOAEL', '3OADPCOAT', '3OXCOAT'])\n",
    "        \n",
    "        # meta pathway reactions should have upper and lower bounds of zero\n",
    "#         zero_bound_reactions = [rxn.id for rxn in model.boundary]\n",
    "#         zero_bound_reactions.extend(['CAT23DOX', 'HMSH', 'HMSD', '4OT'])\n",
    "#         zero_bound_reactions = ['CAT23DOX', 'HMSH', 'HMSD', '4OT']\n",
    "\n",
    "# def rxn_to_GPR(model, rxn):\n",
    "#     parallel_gene_sets = set()\n",
    "#     if rxn.gene_reaction_rule:\n",
    "#         for parallel_gene_string in rxn.gene_reaction_rule.split(' or '):\n",
    "#             parallel_gene_string = parallel_gene_string.strip('() ')\n",
    "#             parallel_gene_list = []\n",
    "            \n",
    "#             for gene_name in parallel_gene_string.split(' and '):\n",
    "#                 gene_name = gene_name.strip('() ')\n",
    "#                 parallel_gene_list.append(gene_name)\n",
    "            \n",
    "#             parallel_gene_sets.add(frozenset(parallel_gene_list))\n",
    "            \n",
    "#     return parallel_gene_sets\n",
    "            \n",
    "# Orginal code\n",
    "#     if r.gene_reaction_rule:\n",
    "#         for x in [x.strip('() ') for x in r.gene_reaction_rule.split(' or ')]:\n",
    "#             temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "#         return temp\n",
    "#     else:\n",
    "#         return {}\n",
    "    \n",
    "# Christina's version\n",
    "#     if r.gene_reaction_rule:\n",
    "#         t2 = 0\n",
    "#         #If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "#         #If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "#         #If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "#         for p in gpr_dict[r.id]:\n",
    "#             xg = []\n",
    "#             for g in p:\n",
    "#                 if g in transcriptomics_dict.keys():\n",
    "#                     print(transcriptomics_dict.get(g))\n",
    "#                     xg.append(transcriptomics_dict.get(g))\n",
    "#                 else:\n",
    "#                     xg.append(np.Inf)\n",
    "#                 t1 = np.min(xg)\n",
    "#             t2 = t2 + t1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "61a35c08-977d-4a61-acb5-4868b5e22b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# _ = [print(rxn, rxn.upper_bound) for rxn in model.reactions if rxn.upper_bound != 1000] \n",
    "# print()\n",
    "# _ = [print(rxn, rxn.lower_bound) for rxn in model.reactions if ( rxn.lower_bound != -1000 and rxn.lower_bound != 0)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e3a84421-52c3-4620-95c9-dea6ed91b142",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'fluxes'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-44-2a442123d3bb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0meflux2sol_WTPR1_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msol_WTPR1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfluxes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'fluxes'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0meflux2sol_WTPR1_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'fluxes'"
     ]
    }
   ],
   "source": [
    "eflux2sol_WTPR1_df = pd.DataFrame(sol_WTPR1.fluxes, columns=['fluxes'])\n",
    "eflux2sol_WTPR1_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b93e9bf6-7128-4a3d-bb1e-a45fd3ce49e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6d92333-13e8-43a4-bf5c-4fb31fd006c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "99fefd29-1e17-4bd1-92f3-b06e4d0b1d1c",
   "metadata": {},
   "source": [
    "##### Just linear scale up to solution to see how it compares:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a49529-2838-471e-9f6c-206ddaa49cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "testscaleeflux2 = eflux2sol_WTPR1_df*(100/1.76124)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "999610be-d11c-4525-9c73-c93059ae8622",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(testscaleeflux2.index,testscaleeflux2.fluxes)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c23b85-c646-4b05-8297-7b085fac11d9",
   "metadata": {},
   "source": [
    "Old efflux code saved as reference material"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "171abbc4-d08e-4695-81db-8f43c51d9562",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def EFlux2test(model, edd_df, line_name, min_bound = 10):\n",
    "    # covert transcriptomic df to dictionary of genes and transcript values\n",
    "#     transcriptomics_dict = {row['Measurement Type']:row['Value'] for _, row in trans_df.iterrows()}\n",
    "    \n",
    "#     # plot histogram of transcriptomic measurements\n",
    "#     plot_histogram(transcriptomics_dict.values(), 'Transcriptomic values')\n",
    "    \n",
    "#     # convert model into dictionary of genes and their associated GPRs\n",
    "#     gpr_dict = model_to_GPR_dictionary(model)\n",
    "    \n",
    "#     with model:        \n",
    "#         # Ensure that medium is correct, and the carbon source is the only limiting substrate\n",
    "#         medium = {key:np.Inf for (key,value) in model.medium.items()}\n",
    "#         medium[\"EX_glc__D_e\"] = 0.0\n",
    "#         medium[\"EX_phenol_e\"] = 1.76124#100.0\n",
    "#         model.medium = medium\n",
    "#         print(model.medium) # print the media to prevent inivisible errors\n",
    "        \n",
    "#         # calculate maximum biomass yield without constraints\n",
    "#         model.tolerance = 1e-9\n",
    "#         unbounded_sol = model.optimize()\n",
    "#         print(f'\\nWithout constraints, flux balance analysis found that {unbounded_sol.objective_value:.3f} grams of biomass could be made.\\n')\n",
    "\n",
    "#         # calculate maximum biomass yield without constraints\n",
    "#         bounds = []\n",
    "#         bounded_model = model.copy()\n",
    "#         with bounded_model:\n",
    "#             for rxn_id in list(gpr_dict.keys()):\n",
    "#                 #exclude phenol related rxns from transcript bounds for now:\n",
    "# #                 if bounded_model.reactions.get_by_id(rxn_id).upper_bound != 1000:\n",
    "# #                     display('Phenol related rxns excluded from transcript bounds',rxn_id)\n",
    "# #                     continue\n",
    "#                 bound = 0\n",
    "#                 for genes_in_series in gpr_dict[rxn_id]:\n",
    "#                     bound += min([transcriptomics_dict[gene] if gene in transcriptomics_dict.keys() else 0 for gene in genes_in_series ])\n",
    "\n",
    "#                 bound = max(bound, min_bound) # min_bound is a parameter of this function\n",
    "#                 bounds.append(bound)\n",
    "\n",
    "#                 if bounded_model.reactions.get_by_id(rxn_id).upper_bound > 0:\n",
    "#                     bounded_model.reactions.get_by_id(rxn_id).upper_bound = bound\n",
    "\n",
    "#                 elif bounded_model.reactions.get_by_id(rxn_id).lower_bound < 0:\n",
    "#                     bounded_model.reactions.get_by_id(rxn_id).lower_bound = -1*bound\n",
    "                    \n",
    "#         #Alternative to external function:\n",
    "#         # I think this is more readable than the nested loops. \n",
    "#         #         xg = []\n",
    "#         #         t2 = 0\n",
    "#         #         for r in model.reactions:\n",
    "#         #             if r.gene_reaction_rule:\n",
    "#         #                 t2 = 0\n",
    "#         #                 #If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "#         #                 #If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "#         #                 #If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "#         #                 for p in gpr_dict[r.id]:\n",
    "#         #                     xg = []\n",
    "#         #                     for g in p:\n",
    "#         #                         if g in transcriptomics_dict.keys():\n",
    "#         #                             print(transcriptomics_dict.get(g))\n",
    "#         #                             xg.append(transcriptomics_dict.get(g))\n",
    "#         #                         else:\n",
    "#         #                             xg.append(np.Inf)\n",
    "#         #                         t1 = np.min(xg)\n",
    "#         #                     t2 = t2 + t1\n",
    "\n",
    "#             # plot histogram of absolute value of bounds inforced\n",
    "#             plot_histogram(bounds, 'Absolute value of bounds')\n",
    "\n",
    "#             bounded_sol = bounded_model.optimize()\n",
    "#             print(f'\\nGiven the bounds from transcriptmics and a minimum absolute bound of {min_bound}, flux balance analysis found that {bounded_sol.objective_value:.3f} grams of biomass could be made.\\n')\n",
    "\n",
    "#         # Constrain the biomass to the optimal value\n",
    "#             print('FBA status', bounded_sol.status)\n",
    "#             print('FBA solution', bounded_sol.objective_value)\n",
    "#             display(bounded_model.objective)\n",
    "#             # Constrain the biomass to the optimal value\n",
    "#             for rxn_id in bounded_model.reactions:\n",
    "#                 if rxn_id.objective_coefficient:\n",
    "#                     rxn_id.lower_bound = bounded_sol.objective_value\n",
    "#             # minimize the sum of squared flux values\n",
    "#             bounded_model.objective = bounded_model.problem.Objective(add([rxn_id.flux_expression**2 for rxn_id in bounded_model.reactions]), direction='min')\n",
    "#             display(bounded_model.problem.Objective)\n",
    "#             eflux2_sol = bounded_model.optimize()\n",
    "#             print('EFlux2 status', eflux2_sol.status)\n",
    "#             print('EFlux2 solution', eflux2_sol.objective_value)\n",
    "#     return eflux2_sol"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4f678b3-8c1c-4444-9e30-97c812e5fe7c",
   "metadata": {},
   "source": [
    "#### **Plot solutions: Comparison of SPOT and EFLUX2 solution**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4f7d0ac-a223-4e43-a666-f75507694c95",
   "metadata": {},
   "source": [
    "##### **Load 13cdata**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bbd32f4-f440-47d7-bf5f-f669e93204ba",
   "metadata": {},
   "source": [
    "####  **Get 13C MFA measured fluxes for phenol**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae1c076f-74cd-44e5-82a9-adeb0f9c0726",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenol_fluxes = pd.read_csv('../13C_flux_data/13C_phenol_flux_data.csv')\n",
    "phenol_fluxes.dropna(inplace=True)\n",
    "print(f'There are {len(phenol_fluxes)} fluxes that can be compared between the MFA and FBA')\n",
    "phenol_fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9408c4ea-55a7-4859-838a-4ada1879d93f",
   "metadata": {},
   "source": [
    "### Add Phenol EFLUX2 and SPOT flux values to phenol fluxes dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5864b3e7-3e48-4d32-b096-4d88a74f7c4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_flux_value(reaction_id, solution):\n",
    "    if reaction_id.startswith('reverse_'):\n",
    "        reaction_id = reaction_id.split('reverse_')[1]\n",
    "        return -1*solution.fluxes[reaction_id]\n",
    "    else:\n",
    "        return solution.fluxes[reaction_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0f84d6-4bfe-400b-8071-da8337ac9b95",
   "metadata": {},
   "outputs": [],
   "source": [
    "EFLUX2_fluxes = []\n",
    "#SPOT_fluxes = []\n",
    "for _, row in phenol_fluxes.iterrows():\n",
    "    reactions = row['Forward Reactions']\n",
    "    flux_value_eflux2 = 0\n",
    "    #flux_value_spot = 0\n",
    "    for x in [x.strip('() ') for x in reactions.split(' or ')]:\n",
    "        and_split = [y.strip('() ') for y in x.split(' and ')]\n",
    "        flux_value_eflux2 += min([get_flux_value(v, testscaleeflux2) for v in and_split])\n",
    "        #flux_value_spot += min([get_flux_value(v,spotsoldf) for v in and_split])\n",
    "    EFLUX2_fluxes.append(flux_value_eflux2)\n",
    "    #SPOT_fluxes.append(flux_value_spot)\n",
    "    \n",
    "phenol_fluxes['EFLUX2 Value'] = EFLUX2_fluxes\n",
    "#phenol_fluxes['SPOT Value'] = SPOT_fluxes\n",
    "phenol_fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c48b766-9871-40fd-8b05-afba0f5877b9",
   "metadata": {},
   "source": [
    "### Plot EFLUX2 vs 13C MFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bb88ab1-a3c2-4d9c-a132-238fef787cba",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.plot(ax.get_xlim(), ax.get_ylim(), ls=\"--\", c=\".3\")\n",
    "\n",
    "for pathway in list(set(phenol_fluxes['Pathway'])):\n",
    "    pathway_df = phenol_fluxes[phenol_fluxes['Pathway'] == pathway]\n",
    "    \n",
    "    measured_flux_list = list(pathway_df['Flux'])\n",
    "    simulated_flux_list = list(pathway_df['EFLUX2 Value'])\n",
    "    \n",
    "    ax.scatter(measured_flux_list, simulated_flux_list, label=pathway)\n",
    "\n",
    "\n",
    "# Dashed line\n",
    "x = np.linspace(*ax.get_xlim())\n",
    "ax.plot(x, x, ls=\"--\", c=\".3\")\n",
    "\n",
    "plt.title(r'Phenol 13C MFA vs. EFLUX2 Fluxes', fontSize=20)\n",
    "plt.xlabel(r'13C MFA flux (per 100 mmol of phenol uptake)')\n",
    "plt.ylabel(r'EFLUX2 flux (per 100 mmol of phenol uptake)')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1970487-c86c-450c-94c9-07d133446f92",
   "metadata": {},
   "source": [
    "-- Stopping here --"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e2eebd-f0a6-4fd9-afb9-d0ead616d55c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "565aadd8-ae1c-48fc-a175-d4476368b31e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17a82a57-e872-4726-88e9-b52500318e0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3510a539-0186-4bb5-8cf7-f94d9b9cb6e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f5b67a1-3478-4b65-b7bd-ec3a9e8b4c97",
   "metadata": {},
   "outputs": [],
   "source": [
    "from optlang.symbolics import add\n",
    "\"\"\"\n",
    "    Provides EFLUX2 predictions as explained in Machado et. al (2014) \n",
    "    \n",
    "        Parameters\n",
    "        ----------\n",
    "        model : cobrapy model.\n",
    "        Transcriptomics : pandas dataframe with transcriptomics data.\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        eflux2_sol as output from eflux2_model.optimize().\n",
    "        \n",
    "\"\"\"\n",
    "#Code only works for GPRs written in disjunctive normal form (DNF). Majority of models have them in DNF but there are some exceptions. \n",
    "\n",
    "def EFlux2test(model, df_trans):\n",
    "    # Convert EDD style dataframe to dictionary\n",
    "    transcriptomics = {row['Measurement Type']:row['Value'] for _, row in df_trans.iterrows()}\n",
    "    print(list(transcriptomics.keys())[:5])\n",
    "    with model:\n",
    "        \n",
    "        # Parse GPR into a dict containing isozymes (separated by 'or')\n",
    "        # Each isozyme has a set of subunits (separated by 'and')\n",
    "        #'and' and 'or' can occur at the same time, or can occur by itself.\n",
    "        gpr_dict = dict()\n",
    "        for r in model.reactions:\n",
    "            if r.gene_reaction_rule:\n",
    "                temp = set()\n",
    "                for x in [x.strip('() ') for x in r.gene_reaction_rule.split(' or ')]:\n",
    "                    temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "                gpr_dict[r.id] = temp\n",
    "                \n",
    "        # define a variable to hold all bounds\n",
    "        all_bounds = []\n",
    "                \n",
    "        # Set the bounds using the transcriptomics data\n",
    "        for r in model.reactions:\n",
    "            if r.gene_reaction_rule:\n",
    "                \n",
    "                #If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "                #If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "                #If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "                \n",
    "#                 t = np.sum([np.min([Transcriptomics.loc[g] if g in Transcriptomics.index\n",
    "#                                     else np.array([np.Inf]) for g in p])\n",
    "#                             for p in gpr_dict[r.id]])\n",
    "                t = np.sum([np.min([transcriptomics[g] if g in transcriptomics.keys()\n",
    "                                    else np.array([np.Inf]) for g in p])\n",
    "                            for p in gpr_dict[r.id]])\n",
    "                \n",
    "                # add bound to the list\n",
    "                if t != np.Inf:\n",
    "                    all_bounds.append(t)\n",
    "                \n",
    "                # Enforce transcriptomic bound\n",
    "                if r.lower_bound < 0.0:\n",
    "                    r.lower_bound = -t\n",
    "                if r.upper_bound > 0.0:\n",
    "                    r.upper_bound = t\n",
    "            else:\n",
    "                #When there is no GPR, the arbitrary bounds are removed. \n",
    "                #Common arbitrary bound value of 1000 for E.coli, might be different depending on the model, e.g., 99999.0 for iMM904 yeast model in BiGG\n",
    "                if r.lower_bound <= -1000.0:\n",
    "                    r.lower_bound = -np.Inf\n",
    "                if r.upper_bound >= 1000.0:\n",
    "                    r.upper_bound = np.Inf\n",
    "              \n",
    "        # These statements can help with scaling bounds to reasonable limits\n",
    "        print(all_bounds[:20])\n",
    "        print(len(all_bounds))\n",
    "        print(f'The average bound was {sum(all_bounds) / len(all_bounds)}')\n",
    "        print(f'The max bound was {max(all_bounds)}')\n",
    "        \n",
    "        # print the media to prevent inivisible errors\n",
    "        display(model.medium)\n",
    "        \n",
    "        # solve FBA with transcriptomic enforced bounds to calculate the maximum biomass\n",
    "        model.tolerance = 1e-9\n",
    "        fba_sol = model.optimize()\n",
    "        \n",
    "        print(f'The FBA solution was {fba_sol.status}.' )\n",
    "        print(f'Given the bounds from transcriptomics, flux balance analysis found that{fba_sol.objective_value:.3f} grams of biomass could be made.\\n')\n",
    "        \n",
    "        # Constrain the biomass to the optimal value\n",
    "        for r in model.reactions:\n",
    "            if r.objective_coefficient:\n",
    "                r.lower_bound = fba_sol.objective_value\n",
    "                \n",
    "        # minimize the sum of squared flux values\n",
    "        model.objective = model.problem.Objective(add([r.flux_expression**2 for r in model.reactions]), direction='min')\n",
    "        eflux2_sol = model.optimize()\n",
    "        \n",
    "        print(f'The EFLUX2 solution was {eflux2_sol.status}.' )\n",
    "        print(f'The minimum sum of squared fluxes to make {fba_sol.objective_value:.3f} grams of biomass given the bounds from transcriptomics is {eflux2_sol.objective_value:.1f}.\\n')\n",
    "\n",
    "    return eflux2_sol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dbde7cf-10d7-4adb-ba13-230142c780ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91390bb8-99d6-4895-a78a-2f79c9806f10",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcc4dc6c-620b-4c1b-89bc-b657aa0ed3aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eflux2sol_WTPR1, spotsol_WTPR1 = eflux2spotpred(model, df_Trans2, 'WT-P-R1', 'phenol', 100)\n",
    "# eflux2sol_WTPR2, spotsol_WTPR2 = eflux2spotpred(model, df_Trans2, 'WT-P-R2', 'phenol', 100)\n",
    "# eflux2sol_WTPR3, spotsol_WTPR3 = eflux2spotpred(model, df_Trans2, 'WT-P-R3', 'phenol', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edab2d08-2018-4ae7-a59f-f09441feba72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eflux2sol_WTPR1.fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d8d0611-3d4c-4027-9c3d-d527dc6c001d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eflux2sol_WTPR1_df = pd.DataFrame(eflux2sol_WTPR1.fluxes, columns=['fluxes'])\n",
    "# eflux2sol_WTPR2_df = pd.DataFrame(eflux2sol_WTPR2.fluxes, columns=['fluxes'])\n",
    "# eflux2sol_WTPR3_df = pd.DataFrame(eflux2sol_WTPR3.fluxes, columns=['fluxes'])\n",
    "# eflux2sol_WTPR = pd.concat([eflux2sol_WTPR1_df, eflux2sol_WTPR2_df, eflux2sol_WTPR3_df], axis=1)\n",
    "# eflux2sol_WTPR.rename(columns = {'fluxes':'R1', 'fluxes':'R2','fluxes':'R3'}, inplace=False)\n",
    "# eflux2sol_WTPR.mean(axis=1)\n",
    "# eflux2sol_WTPR.std(axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb8788d4-faa3-4075-9926-4559bfe9e5cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eflux2sol_WTPR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abb2c28f-530e-415e-ab92-03fdc08091f5",
   "metadata": {},
   "source": [
    "##### **Modularized way**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e44e04e9-5d89-4c5d-a80c-ca5f60b6ffa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function for EFLUX2 and SPOT Predictions:\n",
    "def eflux2spotpred(model, transcript_df, line_name, substrate, sub_uptake_rate):    \n",
    "    with model:\n",
    "        \n",
    "        # This is to prevent medium components other than glucose from being growth limitations\n",
    "        medium = model.medium\n",
    "        medium = {key:1000 for (key,value) in model.medium.items()}\n",
    "    \n",
    "        if substrate == 'phenol':\n",
    "            model.objective = 'Growth_Phenol'\n",
    "            model.reactions.get_by_id('Growth_Glucose').upper_bound = 0\n",
    "            model.reactions.get_by_id('Growth_Glucose').lower_bound = 0\n",
    "            medium[\"EX_glc__D_e\"] = 0.0\n",
    "            medium[\"EX_phenol_e\"] = sub_uptake_rate\n",
    "            \n",
    "        elif substrate == 'glucose':\n",
    "            model.objective = 'Growth_Glucose'\n",
    "            model.reactions.get_by_id('Growth_Phenol').upper_bound = 0\n",
    "            model.reactions.get_by_id('Growth_Phenol').lower_bound = 0\n",
    "            medium[\"EX_glc__D_e\"] = sub_uptake_rate\n",
    "            medium[\"EX_phenol_e\"] = 0.0\n",
    "            \n",
    "        else:\n",
    "            print('Unknown substrate: Please choose among phenol and glucose')\n",
    "            \n",
    "        model.medium = medium\n",
    "        \n",
    "        eflux2sol = EFlux2test(model, transcript_df[transcript_df['Line Name'] == line_name])\n",
    "        spotsol = SPOT(model, transcript_df[transcript_df['Line Name'] == line_name])\n",
    "    return eflux2sol, spotsol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cfa2911-aa1d-4f50-9da0-3ac46efa1883",
   "metadata": {},
   "outputs": [],
   "source": [
    "eflux2sol_WTPR1, spotsol_WTPR1 = eflux2spotpred(model, df_trans, 'WT-P-R1', 'phenol', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "391ad0d3-c426-4b20-97ac-13d3108ceada",
   "metadata": {},
   "outputs": [],
   "source": [
    "#display(eflux2sol_WTPR1.fluxes['Growth_Phenol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c6c4753-aba5-460e-a3f9-a4491b686e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#spotsol_WTPR1df =pd.DataFrame(spotsol_WTPR1, columns=['fluxes'])\n",
    "#spotsol_WTPR1df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "212e000a-362e-4ae7-87da-ca33b132d891",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function for predictions for three replicates and averaging the solutions and calculating the standard deviation:\n",
    "def eflux2spotpredthreereps(model, transcript_df, line_name1, line_name2, line_name3, substrate, sub_uptake_rate):\n",
    "    \n",
    "    #call prediction functions for individual EFLUX2 predictions for all 3 replicates:\n",
    "    eflux2sol_R1, spotsol_R1 = eflux2spotpred(model, transcript_df, line_name1, substrate, sub_uptake_rate)\n",
    "    eflux2sol_R2, spotsol_R2 = eflux2spotpred(model, transcript_df, line_name2, substrate, sub_uptake_rate)\n",
    "    eflux2sol_R3, spotsol_R3 = eflux2spotpred(model, transcript_df, line_name3, substrate, sub_uptake_rate)\n",
    "    \n",
    "    #EFLUX2 calculations:\n",
    "    eflux2sol_R1_df = pd.DataFrame(eflux2sol_R1.fluxes, columns=['fluxes'])\n",
    "    eflux2sol_R2_df = pd.DataFrame(eflux2sol_R2.fluxes, columns=['fluxes'])\n",
    "    eflux2sol_R3_df = pd.DataFrame(eflux2sol_R3.fluxes, columns=['fluxes'])\n",
    "    eflux2sol_all = pd.concat([eflux2sol_R1_df, eflux2sol_R2_df, eflux2sol_R3_df], axis=1)\n",
    "\n",
    "    eflux2sol = pd.DataFrame(eflux2sol_all.mean(axis=1), columns=['fluxes'])\n",
    "    eflux2sol_std = eflux2sol_all.std(axis=1)\n",
    "    \n",
    "    #SPOT calculations:\n",
    "    spotsol_R1_df = pd.DataFrame(spotsol_R1, columns=['fluxes'])\n",
    "    spotsol_R2_df = pd.DataFrame(spotsol_R2, columns=['fluxes'])\n",
    "    spotsol_R3_df = pd.DataFrame(spotsol_R3, columns=['fluxes'])\n",
    "    spotsol_all = pd.concat([spotsol_R1_df, spotsol_R2_df, spotsol_R3_df], axis=1)\n",
    "    \n",
    "    spotsol = spotsol_all.mean(axis=1)\n",
    "    spotsol_std = spotsol_all.std(axis=1)\n",
    "    \n",
    "    return eflux2sol, eflux2sol_std, spotsol, spotsol_std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fbb0754-d6a2-4a38-b1da-bc449a52816d",
   "metadata": {},
   "outputs": [],
   "source": [
    "eflux2sol, eflux2sol_std, spotsol, spotsol_std = eflux2spotpredthreereps(model, df_trans, 'WT-P-R1', 'WT-P-R2','WT-P-R3', 'phenol', 100)\n",
    "#eflux2sol, eflux2sol_std = eflux2spotpredthreereps(model, df_Trans2, 'WT-P-R1', 'WT-P-R2','WT-P-R3', 'phenol', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "267b9c37-7c79-475a-98dc-f529bfe19658",
   "metadata": {},
   "outputs": [],
   "source": [
    "eflux2soldf = pd.DataFrame(eflux2sol, columns=['fluxes'])\n",
    "eflux2solstddf = pd.DataFrame(eflux2sol_std, columns=['fluxes'])\n",
    "spotsoldf = pd.DataFrame(spotsol, columns=['fluxes'])\n",
    "spotsolstddf = pd.DataFrame(spotsol_std, columns=['fluxes'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9759f638-53ff-4067-a6ff-09cbf08b90c6",
   "metadata": {},
   "source": [
    "#### **Plot solutions: Comparison of SPOT and EFLUX2 solution**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0782d2e4-632a-4d5b-ad87-0643d4611df7",
   "metadata": {},
   "source": [
    "##### **Load 13cdata**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9c82e93-eb7d-4b72-91c6-4a27927b2fcd",
   "metadata": {},
   "source": [
    "####  **Get 13C MFA measured fluxes for phenol**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7807b73-d258-4fec-9c26-652535e22262",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenol_fluxes = pd.read_csv('../13C_flux_data/13C_phenol_flux_data.csv')\n",
    "phenol_fluxes.dropna(inplace=True)\n",
    "print(f'There are {len(phenol_fluxes)} fluxes that can be compared between the MFA and FBA')\n",
    "phenol_fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7992430a-5aee-4f12-9727-236620de1f0f",
   "metadata": {},
   "source": [
    "### Add Phenol EFLUX2 and SPOT flux values to phenol fluxes dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95187395-9c83-44d3-b7b8-b14ae4e31142",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_flux_value(reaction_id, solution):\n",
    "    if reaction_id.startswith('reverse_'):\n",
    "        reaction_id = reaction_id.split('reverse_')[1]\n",
    "        return -1*solution.fluxes[reaction_id]\n",
    "    else:\n",
    "        return solution.fluxes[reaction_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c49f728-3eb0-4cd9-aa59-84a593543913",
   "metadata": {},
   "outputs": [],
   "source": [
    "EFLUX2_fluxes = []\n",
    "SPOT_fluxes = []\n",
    "for _, row in phenol_fluxes.iterrows():\n",
    "    reactions = row['Forward Reactions']\n",
    "    flux_value_eflux2 = 0\n",
    "    flux_value_spot = 0\n",
    "    for x in [x.strip('() ') for x in reactions.split(' or ')]:\n",
    "        and_split = [y.strip('() ') for y in x.split(' and ')]\n",
    "        flux_value_eflux2 += min([get_flux_value(v, eflux2soldf) for v in and_split])\n",
    "        flux_value_spot += min([get_flux_value(v,spotsoldf) for v in and_split])\n",
    "    EFLUX2_fluxes.append(flux_value_eflux2)\n",
    "    SPOT_fluxes.append(flux_value_spot)\n",
    "    \n",
    "phenol_fluxes['EFLUX2 Value'] = EFLUX2_fluxes\n",
    "phenol_fluxes['SPOT Value'] = SPOT_fluxes\n",
    "phenol_fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e59f7309-6a75-43af-9490-dd865315e675",
   "metadata": {},
   "source": [
    "### Plot EFLUX2 vs 13C MFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "259d24b2-2bfd-46d3-ac7f-5461ee13a559",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.plot(ax.get_xlim(), ax.get_ylim(), ls=\"--\", c=\".3\")\n",
    "\n",
    "for pathway in list(set(phenol_fluxes['Pathway'])):\n",
    "    pathway_df = phenol_fluxes[phenol_fluxes['Pathway'] == pathway]\n",
    "    \n",
    "    measured_flux_list = list(pathway_df['Flux'])\n",
    "    simulated_flux_list = list(pathway_df['EFLUX2 Value'])\n",
    "    \n",
    "    ax.scatter(measured_flux_list, simulated_flux_list, label=pathway)\n",
    "\n",
    "\n",
    "# Dashed line\n",
    "x = np.linspace(*ax.get_xlim())\n",
    "ax.plot(x, x, ls=\"--\", c=\".3\")\n",
    "\n",
    "plt.title(r'Phenol 13C MFA vs. EFLUX2 Fluxes', fontSize=20)\n",
    "plt.xlabel(r'13C MFA flux (per 100 mmol of phenol uptake)')\n",
    "plt.ylabel(r'EFLUX2 flux (per 100 mmol of phenol uptake)')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01734cb6-69c9-4952-81e3-13c754d2bd38",
   "metadata": {},
   "source": [
    "### Plot SPOT vs 13C MFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a761d23a-ae2e-4566-88db-63f0bc558078",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.plot(ax.get_xlim(), ax.get_ylim(), ls=\"--\", c=\".3\")\n",
    "\n",
    "for pathway in list(set(phenol_fluxes['Pathway'])):\n",
    "    pathway_df = phenol_fluxes[phenol_fluxes['Pathway'] == pathway]\n",
    "    \n",
    "    measured_flux_list = list(pathway_df['Flux'])\n",
    "    simulatedspot_flux_list = list(pathway_df['SPOT Value'])\n",
    "    \n",
    "    ax.scatter(measured_flux_list, simulatedspot_flux_list, label=pathway)\n",
    "\n",
    "\n",
    "# Dashed line\n",
    "x = np.linspace(*ax.get_xlim())\n",
    "ax.plot(x, x, ls=\"--\", c=\".3\")\n",
    "\n",
    "plt.title(r'Phenol 13C MFA vs. SPOT Fluxes', fontSize=20)\n",
    "plt.xlabel(r'13C MFA flux (per 100 mmol of phenol uptake)')\n",
    "plt.ylabel(r'SPOT flux (per 100 mmol of phenol uptake)')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e89a67cb-b755-40a8-b3cf-9611298e9ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "rxns = [r.id for r in model.reactions]\n",
    "# #EFlux2_fluxdict = dict(zip(rxns,EFlux2_min.solution.get_values()))\n",
    "# EFlux2_fluxdict = dict(zip(rxns,eflux2solfluxes.to_numpy()))#replace previous .get_values()\n",
    "# SPOT_fluxdict = dict(zip(rxns,spotsolWTPR1.to_numpy()))#replace previous .get_values()\n",
    "# #display(SPOT_fluxdict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97cabe14-623a-45fa-a4d7-913e86973ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scatterplot of solutions\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "\n",
    "def maefunc(y_predict, y_actual):\n",
    "    \"\"\"Mean Absolute Error.\n",
    "    Multioutput case included.\"\"\"\n",
    "\n",
    "    if y_actual.ndim == 1:\n",
    "        return np.mean(np.abs([y_a - y_p for y_a, y_p in zip(y_actual, y_predict)]))\n",
    "    else:\n",
    "        return [\n",
    "            np.mean(\n",
    "                np.abs([y_a - y_p for y_a, y_p in zip(y_actual[:, i], y_predict[:, i])])\n",
    "            )\n",
    "            for i in range(y_actual.shape[1])\n",
    "        ]\n",
    "    \n",
    "def scatterplotcompare(observed, predicted, labels, strtitle):\n",
    "    fig, ax = plt.subplots(figsize=(8, 8))\n",
    "    #res = stats.linregress(observed.to_numpy().astype(float), predicted.to_numpy())\n",
    "    #print(res.rvalue**2)\n",
    "    lims = [\n",
    "                np.min([observed, predicted]),  # min of both axes\n",
    "                np.max([observed, predicted]),  # max of both axes\n",
    "            ]\n",
    "    ax.set_xlim(lims)\n",
    "    #ax.plot([0.0, 0.2], [0.0, 0.2], 'k',  c='0.55')\n",
    "    # Plot Diagonal Dashed Line\n",
    "    ax.plot(lims, lims, ls=\"--\", color=\".8\", zorder=0)\n",
    "    for i in range(0, len(observed)):\n",
    "        ax.scatter(observed[i], predicted[i])\n",
    "        #ax.annotate(str(labels[i]),(observed[i],predicted[i]))\n",
    "    #ax.annotate(\"r-squared = {:.3f}\".format(round(r2_score(observed, predicted))), (0, 1))\n",
    "    #plt.plot(observed, res.intercept + res.slope*observed, 'k', label='fitted line')\n",
    "    rss = np.sum((predicted - observed) ** 2)\n",
    "    tss = np.sum((predicted - predicted.mean()) ** 2)\n",
    "    r2_1 = np.round(1 - rss / tss,2)\n",
    "    #display(r2_1)\n",
    "    #r2 = round(r2_score(observed, predicted),2)\n",
    "    mse = np.round(1/len(observed) * rss,2) \n",
    "    rmse = np.round(np.sqrt(1/len(observed) * rss),2)\n",
    "    #display(rmse)\n",
    "    #display(r2_score(observed, predicted))\n",
    "    mae_score = np.round(maefunc(observed, predicted),2)\n",
    "    plt.xlabel(r'EFLUX2')#Observed growth rates [$mmol/gDW/hr$]')\n",
    "    plt.ylabel(r'SPOT')#Predicted growth rates [$mmol/gDW/hr$]')\n",
    "    #plt.title(r'Growth rates: Observed vs. Predicted ('+strtitle+'), \\n' + f\"$R^2$={r2_1}, MAE={mae_score}, MSE = {mse}, RMSE={rmse}\", fontsize=18)\n",
    "    plt.title(r'EFLUX2 vs. SPOT ('+strtitle+'), \\n' + f\"$R^2$={r2_1}, MAE={mae_score}, MSE = {mse}, RMSE={rmse}\", fontsize=18)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0537a8eb-489b-4d5a-a409-f94185068f80",
   "metadata": {},
   "outputs": [],
   "source": [
    "scatterplotcompare(eflux2soldf.to_numpy(), spotsoldf.to_numpy(), labels='', strtitle='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cdf2f9a-363e-4b17-acbe-b9f9db7f4e86",
   "metadata": {},
   "outputs": [],
   "source": [
    "scatterplotcompare(eflux2soldf.loc['Growth_Phenol',:], spotsoldf.loc['Growth_Phenol',:], labels='Growth', strtitle='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a15513-f561-4e16-825c-2caf0b75e264",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c2a0376-c9c9-45ae-b02b-3d0c51f0cd7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from optlang.symbolics import add\n",
    "\"\"\"\n",
    "    Provides EFLUX2 predictions as explained in Machado et. al (2014) \n",
    "    \n",
    "        Parameters\n",
    "        ----------\n",
    "        model : cobrapy model.\n",
    "        Transcriptomics : pandas dataframe with transcriptomics data.\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        eflux2_sol as output from eflux2_model.optimize().\n",
    "        \n",
    "\"\"\"\n",
    "#Code only works for GPRs written in disjunctive normal form (DNF). Majority of models have them in DNF but there are some exceptions. \n",
    "\n",
    "def EFlux2test(model, Transcriptomics):\n",
    "    #with model as eflux2_model:\n",
    "    eflux2_model = model.copy()\n",
    "    # Parse GPR into a dict containing isozymes (separated by 'or')\n",
    "    # Each isozyme has a set of subunits (separated by 'and')\n",
    "    #'and' and 'or' can occur at the same time, or can occur by itself.\n",
    "    gpr_dict = dict()\n",
    "    for r in eflux2_model.reactions:\n",
    "        if r.gene_reaction_rule:\n",
    "            temp = set()\n",
    "            for x in [x.strip('() ') for x in r.gene_reaction_rule.split(' or ')]:\n",
    "                temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "            gpr_dict[r.id] = temp\n",
    "    #display(gpr_dict)\n",
    "    # Set the bounds using the transcriptomics data\n",
    "    for r in eflux2_model.reactions:\n",
    "        t2=0\n",
    "        if r.gene_reaction_rule:\n",
    "            #If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "            #If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "            #If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "            t = np.sum([np.min([Transcriptomics.loc[g].values if g in Transcriptomics.index\n",
    "                                else np.array([np.Inf]) for g in p])\n",
    "                        for p in gpr_dict[r.id]])\n",
    "            for p in gpr_dict[r.id]:\n",
    "                xg=[]\n",
    "                for g in p:\n",
    "                    if g in Transcriptomics.index:\n",
    "                        xg.append(Transcriptomics.loc[g].values)\n",
    "                    else:\n",
    "                        xg.append(np.array([np.Inf]))\n",
    "                    t2=np.min(xg)\n",
    "                t2+=t2\n",
    "            if t2!=t:\n",
    "                display('difference',t2, t)\n",
    "            if t != np.inf:\n",
    "                display(t)\n",
    "            if r.lower_bound < 0.0:\n",
    "                r.lower_bound = -t\n",
    "            else:\n",
    "                pass\n",
    "            if r.upper_bound > 0.0:\n",
    "                r.upper_bound = t\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "            #When there is no GPR, the arbitrary bounds are removed. \n",
    "            #Common arbitrary bound value of 1000 for E.coli, might be different depending on the model, e.g., 99999.0 for iMM904 yeast model in BiGG\n",
    "            if r.lower_bound < -99999:\n",
    "                r.lower_bound = -np.Inf\n",
    "            if r.upper_bound > 99999:\n",
    "                r.upper_bound = np.Inf \n",
    "            if r.lower_bound >= 0:\n",
    "                r.lower_bound = 0\n",
    "            if r.upper_bound <= 0:\n",
    "                r.upper_bound = 0\n",
    "\n",
    "#             if r.lower_bound < 0.0 and r.lower_bound > -1000.0:\n",
    "#                 print(r.id, r.lower_bound, r.upper_bound)\n",
    "#                 r.lower_bound = -1000.0\n",
    "#             elif r.lower_bound > 0.0:\n",
    "#                 print(r.id, r.lower_bound, r.upper_bound)\n",
    "#                 r.lower_bound = 0.0\n",
    "#             if r.upper_bound > 0.0 and r.upper_bound < 1000.0:\n",
    "#                 print(r.id, r.lower_bound, r.upper_bound)\n",
    "#                 r.upper_bound = 1000.0\n",
    "#             elif r.upper_bound < 0.0:\n",
    "#                 print(r.id, r.lower_bound, r.upper_bound)\n",
    "#                 r.upper_bound = 0.0\n",
    "#             if r.lower_bound == -1000.0:\n",
    "#                 r.lower_bound = -np.Inf\n",
    "#             if r.upper_bound == 1000.0:\n",
    "#                 r.upper_bound = np.Inf\n",
    "\n",
    "#             if r.lower_bound < 0.0:\n",
    "#                 r.lower_bound = -np.Inf\n",
    "#             elif r.lower_bound >= 0.0:\n",
    "#                 r.lower_bound = 0.0\n",
    "#             elif r.upper_bound > 0.0:\n",
    "#                 r.upper_bound = np.Inf\n",
    "#             elif r.upper_bound <= 0.0:\n",
    "                r.upper_bound = 0.0\n",
    "#                 if r=='Phenol_e':\n",
    "#                     display(r)\n",
    "#                     r.lower_bound = -100\n",
    "#                     r.upper_bound = 0\n",
    "    display(eflux2_model.medium)\n",
    "    # solve FBA to calculate the maximum biomass\n",
    "    eflux2_model.tolerance = 1e-9\n",
    "    fba_sol = eflux2_model.optimize()\n",
    "    print('FBA status', fba_sol.status)\n",
    "    print('FBA solution', fba_sol.objective_value)\n",
    "    display(eflux2_model.objective)\n",
    "    # Constrain the biomass to the optimal value\n",
    "    for r in eflux2_model.reactions:\n",
    "        if r.objective_coefficient:\n",
    "            r.lower_bound = fba_sol.objective_value\n",
    "    # minimize the sum of squared flux values\n",
    "    eflux2_model.objective = eflux2_model.problem.Objective(add([r.flux_expression**2 for r in eflux2_model.reactions]), direction='min')\n",
    "    display(eflux2_model.problem.Objective)\n",
    "    eflux2_sol = eflux2_model.optimize()\n",
    "    print('EFlux2 status', eflux2_sol.status)\n",
    "    print('EFlux2 solution', eflux2_sol.objective_value)\n",
    "    # return eflux2 solution\n",
    "    return eflux2_sol"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biodesign_3.7",
   "language": "python",
   "name": "biod_3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
