{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7563604a-2ba5-4e5d-b9e0-309615df0b5b",
   "metadata": {},
   "source": [
    "# **Notebook to make predictions with reprocessed Yoneda and Henson data and EFLUX2 and SPOT** "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a6edbe7-d3d0-49ac-9d13-8f43882945f0",
   "metadata": {},
   "source": [
    "### **Import python packages**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5e37f811-1a95-45e0-932f-962edb144307",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cobra\n",
    "from statistics import mean, median\n",
    "%matplotlib inline\n",
    "from matplotlib import pyplot as plt\n",
    "from edd_utils import login, export_study, export_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e20d0fe-15c8-4248-b6be-aa65f868fc95",
   "metadata": {},
   "source": [
    "### **Import Joonhoons Code**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "faf7a9d4-095c-42c6-a32b-5b5fd19ffca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OptlangDir = '../EFLUX2_SPOT'\n",
    "# sys.path.append(OptlangDir)\n",
    "# from spot_cplex import SPOT #spot not working yet because quadratic constraints not supported yet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "244d4ba7-9d90-4f21-8de8-18a77f1a87dd",
   "metadata": {},
   "source": [
    "### **Sign into EDD**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e8232d35-b2bf-4b51-9625-3a12a76661f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Password for garrettroell:  ················\n"
     ]
    }
   ],
   "source": [
    "study_slug = 'biodesign_henson2018_reprocessed'\n",
    "edd_server = 'public-edd.jbei.org'\n",
    "user = 'garrettroell' # 'schenkch'\n",
    "session = login(edd_server=edd_server, user=user)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8bfde81-2fb8-4b07-9273-4070bc5fea43",
   "metadata": {},
   "source": [
    "### **Get transcriptomic data from [EDD](https://public-edd.jbei.org/s/biodesign_henson2018_reprocessed/)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f177a0d5-a877-4bd0-af30-304b0badbc93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a38864e51da64a59acb4cbccf829c88f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=440493.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study ID</th>\n",
       "      <th>Study Name</th>\n",
       "      <th>Line ID</th>\n",
       "      <th>Line Name</th>\n",
       "      <th>Line Description</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Assay ID</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Formal Type</th>\n",
       "      <th>Measurement Type</th>\n",
       "      <th>Compartment</th>\n",
       "      <th>Units</th>\n",
       "      <th>Value</th>\n",
       "      <th>Hours</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14905</td>\n",
       "      <td>WT-M-R1</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14948</td>\n",
       "      <td>WT-M-R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14905</td>\n",
       "      <td>WT-M-R1</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14948</td>\n",
       "      <td>WT-M-R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14906</td>\n",
       "      <td>WT-M-R2</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14949</td>\n",
       "      <td>WT-M-R2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14906</td>\n",
       "      <td>WT-M-R2</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14949</td>\n",
       "      <td>WT-M-R2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14409</td>\n",
       "      <td>Biodesign_Henson2018_reprocessed</td>\n",
       "      <td>14907</td>\n",
       "      <td>WT-M-R3</td>\n",
       "      <td>R. Opacus wild type strain with Mixture of all...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>14950</td>\n",
       "      <td>WT-M-R3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Study ID                        Study Name  Line ID Line Name  \\\n",
       "0     14409  Biodesign_Henson2018_reprocessed    14905   WT-M-R1   \n",
       "1     14409  Biodesign_Henson2018_reprocessed    14905   WT-M-R1   \n",
       "2     14409  Biodesign_Henson2018_reprocessed    14906   WT-M-R2   \n",
       "3     14409  Biodesign_Henson2018_reprocessed    14906   WT-M-R2   \n",
       "4     14409  Biodesign_Henson2018_reprocessed    14907   WT-M-R3   \n",
       "\n",
       "                                    Line Description         Protocol  \\\n",
       "0  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "1  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "2  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "3  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "4  R. Opacus wild type strain with Mixture of all...  Transcriptomics   \n",
       "\n",
       "   Assay ID Assay Name Formal Type Measurement Type  Compartment Units  Value  \\\n",
       "0     14948    WT-M-R1         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "1     14948    WT-M-R1         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "2     14949    WT-M-R2         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "3     14949    WT-M-R2         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "4     14950    WT-M-R3         NaN   WP_000104864_1            0  FPKM    0.0   \n",
       "\n",
       "   Hours  \n",
       "0   20.0  \n",
       "1   32.0  \n",
       "2   20.0  \n",
       "3   32.0  \n",
       "4   20.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = export_study(session, study_slug, edd_server=edd_server)\n",
    "trans_df = df[df['Protocol'].str.contains('Transcriptomics')]\n",
    "trans_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7252a93-7cea-4b7c-8375-c03a194a9524",
   "metadata": {},
   "source": [
    "### **Load Genome Scale Model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f6b96da-3aaf-49fd-9dec-b674a678d243",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name =  '../GSMs/Ropacus_annotated_curated_with_phenol_custom_biomass.xml'\n",
    "model = cobra.io.read_sbml_model(file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c050bfe4-28bd-42d3-a42d-7eb33402d132",
   "metadata": {},
   "source": [
    "### **Define functions used in main EFlux function**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f157a170-aa60-403a-9d41-903bb8b454d4",
   "metadata": {},
   "source": [
    "Define function to plot histogram from list of numbers. Will be used for transcriptomic measurements and flux bounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b39e769d-8315-4adb-b033-7b5dbc6a8cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_histogram(data, title):\n",
    "\n",
    "    bins = np.arange(0, 400, 5) # x axis goes from 0 to 40 with bars of width 5\n",
    "    plt.hist(data, bins=bins, alpha=0.5) \n",
    "    plt.title(f'{title}\\nmean = {mean(data):.3f}, median={median(data):.3f} max={max(data):.3f}')\n",
    "    plt.xlabel('Transcriptomics value')\n",
    "    plt.ylabel('count')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a779be49-3fc2-4cf7-a1e2-00df3b5ddb89",
   "metadata": {},
   "source": [
    "Define a function to find the transcriptomic value for a given reaction id. The function takes into account genes that are required in parallel and in series."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e82e91bd-a148-4bdd-a4d3-fb4a1fc5bc60",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rxn_id_to_upper_bound(model, trans_df, rxn_id):\n",
    "    associated_genes = model.reactions.get_by_id(rxn_id).gene_reaction_rule\n",
    "    parallel_transcript_values = []\n",
    "    \n",
    "    if associated_genes:\n",
    "        for parallel_gene_string in associated_genes.split(' or '):\n",
    "            parallel_gene_string = parallel_gene_string.strip('() ')\n",
    "            \n",
    "            series_transcript_values = []\n",
    "            for gene_name in parallel_gene_string.split(' and '):\n",
    "                gene_name = gene_name.strip('() ')\n",
    "                if gene_name == 'spontaneous':\n",
    "                    series_transcript_values.append(np.Inf) # A spontaneous reaction should not require transcripts\n",
    "                else:\n",
    "                    transcript_value = float(trans_df[trans_df['Measurement Type'] == gene_name].Value)\n",
    "                    series_transcript_values.append(transcript_value)\n",
    "            \n",
    "            parallel_transcript_values.append(min(series_transcript_values))\n",
    "            \n",
    "        return sum(parallel_transcript_values)\n",
    "    else:\n",
    "        return np.Inf # Reactions without transcript measurements shouldn't be constrained"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32f10d13-e30e-4d78-bc3c-3e386c47a181",
   "metadata": {},
   "source": [
    "### **Main EFlux function**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f38895b6-054a-404e-a293-56847f42a1f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from optlang.symbolics import add\n",
    "\n",
    "def EFlux2test(model, edd_df, line_name, min_bound = 10):\n",
    "    \n",
    "    # get transcriptomics measurement for a single experiment and plot values\n",
    "    trans_df = edd_df[edd_df['Line Name'] == line_name]\n",
    "    plot_histogram(trans_df['Value'], 'Transcriptomic values by genes')\n",
    "    \n",
    "    # make a dataframe mapping reaction to transcript levels using cobrapy's gene_reaction_rule\n",
    "    reaction_transcript_df = pd.DataFrame()\n",
    "    reaction_transcript_df['reaction_id'] = [rxn.id for rxn in model.reactions]\n",
    "    reaction_transcript_df['transcipt_level'] = [rxn_id_to_upper_bound(model, trans_df, rxn_id) for rxn_id in reaction_transcript_df.reaction_id]\n",
    "    plot_histogram(reaction_transcript_df['transcipt_level'], 'Transcriptomic values by reactions')\n",
    "    display(reaction_transcript_df.head(3))\n",
    "    \n",
    "    with model:        \n",
    "        # Ensure that medium is correct, and the carbon source is the only limiting substrate\n",
    "        medium = {key:np.Inf for (key,value) in model.medium.items()}\n",
    "        medium[\"EX_glc__D_e\"] = 0.0\n",
    "        medium[\"EX_phenol_e\"] = 100.0 # 1.76124\n",
    "        model.medium = medium\n",
    "\n",
    "        # calculate maximum biomass yield without constraints\n",
    "        model.tolerance = 1e-9\n",
    "        unbounded_sol = model.optimize()\n",
    "        print(f'\\nWithout constraints, flux balance analysis found that {unbounded_sol.objective_value:.3f} grams of biomass could be made.\\n')\n",
    "\n",
    "        # boundry reactions should also not be given infinite bounds/leads to media problems.\n",
    "        # TODO: Need to add phenol uptake reactions to this list too.\n",
    "        reactions_to_exclude = [rxn.id for rxn in model.boundary]\n",
    "        \n",
    "        # calculate maximum biomass yield with constraints\n",
    "        bounded_model = model.copy()\n",
    "        with bounded_model:\n",
    "            for rxn_id in reaction_transcript_df['reaction_id']:\n",
    "                if rxn_id in reactions_to_exclude:\n",
    "                    pass\n",
    "                else:\n",
    "                    bound = rxn_id_to_upper_bound(bounded_model, trans_df, rxn_id) # can replace with a dataframe method to access data in reaction_transcript_df\n",
    "                    bound = max(bound, min_bound) # min_bound is a parameter of this function\n",
    "                    bounded_model.reactions.get_by_id(rxn_id).upper_bound = bound\n",
    "                    bounded_model.reactions.get_by_id(rxn_id).lower_bound = -bound\n",
    "                    \n",
    "            medium = {key:np.Inf for (key,value) in bounded_model.medium.items()}\n",
    "            medium[\"EX_glc__D_e\"] = 0.0\n",
    "            medium[\"EX_phenol_e\"] = 100.0 # 1.76124\n",
    "            bounded_model.medium = medium\n",
    "            print(bounded_model.medium)\n",
    "\n",
    "            bounded_sol = bounded_model.optimize()\n",
    "            print(f'\\nGiven the bounds from transcriptmics and a minimum absolute bound of {min_bound}, flux balance analysis found that {bounded_sol.objective_value:.3f} grams of biomass could be made.\\n')\n",
    "\n",
    "#         # Constrain the biomass to the optimal value\n",
    "#             print('FBA status', bounded_sol.status)\n",
    "#             print('FBA solution', bounded_sol.objective_value)\n",
    "#             display(bounded_model.objective)\n",
    "#             # Constrain the biomass to the optimal value\n",
    "#             for rxn_id in bounded_model.reactions:\n",
    "#                 if rxn_id.objective_coefficient:\n",
    "#                     rxn_id.lower_bound = bounded_sol.objective_value\n",
    "#             # minimize the sum of squared flux values\n",
    "#             bounded_model.objective = bounded_model.problem.Objective(add([rxn_id.flux_expression**2 for rxn_id in bounded_model.reactions]), direction='min')\n",
    "#             display(bounded_model.problem.Objective)\n",
    "#             eflux2_sol = bounded_model.optimize()\n",
    "#             print('EFlux2 status', eflux2_sol.status)\n",
    "#             print('EFlux2 solution', eflux2_sol.objective_value)\n",
    "#     return eflux2_sol"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f441bd3-e7cf-469c-a0a2-db836dd8818f",
   "metadata": {},
   "source": [
    "### **Main EFlux function test**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1ca8506a-4510-4ddc-b388-b1e3c1ec68b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAElCAYAAAAV9s4VAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3debxVdb3/8ddbcAocQHECFAfMuJZoXCHT4joPmVZe0yzJIcq0rCxDvSWllb+6pWmmkeI8a+WQpuSQWdcBTHFWUksQAUUU0Mzh8/vj+92y2Ox91j569tn7wPv5eOzHWev7XcNnr73P/qz1XWt9lyICMzOzjizX6gDMzKz9OVmYmVkpJwszMyvlZGFmZqWcLMzMrJSThZmZlXKysB5D0nckndnqOAAk3STpgG5c36GSbuuu9eV13iHp8925TmtfvVsdgHU/SQsKo+8BXgPezONfjIiLuj+qchFxQqPTSjoRGBQRn29SLDs3Y7lm7crJYhkUEX0rw5KeBg6NiD/Wm15S74h4oztia+cYzJZlboayJUg6UdJlki6RNB/4rKQPSbpT0jxJMyWdKmn5PH1vSSHpi5KmSXpR0qmF5W0q6XZJL0l6XtLFhbr3S/qjpLmSnpN0dAcxnCjp3Fy/SV7nFyQ9m19fz3UfA44GDpC0QNKUXD5I0nV5XU9IOrjqPV+a17dA0v2SNpb0P5LmSPqnpB0L0y/WRJPf+6OS5kt6UNIWNbbrryWdVFX2e0lfzcP/I+nJvIyHJH28zueziaSoKquO59Acz4uSbpA0OJcvlz+72fnzmCppWM0vQjJU0uQ87W8l9cvLuVHSYVUxPCxpzzoxH5S34fOSjpU0XdLoQkzHSvp7rr+0sJ7K53xgnmeOpHGF5XY073skXSzphfy9vVvSmh28V+tIRPi1DL+Ap4Edq8pOBP4N7EnaoVgZ+E9gJOlodCPgceCIPH1vIICrgdWAIcDcynKBK4Bv52WtBHw4l68GzAKOBFYEVgW27iCGE4Fzc/0meZ0XkJrStgBeAEYX5j+36n39BTgtx7AV8Dzw0cL0rwI75vdzMfAUMC6PHwY8UVjWHcDn8/D+wDPABwEBmwKDa2zr7fP2Vh5fI69z7Ty+L7Bufr+fARYU6g4Fbiu+96plF+P5FPAY8N4c+3jgz7luD+DuvO2XA4YB69T5btyR39cwoA/wu8L2/wzwl8K0HwRmA71rLOf9wHxgm/w5nwy8UfisjsqfzcD82ZwFXFD1OZ9Z+NxeA4Y2MO/hOeaVgV7ACKBvq//neuqr5QH41eIvQP1kcUvJfN8ErsjDlWQxqlD/G+Cbefhi4AxgYNUyPgfcU2f5S8RA7WSxSaH+Z8CvqqfN4xsCrwN9CmU/Ac4qTH9Doe4TwEvAcnm8X15f3zxe/HG+GTi8gW29HDAD2CaPHwbc1MH0DwJ75OHOJItJwJhCXe/8AzsQ2Bl4lJT4lyuJ9w7gxML4B4B/kRLiysA8YMNcdwpwap3lfL/yA57H+7B4sniCnLTz+OC8nuUKn/M6hfp7gX0amHdsfg/vb/X/2dLwcjOU1fNMcUTSZrnJ5DlJL5N+AKoP6Z8rDL8CVM6NHAUsD0yW9ICkMbl8MPD3RmNoYJp/AOvVmW494PmIWFg1/cDC+KzC8KvAnIh4qzAOi95TUdn7ACAv6zLSkQikvfO3LyaQ9Pnc/DVP0jxgM5bcxo3YADi9sJzngbdIJ/xvIu2lnwHMknSmpFU6WFb19l0R6B8RrwJXkpoHewH7kY7yalmvuJz8GbxYqF8fuLYQ7wO5fK3CPPW+Wx3Ney7wR+BySTMknSTJ52nfIScLq6e6O+JfkfZ0N4mIVYHvkvYwyxcUMTMiDo2IdUlNAxMkbUj6Adm4EzHUMrgwvD7wbJ15nwXWlNSnavoZDayjTNn7KLoE+O/8/rciHYEhaSPSD/hhwBoRsTrpCKDWNl6Y53lPoWydqngOiYjVC6+VI+IugIg4JSK2AjYnNTF9o4N4q7fva6QmRoDzgANIRysvRsQ9dZYxExhUGcmfQb9C/XRgp6p4V6pKEPXUnTci/h0R4yPifcC2pKPFbrvceWnjZGGNWoXULLNQ0vuALzY6o6R9JVX24OeRfsjfBK4B1pd0hKQVJa0qaetOxvUdSStLej8whrTnDukoYYgkAUTEU8Bk4Id5XcOBg4ALO7m+Ws4Cjpa0pZKhlRPK1fIP6svABOD6iJifq/qStsscQJK+QDqyqOW5/PqspF6SxpKOJirOBI7LnxOSVpe0Tx7eOr96k5LOv0lHHfUcmI8q+wDfAy6P3N5DauJZAfh/1D+qgHTOam9JoyStQDoqLTqT9Lmsn2Ncq97J/Rrqzitpe0mbS1qOtM1fL3mv1gEnC2vUUaQf4/mko4zLOp58MSOBeyQtJO1JHx4R/4yIl4CdSCdkZ5FOmn+0k3HdATwJ3AT8KCJuyeWXkX7I5kq6O5d9GhhK+qG9Ejg2Im7r5PqWEBGXkH4wLyP9KP2Gxfecq11COpH+9lVhETGVdPL9btKe+HuBu+qsL4AvAMeSmpg2KU4bEVeQzt9ckZsMpwK75OrVgbNJSfvpvK6fdRDrBaSEOpN0kvhrVXGcTzpCqXtvTn5vXycljWdJFyK8QDpKIa//D8DNSle+/ZV0QUUjOpp3PdJn8TLwEKlJ6uJaC7FyWrSTYNZzSNqEdHVSQ01h1hxKlx8fGBGjOzHPqqRktUFENHJeytqAjyzM7B3JTVNfJjWplU378XzfQ1/gp8C9ThQ9i5OFmXWapD1I91X8k8aaJD9BaoKaTroPZ/8Op7a242YoMzMr5SMLMzMr5WRhVkO+Qe6OwviCfC+E2TLJyWIple9dmCzpNeXO9wp1oyRNUupQb46kKyStW6hfMd/ZOytPc23hPola6xouaYqkV/Lf4YW6byl1rDdf0lOSvlVj/iNz3UJJj0jatIs2Q5eJiL4R8WR3rU/SmLwtX1bqQO/HxbuPJfVX6thvoaR/SPpMB8v6ulIHhS8rdbh4ctWyhki6NX9+j2rxDhPPzImy8notX6LaI3T0/cv3ZFySt8lLkv4iaWShfrSkt6re/5hC/fsk3ZLnnSbpE3Vi+K5SZ4g71qkvi2MPpY4i5yn1oHCWqu66l7SjpHvz92G6pH3fzXarxcli6fUsqb+jiTXq+pGuYBlCuplrPnBOof5I4EOkvoDWI3XNcFqtlSjdZHU16Vr8fqS7eq/O5ZDuQD4w1+0KHCFpv8L8hwKHkDq46wt8jHTvwLLuPaR7GtYk3aeyA6k/rorTSTfUrU26K/kMSf9RZ1nXAFvlO+83J3W6+NVC/SXA30gdGx4HXClpAEBEfCknyr6Rura/hHS/RE/R0fevL3APqRPE/qTv7u/zFVsVzxbff0ScB6mnZdL3/ro871jgwuodHUkbA/9Nuk+lnrI4ViP9L68HvI/URc1PCusYRrp/5Lg87RbAlIa2Tme0unOqdnmRblD6FukGpoWkG5fWBm4g/Zj+EehXmH4U6QagecD95E7Rct1BwCN5vidJDxSq1I0mXRFyFOlqkpnAQU18X0v0vlpjmq2A+YXxM4AfF8b3AB6rM+/OpC4zVCj7J7BrnelPBU7Lw8uRuqbY4R2+t9vy+/srqYfWa0k/eBeRbsS6BxhSmH4zUid7c0m9su5bqFuD9KP6MunGuBOAOwr1b3damLfH3/K0zwDjC9MNydOOydvheeC4LvgcvwFcm4f7kBLFpoX6C4CTGljOGvm7/Ms8vinp5rhVCtP8GfhSjXn75O/0RztYfpAup30iT3sCqSuUv+btdTmwQp62H+nHdg5ph+Q6Uv9VkH40pwN75vG+wDTSPR3vZju+/f2rU/8y8ME8PBqYXme6zfN3rvi9vwk4oWq6PwC7U6PDzpI4346jRt0ngQcK4xdXr7cZLx9ZLO5TpDuKNyV1jX0D6S7ZAaQftspzBwYCvyf9UPUn7fFdVdkbIyWBj5G63D4IOFnSVoX1rEPaAxhI2qs+XbkP/mqSfpkPP2u9pnbR+/4I6Q7XirOBD0taT6n/oQNI26KW/wCmRv7WZlNzefV7EbBdYV2D8mtzSc/kZoLvKXXP0Kj9SL3XDiT9KP0f6SipPylhH5/X3YeUKC4mdTK3H/BLLXqWw+mk3krXBQ7Or3oWkvZWVycljsMk7V01zbaku7B3AL6rRV1vfKaDz3OecrcVNRQ/o02BNyLi8UL9/dTY5hV5vS+TktcWpLvwyfM8GYu6HeloWZ8i/bDfXm892S6kveRRpOeKTAA+S+pnanMWXTa7HOmz2oDU79SrwC8AImIu6TP4taS1SN2a3xcR5+f3M66j7VhnG1R//6rrh5Pu+p9WKF5LqTn2qdx816fWvJVF5PdXWd5/A69FxPUdzNNoHEXV/6+j8nwPKD1r5kJJ/TuzzoY0Oxv1lBcp8x9QGL8KOKMw/hXgd3n42xS6XM5lN1LoFrqq7nfAkXl4NOmfonehfjaF7r27+H11eGRBamqaC2xXKFsNuJS0l/gGaS+6f535vwNcWlV2EYW97UL590g/RCvm8W3yOn5P+uEdQury4wsNvrfbKOy1k272KnYzvifpBwZSVx9/rpr/V6Rk0ovUb9BmhbofUufIokYcpwAn5+EhedpBhfq7gf3exWd4MGkve808vh3wXNU0XyB3YV6yrKGkvf118vjngDurpvlBre8MqSv2JT7XqmmC/LySPD4F+HbVZ3RKnXmHkzokLJadRupJdgapg8V387+w2Pevqm7VvJ5jCmXrkDpaXI7Uxf3tLOoCf3lSq8HReXhn0tHejbl+FdLR1ZA8/jQNHFnUiqOqfifSUVjxqPLfefmbko7ArgIuejfbqtbLRxaLq+6iunq80oa4Aann0OKezLakvVIk7ab0VLm5uW53Fu9q+oVY/BGhxS6Xu41Slxk3kBLZnwtVp5O6ol6D1PTwG+ofWSwgfcGLViU1QRTXdQRpb3yPiKj0CVTp9vvHETEvIp4m/YDv3om30ZnPbGTVZ3YA6QdhAOmZD9XdcdckaaTSCeE5kl4CvkTj3bV3Sj5i+RGwW0RUzuU0tM1riYgnSHulv+zMsvIRz2hSX1BlGvpMlO7o/pXSCfqXST/Gqyt1eV4xgbS3fm5EvNDAumuq8/2r1K1MasK8MyJ+VCmP1HPtwxHxVqSOKI8mHV0REa8De5OOLJ8jNStfTkrqkB44dUH+TjcaY804CvWjSEfG+8TiR5WvAudExOMRsYC0o9OZ/6GGOFm8M8+QvgjFbpH7RMRJklYkZfb/JT3lbHXgehrszrualrwapfiqeTjd4HI3ILVdnxAR1T2GDif9c87N/1inAVur9iMpHwI+kA/xKz5A4TBZqf+gcaRzE9ML0z1G2isqNmE16y7RZ4A/VX1mfSPiMFLTyhss2R13PReTzm8MjojVSD2fNvT5Sjqgg89zQbEZStKuwK9J7fYPFBbzONBb0tBC2RbUaV6poTeLulR/CNhIi19dU2tZnyM9Ga8rrwg7itRUNzLSyfeP5HIB5KQxgZSgvpx3bsh1x3a0HYsr6eD7R/5//R3pR76sJ+Wg8JsZEVMj4qMRsUZE7EJ6gmSl08odgK8qXb30HOm7dbmkb9dacFkckrYkfecOjoibq6qn0g3/Q04W78yFwJ6SdlHqInolpcvsBpHaGlck/wBJ2o10iPqORNXVKFWvjtqoe0taidTEUomxd64bCNwC/CIizqwx+z2krqlXU3rO9pdJV4XUukrpNlJ3419VuuT2iFx+S17XAaQ9nZ2qf2gi4hVSVxFHS1olb7+xpBOdlUs6Q9KQBjZVmeuATSV9TtLy+fWfkt4XEW+Sjp7G573dYaQT1PWsAsyNiH8pdale97LVahFxUQefZ9+I+Cek7rVJzXmfioi7q5ZR6b33+5L6SPowsBd1uglXeh73Wnl4GHAMqUmJvId6H3B8/o58gpTsr6pazIGkhwl1pVVIe8Xzchv78VX1x5J++A4mXf1zfuWoIyJ+2NF2rCygo+9f/m5fmWMYE4sedFWp/y9JGygZDJxEugKqUv+BvM3eI+mbpJaFc3P1DqQjouH59SwpCZxevREaiGNz0onyr0TEtTW24znAQZI2UjrHOI78P9Slurpdq6e+qGpTJCWE8YXxQ4E/FsZHAn8itffPIbW7r5/rDicdes8j/QNfSn48JTWusKhedxe9n/Gkf7Tia3yuOz6PLyi+CvNWriiand/DHeRnY+f6G0jde1fGtyS1Tb9KeuTlloW6p0jnA4rrOrNQv2rePvNJe//fZVE3NNvlbbN8nfd4G3BoYbz6Uao7AtMK4+/Nn9McUhfZtwDDc90A0j9YI1dD7UNqppqf5/kFcGGuG5Kn7V0vzgY/v1tJRzvF7VY8H9OftCe6kHTV1WcKddtVfZ7n5O/jwrw9fwKsVKgfkmN8lXS0V/2Y3Q/leVdpIO7Fzu1QeNxr4TOqPMp2vbzeBaSjpS9Wth3pBPmLhW3ei/Ss7U5dWdbR94/UHX6QmgmL9dvl+m+QzpW8QvpunsriV439JMe4gPQ/UfOcVp3flzM7Ecc5pOdwFOseqlr+90jf6zmk35x+ndlOjbzcN5S1LUn/Q3q06a9KJzazpnKyMDOzUj5nYWZmpZwszMyslJOFmZmV6l0+Sc+z5pprxpAhQ1odhplZjzJlypTnI2JArbqlMlkMGTKEyZMntzoMM7MeRVLdngvcDGVmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZWysnCzMxKOVmYmVkpJwszMyvlZGFmZqWWyju4362TJz2+2PjXd9q0RZGYmbUHH1mYmVkpJwszMyvlZGFmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZWysnCzMxKOVmYmVkpJwszMyvlZGFmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZWqmnJQtJgSbdKeljSQ5KOzOX9JU2S9ET+2y+XS9KpkqZJmippq8KyxuTpn5A0plkxm5lZbc08sngDOCoihgGjgMMlDQPGATdHxFDg5jwOsBswNL/GAmdASi7A8cBIYGvg+EqCMTOz7tG0ZBERMyPi3jw8H3gEGAjsBZyXJzsP2DsP7wWcH8mdwOqS1gV2ASZFxNyIeBGYBOzarLjNzGxJ3XLOQtIQYEvgLmDtiJiZq54D1s7DA4FnCrNNz2X1ys3MrJs0PVlI6gtcBXwtIl4u1kVEANFF6xkrabKkyXPmzOmKRZqZWdbUZCFpeVKiuCgifpOLZ+XmJfLf2bl8BjC4MPugXFavfDERMSEiRkTEiAEDBnTtGzEzW8Y182ooAWcDj0TEzwpV1wCVK5rGAFcXyg/MV0WNAl7KzVU3AjtL6pdPbO+cy8zMrJv0buKyPwx8DnhA0n257FjgJOBySYcA/wD2zXXXA7sD04BXgIMAImKupBOAe/J034+IuU2M28zMqjQtWUTEHYDqVO9QY/oADq+zrInAxK6LzszMOsN3cJuZWSknCzMzK+VkYWZmpZwszMyslJOFmZmVcrIwM7NSThZmZlbKycLMzEo5WZiZWSknCzMzK+VkYWZmpZwszMyslJOFmZmVcrIwM7NSThZmZlbKycLMzEo5WZiZWSknCzMzK+VkYWZmpZwszMyslJOFmZmVcrIwM7NSThZmZlbKycLMzEo5WZiZWSknCzMzK+VkYWZmpZwszMyslJOFmZmVcrIwM7NSThZmZlbKycLMzEo5WZiZWSknCzMzK+VkYWZmpZwszMyslJOFmZmVcrIwM7NSTUsWkiZKmi3pwULZeEkzJN2XX7sX6o6RNE3SY5J2KZTvmsumSRrXrHjNzKy+Zh5ZnAvsWqP85IgYnl/XA0gaBuwH/Eee55eSeknqBZwO7AYMA/bP05qZWTfq3awFR8TtkoY0OPlewKUR8RrwlKRpwNa5blpEPAkg6dI87cNdHK6ZmXWgFecsjpA0NTdT9ctlA4FnCtNMz2X1ypcgaaykyZImz5kzpxlxm5kts7o7WZwBbAwMB2YCP+2qBUfEhIgYEREjBgwY0FWLNTMzmtgMVUtEzKoMS/o1cF0enQEMLkw6KJfRQbmZmXWTbj2ykLRuYfQTQOVKqWuA/SStKGlDYChwN3APMFTShpJWIJ0Ev6Y7YzYzsyYeWUi6BBgNrClpOnA8MFrScCCAp4EvAkTEQ5IuJ524fgM4PCLezMs5ArgR6AVMjIiHmhWzmZnV1syrofavUXx2B9P/APhBjfLrgeu7MDQzM+sk38FtZmalnCzMzKyUk4WZmZVysjAzs1JOFmZmVsrJwszMSjlZmJlZKScLMzMr5WRhZmalnCzMzKyUk4WZmZVysjAzs1JOFmZmVsrJwszMSjlZmJlZqYaShaSbGykzM7OlU4cPP5K0EvAe0tPu+gHKVasCA5scm5mZtYmyJ+V9EfgasB4whUXJ4mXgF02My8zM2kiHySIifg78XNJXIuK0borJzMzaTEPP4I6I0yRtAwwpzhMR5zcpLjMzayMNJQtJFwAbA/cBb+biAJwszMyWAQ0lC2AEMCwiopnBmJlZe2r0PosHgXWaGYiZmbWvRo8s1gQelnQ38FqlMCI+3pSozMysrTSaLMY3MwgzM2tvjV4N9admB2JmZu2r0auh5pOufgJYAVgeWBgRqzYrMDMzax+NHlmsUhmWJGAvYFSzgjIzs/bS6V5nI/kdsEsT4jEzszbUaDPUJwujy5Huu/hXUyIyM7O20+jVUHsWht8AniY1RZmZ2TKg0XMWBzU7EDMza1+NPvxokKTfSpqdX1dJGtTs4MzMrD00eoL7HOAa0nMt1gOuzWVmZrYMaDRZDIiIcyLijfw6FxjQxLjMzKyNNJosXpD0WUm98uuzwAvNDMzMzNpHo8niYGBf4DlgJrAP8PkmxWRmZm2m0Utnvw+MiYgXAST1B/6XlETMzGwp1+iRxQcqiQIgIuYCW3Y0g6SJ+cqpBwtl/SVNkvRE/tsvl0vSqZKmSZoqaavCPGPy9E9IGtO5t2dmZl2h0WSxXOWHHd4+sig7KjkX2LWqbBxwc0QMBW7O4wC7AUPzayxwRmE9xwMjga2B44txmJlZ92g0WfwU+D9JJ0g6Afgr8OOOZoiI24G5VcV7Aefl4fOAvQvl5+d+p+4EVpe0Lqn/qUkRMTcf2UxiyQRkZmZN1ugd3OdLmgxsn4s+GREPv4P1rR0RM/Pwc8DaeXgg8Exhuum5rF75EiSNJR2VsP7667+D0MzMrJ5GT3CTk8M7SRD1lheSonzKhpc3AZgAMGLEiC5brpmZvYMuyt+lWbl5ifx3di6fAQwuTDcol9UrNzOzbtTdyeIaoHJF0xjg6kL5gfmqqFHAS7m56kZgZ0n98ontnXOZmZl1o4aboTpL0iXAaGBNSdNJVzWdBFwu6RDgH6Qb/QCuB3YHpgGvAAdBukQ3n1C/J0/3/XzZrpmZdaOmJYuI2L9O1Q41pg3g8DrLmQhM7MLQzMysk7q7GcrMzHogJwszMyvlZGFmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZWysnCzMxKNe2mvKXJyZMef3v46ztt2sJIzMxaw0cWZmZWysnCzMxKOVmYmVkpJwszMyvlZGFmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZWysnCzMxKOVmYmVkpJwszMyvlZGFmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZWysnCzMxKOVmYmVkpJwszMyvlZGFmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZWysnCzMxKOVmYmVkpJwszMyvlZGFmZqVakiwkPS3pAUn3SZqcy/pLmiTpify3Xy6XpFMlTZM0VdJWrYjZzGxZ1ruF6/6viHi+MD4OuDkiTpI0Lo9/G9gNGJpfI4Ez8t+WOHnS44uNf32nTVsUiZlZ92mnZqi9gPPy8HnA3oXy8yO5E1hd0rqtCNDMbFnVqmQRwE2Spkgam8vWjoiZefg5YO08PBB4pjDv9Fy2GEljJU2WNHnOnDnNitvMbJnUqmaobSNihqS1gEmSHi1WRkRIis4sMCImABMARowY0al5zcysYy05soiIGfnvbOC3wNbArErzUv47O08+AxhcmH1QLjMzs27S7clCUh9Jq1SGgZ2BB4FrgDF5sjHA1Xn4GuDAfFXUKOClQnOVmZl1g1Y0Q60N/FZSZf0XR8QfJN0DXC7pEOAfwL55+uuB3YFpwCvAQd0fspnZsq3bk0VEPAlsUaP8BWCHGuUBHN4NoZmZWR3tdOmsmZm1KScLMzMr5WRhZmalWtndx1LB3X+Y2bLARxZmZlbKycLMzEo5WZiZWSknCzMzK+VkYWZmpZwszMyslJOFmZmVcrIwM7NSvimvi/kmPTNbGvnIwszMSjlZmJlZKTdDNZmbpcxsaeAjCzMzK+VkYWZmpZwszMyslJOFmZmV8gnubuYT3mbWE/nIwszMSjlZmJlZKTdDtVixWcpNUmbWrnxkYWZmpXxk0UZ88tvM2pWTRQ/iZGJmreJmKDMzK+UjizZWfSRhZtYqThY9WFkycTOVmXUVN0OZmVkpH1kso3yy3Mw6w8liKeaEYGZdxcliGeIT5mb2TjlZGOCT5WbWMScLa0hnk4n7vDJbujhZWJfoKJmUnTt5t81jTkZmzedkYd2umedO3Jxm1hw9JllI2hX4OdALOCsiTmpxSNYmOpN8OpuoOkouXX21mZvurJ0pIlodQylJvYDHgZ2A6cA9wP4R8XCt6UeMGBGTJ09+x+vzVUNmjXs3zYo9KSm+m52DnnIZu6QpETGiZl0PSRYfAsZHxC55/BiAiPhRremdLMxsWfVuElFHyaKnNEMNBJ4pjE8HRhYnkDQWGJtHF0h67F2sb03g+Xcxf7O0a1zQvrG1a1zQvrG1a1zQvrG1TVzfWLKoM7FtUK+ipySLUhExAZjQFcuSNLledm2ldo0L2je2do0L2je2do0L2je2do0Lui62ntKR4AxgcGF8UC4zM7Nu0FOSxT3AUEkbSloB2A+4psUxmZktM3pEM1REvCHpCOBG0qWzEyPioSauskuas5qgXeOC9o2tXeOC9o2tXeOC9o2tXeOCrmqe7wlXQ5mZWWv1lGYoMzNrIScLMzMr5WRRIGlXSY9JmiZpXBvE87SkByTdJ5pAM6AAAAiISURBVGlyLusvaZKkJ/Lfft0Qx0RJsyU9WCirGYeSU/M2nCppqxbENl7SjLzd7pO0e6HumBzbY5J2aWJcgyXdKulhSQ9JOjKXt3y7dRBbS7ebpJUk3S3p/hzX93L5hpLuyuu/LF/kgqQV8/i0XD+kGXGVxHaupKcK22x4Lu/u/4Nekv4m6bo83vXbLCL8SudtegF/BzYCVgDuB4a1OKangTWryn4MjMvD44D/1w1xfATYCniwLA5gd+AGQMAo4K4WxDYe+GaNaYflz3VFYMP8efdqUlzrAlvl4VVI3dUMa4ft1kFsLd1u+b33zcPLA3flbXE5sF8uPxM4LA9/GTgzD+8HXNbEbVYvtnOBfWpM393/B98ALgauy+Ndvs18ZLHI1sC0iHgyIv4NXArs1eKYatkLOC8Pnwfs3ewVRsTtwNwG49gLOD+SO4HVJa3bzbHVsxdwaUS8FhFPAdNIn3sz4poZEffm4fnAI6SeCFq+3TqIrZ5u2W75vS/Io8vnVwDbA1fm8uptVtmWVwI7SFJXx1USWz3d9nlKGgTsAZyVx0UTtpmTxSK1uhTp6B+oOwRwk6QpSt2ZAKwdETPz8HPA2q0JrW4c7bIdj8iH/xMLTXUtiS0f6m9J2httq+1WFRu0eLvl5pT7gNnAJNJRzLyIeKPGut+OK9e/BKzRjLhqxRYRlW32g7zNTpa0YnVsNeLuaqcARwNv5fE1aMI2c7Job9tGxFbAbsDhkj5SrIx0LNnya5/bJY6CM4CNgeHATOCnrQpEUl/gKuBrEfFysa7V261GbC3fbhHxZkQMJ/XSsDWwWXfHUE91bJI2B44hxfifQH/g290Zk6SPAbMjYkqz1+VksUjbdSkSETPy39nAb0n/PLMqh7P57+wWhVcvjpZvx4iYlf+x3wJ+zaImk26NTdLypB/jiyLiN7m4LbZbrdjaZbvlWOYBtwIfIjXhVG4gLq777bhy/WrAC82Mqyq2XXOTXkTEa8A5dP82+zDwcUlPk5rOtyc996fLt5mTxSJt1aWIpD6SVqkMAzsDD+aYxuTJxgBXtybCunFcAxyYrwYZBbxUaHbpFlVtw58gbbdKbPvlK0I2BIYCdzcpBgFnA49ExM8KVS3fbvVia/V2kzRA0up5eGXS82seIf0w75Mnq95mlW25D3BLPlrrcnVie7SQ+EU6L1DcZk3/PCPimIgYFBFDSL9Zt0TEATRjmzXr7HxPfJGuYHic1E56XItj2Yh0Bcr9wEOVeEjtizcDTwB/BPp3QyyXkJolXie1fx5SLw7S1R+n5234ADCiBbFdkNc9Nf9zrFuY/rgc22PAbk2Ma1tSE9NU4L782r0dtlsHsbV0uwEfAP6W1/8g8N3C/8LdpBPrVwAr5vKV8vi0XL9RE7dZvdhuydvsQeBCFl0x1a3/B3mdo1l0NVSXbzN392FmZqXcDGVmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZWysnC2pKkNQo9eT6nxXtDXaGFcR0u6YCSabbP19Z31TpvrNxz0x0kbZK7tTB7W494rKoteyLiBVK3E0gaDyyIiP8tTpNvhFKkO46bTlLviDi9gUm3B54H7uyK9UZE07pSN2uUjyysR8l7vQ9Luoh0s+K6kiZImqz0nIHvFqadrvSMhr/ljt42zeXbKz2X4D5J9+Y75JF0rNLzQ+6X9INcdkfuIG4yqZO9EyV9rVB3Sl7OA5JGSNoYOBT4Vi7fJvcKcGuOYZJSL6FIulDS6UrPFfi7pI9IOk/So5LOrnoflbuHD8rLuV/SOblsP0kP5rJba2yzK1V4BkVe796SNpb057x9pkgaWWPeQyWdUhj/g6Rt8/Bukv4vb8PLKtvRlk4+srCeaDPgwIioPBBqXETMVerr5lZJV0bEw3naWRGxpaSvkvr8/xLwLWBsRNyl1JnevyTtSeqwceuIeFVS/8L6ekXEiLyuE6tiWTEihkvaHjgrD58FPB8Rp+R5bsh1Fyn1HnwKi7piWC0iRkr6FHAtqS+kR4F7JW0eEcWHOm1B6qhum/x+KzEeD4yOiFmVpFLlMmBf4EZJKwEfJd3p3gvYKSL+JWkzUtfVSySMWiStRXoexw4R8Yqk44AjgR82Mr/1PD6ysJ7o75VEke0v6V7gXuB9pIf1VFQ68JsCDMnDfwF+LukrwKoR8SawIzAxIl4FiIjiMzIu6yCWS/L0twBr5eRTbSSpkzeA84HtCnXX5r8PAM9GxMO5We3hQrwV25MeVjO3Ksa/AOdLOpTa/9O/B3ZS6jxwD1J/QK+RHmZ0ttJTBi9l8e1WZps8/V/z+Y0DasRrSxEfWVhPtLAyIGkoaY9264iYJ+lCUv83Fa/lv2+Sv+8RcaKka0g/nHdK2qHR9dVQ3V9OZ/vPqcT3VmG4Mt7o/+cXSAnpY6Qjki0j4sW3A0p7/neQOr/7NOnpbgBHkZ5t8FnSw3wWsKQ3WDwBVbatgD9ExOcajNF6OB9ZWE+3KjAfeFmpB9DSk8GSNo6IqRHxI9LRyHtJD9o5WKlHUaqaoTry6Tz9aFKT18IcT/HqpTtJzUCQfphvb3DZ1W4BPl2JrRDjRpGexvYd4EVqP2TnMlLT04dI7xVS99QzI3UQN4aUAKo9DWypZAjwwVz+V+CjkjbKsfTJiduWUj6ysJ7uXlKTzaPAP0hNMmW+KWk70t77VOCmiPh3PicwWdLrpOah7zSwrNdzM0wv4KBcdjVwhaRPAofn10RJxwCzCtN1SkTcL+nHwO2S3iA1rR0CnKzUdbjye3mwxux/IJ2TuCIiXs9lvwCulHQwqanqtRrz/Yn0DIRHSBcU3JdjmSXpEOAyLbqU+VhSb7q2FHKvs2bvUG7aOSIifE+CLfXcDGVmZqV8ZGFmZqV8ZGFmZqWcLMzMrJSThZmZlXKyMDOzUk4WZmZW6v8D13Lng8+Zj0YAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAElCAYAAAALP/6mAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3debgcVZnH8e+PhE32JWJIAmEJMuhAYDKAiJphURaZAIMIgxIQDCIoKqIso6IDo4wiiDgwIFsAWQVBBpVdjchygyGEsBhZTGI2lkCCEEl8549zulLp9L23b5Jebu7v8zz93KpT29unb9fbdarqlCICMzMzgFVaHYCZmbUPJwUzMys4KZiZWcFJwczMCk4KZmZWcFIwM7OCk4I1jaSvSbq41XEASLpL0hFN3N6xkh5o1vbyNsdJOqqZ21zRJPWTNF/SZq2Opa/o3+oArOckzS+NvgNYACzK48dFxLXNj6p7EfGf9c4r6SxgcEQc1aBYPtyI9drykTQO+HFEXAkQEYuAtVsaVB/jpNALRUTxJZH0AnBsRNzT2fyS+kfEwmbE1s4x2Irhz3Ll5uajlZCksyTdIOk6SfOAT0h6n6SHJM2VNEPSBZJWzfP3lxSSjpM0RdKrki4orW8bSb+R9JqklyT9pDTtHyXdI+kVSTMlfaWLGM6SdGWevnXe5qcl/SW/vpinfRT4CnBEbjoYn8sHS7ojb+uPkj5V9Z6vz9ubL+lxSVtJ+g9JcyT9WdJepfmXaFrJ7/1pSfMkTZK0Q416vVTSd6rK/k/S5/Pwf0h6Lq/jSUn/2snns7WkqCqrjufYHM+rkn4haUguXyV/drPz5zFR0nY1/xGSYZI68ry3Stogr+dXko6vimGypAM6i1fS0ZL+DNyVy99f+p+aIOmDVfE/leviT5KOrVrnwXmZ1/P/3IclnQO8D7g4f4bnl/43h+bl1pd0Tf5MX5B0miSVtvlrSeflmJ6T9OHSNo/Jy8zL0w7rot76rojwqxe/gBeAvarKzgL+BhxASvxrAv8M7EI6OtwSeBY4Mc/fHwjgNmA9YCjwSmW9wE3AV/O61gDen8vXA2YBJwGrA+sCO3cRw1nAlXn61nmbV5OawHYAXgZGlpa/sup9/Q74YY5hJ+Al4EOl+d8E9srv5yfA88Cpefx44I+ldY0DjsrDhwNTgX8CBGwDDKlR13vk+lYe3yhvc5M8figwML/ffwfml6YdCzxQfu9V6y7H82/AM8C7c+xnAr/N0/YHHsl1vwqwHfCuTv43xuX3tR2wFvCzUv3/O/C70rz/BMwG+tdYT+WzuiJ/VmsCQ/Ln9ZEcxz7589goL3MA6f9Mud7eBLbP03YD5gJ75mWHAO+uroeq/82hefwnwC3AOnn9U4DRpTp+G/gU0A/4HDA1T1sXeA0YlscHAtu1+vvbjq+WB+DXcn6AnSeF+7pZ7svATXm48sXbtTT9FuDLefgnwEXAoKp1fBJ4tJP1LxUDtZPC1qXp3wf+t3rePL5F/sKvVSr7Lqn9uTL/L0rTDso7gVXy+AZ5e2vn8fJO+F7ghDrqehVgOrBbHj8euKuL+ScB++fhniSFuys7utLnswAYBHwYeJqU4FfpJt5xwFml8e2Bt0g76jVJO+Yt8rTzgQs6WU/ls9qsVHYGcEXVfPcCR3SyjjsqdQxcBny3i5iPqnrvQfqhsiqwENimNP0E4J5SHT9dmrZuXnbjPDw3/1+s0YzvZm99uflo5TW1PCJp29zUMVPS68C3SF+Wspml4b+y+ATfyaQvZIekJySNzuVDgD/VG0Md87wIbNrJfJsCL0XEG1XzDyqNzyoNvwnMiYi/l8ah9knL7t4HAHldN5COLCD92i5O6ks6KjdbzZU0F9iWpeu4HpsDPyqt5yXg76QT73cBF5OS9CxJF0tap4t1Vdfv6sCGEfEmcDOpWa8fcBjpqK0r5XVtDhxeiTHHuSv585P0UUkP56a+uaRkVqmLuuq7hneSjgBerHpP5f+B6v9hSD8EXid9bicAM3Mz5DbLEMNKz0lh5VXd/e3/kn65bh0R6wJfJ/1i7H5FETMi4tiIGEj6Ul0iaQvSTmKrHsRQy5DS8GbAXzpZ9i/AxpLWqpp/eh3b6E5376PsOuBj+f3vRDqiQtKWpB318aQmlPVJv+hr1fEbeZl3lMreVRXPMRGxfum1ZkQ8DBAR50fETsB7SU1DX+oi3ur6XUBqGgS4CjiCtMN+NSIe7eqNR/75XYrxiqoY14qI70pak5Rwvk1qPlufdB5CpWU7q++u/mdmk66y27zqPdX1PxARv4iIvUhNR1NI3wmr4qTQd6xDak55Q9I/AMfVu6CkQyVVfo3NJX1xFwG3A5tJOlHS6pLWlbRzD+P6mqQ1Jf0jMJr0SxzSr/6hlZOIEfE80AH8V97WcOBo4Joebq+WHwNfkbSjkmGVE7vV8o7zdeAS4M6ImJcnrU2qlzmAJH2adKRQy8z8+oTSdfhjWHJHdzFwRv6cKidXD8nDO+dXf1Jy+RvpKKIzR+ajxLWAbwI3lnbu44DVgHPo/iih2tXAQZL2zu9hDUn/ImlT0tHIarkuFildOLBnadnLgGPz/KsoXUDw7jxtFulcwVIi4m1SsvkvSWvnxPxF6vgfkDRQ0gE5Ef+NVHdd1Vuf5aTQd5xM2unOI/1CuqHr2ZewC/CopDdIv4xPiIg/R8RrwN6kE6OzSCevP9TDuMYBz5F+SX47Iu7L5TeQdiyvSHokl30cGEbaod4MnB4RD/Rwe0uJiOtIO8YbSDv8W0jnIDpzHemEdnEVVkRMJJ0EfwSYQTpJ/HAn2wvg08DppKahrcvzRsRNpPMrN+WmvomkE7oA65N2qnNJ55Nm5Hk7czVppzmD1PTyhao4xpKOOHp0b0tEvEBqn/8aaef/Z9L/2CoRMZe0s76VdFRyCOmcQmXZB/P7v4D0Q+V+Fh/RnM/iZqla7+uzpJ36C8CvSUc7Y+sIuR9wCqkeXiad7D6h3vfbl2jJI0Kz5pC0NelqoLqasKwxlC7rPTIiRrY6FmsPPlIw66Nyk9JnSU1hZoCTglmfJGl/0onbP9OzpkRbybn5yMzMCj5SMDOzgpOCNYRS3z8j65z33bkfnHnK/Qi1iqQrlXpoRdIHJD3TynhsxZN0hKS7ejD/WUp9fs3sfu7ez0nBGiIi3tODy0W/AtwfEetExAXdzt0kEfHbiHh393OuOPmekAcl/VVdPH9B0pFKHcUdWyqTpHMkvZxf51Tu8+hmm5fndW1dKhsq6U6lDvlmSrow3xtRmT5c0vgc5/h830ivEBHXRp1dpys9x+FkUj9J7+pu/pWBk4K1g82BJ1sdRJt4hXSt/nc6m0Gpp9PTWbrOxgAHkjoX3J7UKV2XNylK2p3adxf/D+lE9EBgOOn+k8/mZVYjdZ54Del+jquA23L5ymYz4OWImN3qQJrFSaFNKHXpe4pSV8hvSLpM0iZK3SbPU+qeeoPS/LvmX5RzlfrbGVmadrQWd1v8nKTjStNGSpom6WSl7pdnSDq6Qe9nrzx8pqQbJY3V4m6lR+Rp9wH/Alyo1F1yt/3R5F+1n1XqPnuepP9U6ib7QaWumG8s76CU+uGZkOvqQUnbl6btKOmxvJ4bSD2wVqaNlDStNH6qUjfQ85S6mT6oNO0ope6vv5d/XT8vad+e1ltE3BMRN7K4u49avk268eulqvLRwLkRMS0ipgPnAkd1tpL8y/+HpN5Eq21Buvv5rYiYCfwSeE+eNpLUUd35EbEgH91VekOttZ0HchPMg/kz/rmkjSRdmz+vR5W7xs7z/0DS1DxtvKQPlKbdKenc0vj1ki7v7D12Es9RSg/zqYyHpM/k/6e5kn6Uj7r2InVOuGmO+8qebKfXanWPfH6lF+kOzYeATUgdfM0GHgN2JO2o7gO+kecdRLorcz9SYt87jw/I0/cn/foT6RfeX4Gd8rSRpJ4mv0Xq5G6/PH2DTuL6H9Lds7VeE7t5P5Wut88k9c65H+nO0m8DD5XmfYD0oKB666rSzfe6pB3VAlIPnVuSupSezOLulHfMdblL3vboHFulK4YXSXffrkq68/Ztcs+iua6mlbb7MVKHb6uQ7q5+AxiYpx2Vl/103s7xpB27lqUeKfWqWlW+M6m7j1Wq6410d/AupfERwLwu6vEU4AelOi33WHsc6U7hd5D+3yYBB+VpX6TUI20uuwM4uZPtPEDqa2ir0ufzLIu7OR9LqcdV4BOkbsn7k5puZpJ7NiX1ETWblICOIN0Nv06etnsXdTwX2L30WY2r+n+6g3S3+GakO7T3qfU/0BdeLQ/Ar/xBpB3VEaXxnwIXlcY/B/wsD38VuLpq+V9R6m65atrPgJPy8EhSj6H9S9NnU+o2ewW+n3JSuKc0bTvgzdL4Eju3OtYd5Gc65PHxwFdL4+eSfsVC6qTuP6uWf4aULD9Iacedpz1IJ0mhRhwTgFF5+ChgSmnaO3KcNZ91UMd7XCopkJJNR+Wzqq43Un9U25bGh+UYVGP9Q0g76vVKdVpOCv+Q63VhnnYlixPc14Drq9Z3LXBmJ+/lAeCMqs+n3M35AcCELuriVWCH0vi/kTrVe4m8o+9h3R7F0klh99L4jcCp9fwPrIwvNx+1l+qun6vHK90+b07qqbPcbfHupPZfJO2r9ESsSrfF+7FkF84vx5KPUyx3k90o1V0ar1E+cbkMelJXJ1fV1RDSL/5NgemRv/1ZuVvmJSid3J1QWs97WbJei/cYEUW3zT18X135LOmo4qFOps8nHT1VrAvMr3p/FecD34rUf9USJK1Cai66hfRwno1J5w7O6WQ7lW3No3P1fl5I+nJu/nwt1/N6LFnPPyclyGciYhwrRmfdxvc5Tgq901TSkUJ1t8XfkbQ66SjjeyzutvhO6uwmu5pSf/3zO3n1hpPDU4Gzq+rqHZE6wZsBDJKWuEJns1orkbQ5cClwIou7xp5EnfW6gupxT1LPpDOVLo/cDThX0oV5+pOkk8wVO9D5Cfw9ge+W1gXwe0n/DmxIqocLI50zeJn01LX9StvZvqretu9iW3XL5w++QnqK3Qa5nl9jyXo+G3gKGCjp8PKyXdTx/PK5Ceuck0LvdA1wgKSPaHG3xSMlDSa1k69OahddmE921nX5XS0R8ZmIWLuT13u6X0PP5ROBL6yg1V0KfEbSLvnk4VqS9ld6MM3vSc0jn5e0qqSDSW32tazF4q6xUTo5/956g6i3HiufJ6k9fZX82a6aJx9FatYZnl8dpO6wz8jTxwJfkjRIqQvrk0nNPrVsQ0oalXVBasa5NSJeIj3K9HilZySvTzoXMzHP9wCpqerzSt2Yn5jLKz3cLo91SJ/JHKC/pK9TOipReg700cCROaYfKnfrHukS4s7qeO2I+O0KiG+l56TQC0XEVGAU6bLEOaRfw6eQui2eB3ye1C76KunpYLe3KNRlNYT0POblFhEdpJO/F5LqYwr5ipyI+BtwcB5/hXTy+JZO1jOZ1Bb+e1LTxz+uqBirfJLUnHIR8IE8fGmOYW5EzKy8SF1Iv15qAvpfUtPKE6SjmP+j9CCZ8q/liJhdtS5IT7arPKHuYNJzl+eQ6uxt0gnmSr0dSNoxzyU9E/nAXL68fkVqunqW1JT3FvmJb5LWJSW+EyNiet7JXwZcUXXUYsvBfR9Z21G62/SkiHiq1bGY9TVOCmZmVnDzkZmZFZwUzMys4KRgZmaF5bl5qOU23njjGDp0aKvDMDPrVcaPH/9SRAyoNa1XJ4WhQ4fS0dHR6jDMzHoVSZ3eue/mIzMzKzQsKeQ7MR9R6tb5SUnfzOVXKnUrPCG/hudySbpA0hSl7qN3alRsZmZWWyObjxYAe0TE/Hyb/jhJv8jTTomIm6vm35fUq+MwUjfHF+W/ZmbWJA07Uohkfh5dNb+6ulNuFDA2L/cQsL6kgY2Kz8zMltbQcwq5c68JpP76746Ih/Oks3MT0Xm5V09ID/KYWlp8Wi6rXucYSR2SOubMmdPI8M3M+pyGJoWIWBQRw4HBwM6S3gucBmwL/DOpi96v9nCdl0TEiIgYMWBAzSuqzMxsGTXl6qOImAvcT3rE3YzcRLSA1Ed7pavi6aTeMSsG5zIzM2uSRl59NCD3w46kNUnPEX66cp4gd3V7IKmLX0jdOx+Zr0LaFXgtImY0Kj4zM1taI68+GghcJakfKfncGBF3SLpP0gDSk5QmAJ/J899JerLTFNLj8I5uYGxmZlZDw5JCREwEdqxRvkcn8wdwQqPiqXbe3c8uMf7Fvbdp1qbNzNqW72g2M7OCk4KZmRWcFMzMrOCkYGZmBScFMzMrOCmYmVnBScHMzApOCmZmVnBSMDOzgpOCmZkVnBTMzKzgpGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmZlZwUjAzs4KTgpmZFZwUzMys4KRgZmYFJwUzMys0LClIWkPSI5Iel/SkpG/m8i0kPSxpiqQbJK2Wy1fP41Py9KGNis3MzGpr5JHCAmCPiNgBGA7sI2lX4BzgvIjYGngVOCbPfwzwai4/L89nZmZN1LCkEMn8PLpqfgWwB3BzLr8KODAPj8rj5Ol7SlKj4jMzs6U19JyCpH6SJgCzgbuBPwFzI2JhnmUaMCgPDwKmAuTprwEb1VjnGEkdkjrmzJnTyPDNzPqchiaFiFgUEcOBwcDOwLYrYJ2XRMSIiBgxYMCA5Y7RzMwWa8rVRxExF7gfeB+wvqT+edJgYHoeng4MAcjT1wNebkZ8ZmaWNPLqowGS1s/DawJ7A0+RksMhebbRwG15+PY8Tp5+X0REo+IzM7Ol9e9+lmU2ELhKUj9S8rkxIu6QNBm4XtJZwB+Ay/L8lwFXS5oCvAIc1sDYzMyshoYlhYiYCOxYo/w50vmF6vK3gI81Kh4zM+ue72g2M7OCk4KZmRWcFMzMrOCkYGZmBScFMzMrOCmYmVnBScHMzApOCmZmVnBSMDOzgpOCmZkVnBTMzKzgpGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmZlZwUjAzs4KTgpmZFZwUzMys4KRgZmYFJwUzMys0LClIGiLpfkmTJT0p6aRcfqak6ZIm5Nd+pWVOkzRF0jOSPtKo2MzMrLb+DVz3QuDkiHhM0jrAeEl352nnRcT3yjNL2g44DHgPsClwj6RtImJRA2M0M7OShh0pRMSMiHgsD88DngIGdbHIKOD6iFgQEc8DU4CdGxWfmZktrSnnFCQNBXYEHs5FJ0qaKOlySRvkskHA1NJi06iRRCSNkdQhqWPOnDkNjNrMrO9peFKQtDbwU+ALEfE6cBGwFTAcmAGc25P1RcQlETEiIkYMGDBghcdrZtaXNTQpSFqVlBCujYhbACJiVkQsioi/A5eyuIloOjCktPjgXGZmZk3SyKuPBFwGPBUR3y+VDyzNdhAwKQ/fDhwmaXVJWwDDgEcaFZ+ZmS2tkVcfvR/4JPCEpAm57HTgcEnDgQBeAI4DiIgnJd0ITCZduXSCrzwyM2uuhiWFiBgHqMakO7tY5mzg7EbFZGZmXfMdzWZmVnBSMDOzgpOCmZkVnBTMzKzgpGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmZlZwUjAzs4KTgpmZFZwUzMys4KRgZmYFJwUzMys4KZiZWcFJwczMCk4KZmZWcFIwM7OCk4KZmRWcFMzMrOCkYGZmBScFMzMrNCwpSBoi6X5JkyU9KemkXL6hpLsl/TH/3SCXS9IFkqZImihpp0bFZmZmtTXySGEhcHJEbAfsCpwgaTvgVODeiBgG3JvHAfYFhuXXGOCiBsZmZmY1NCwpRMSMiHgsD88DngIGAaOAq/JsVwEH5uFRwNhIHgLWlzSwUfGZmdnSmnJOQdJQYEfgYWCTiJiRJ80ENsnDg4CppcWm5bLqdY2R1CGpY86cOQ2L2cysL6orKUi6t56yTpZdG/gp8IWIeL08LSICiHrWU1rmkogYEREjBgwY0JNFzcysG/27mihpDeAdwMb5hLDypHWp8Su+xvKrkhLCtRFxSy6eJWlgRMzIzUOzc/l0YEhp8cG5zMzMmqS7I4XjgPHAtvlv5XUbcGFXC0oScBnwVER8vzTpdmB0Hh6d11UpPzJfhbQr8FqpmcnMzJqgyyOFiPgB8ANJn4uIH/Zw3e8HPgk8IWlCLjsd+A5wo6RjgBeBQ/O0O4H9gCnAX4Gje7g9MzNbTl0mhYqI+KGk3YCh5WUiYmwXy4xjcXNTtT1rzB/ACfXEY2ZmjVFXUpB0NbAVMAFYlIsD6DQpmJlZ71NXUgBGANvlX/NmZraSqvc+hUnAuxoZiJmZtV69RwobA5MlPQIsqBRGxL82JCozM2uJepPCmY0MwszM2kO9Vx/9utGBmJlZ69V79dE8FndHsRqwKvBGRKzbqMDMzKz56j1SWKcynO9UHkXqDtvMzFYiPe4lNXdt/TPgIw2Ix8zMWqje5qODS6OrkO5beKshEZmZWcvUe/XRAaXhhcALpCYkMzNbidR7TsGd05mZ9QH1PmRnsKRbJc3Or59KGtzo4MzMrLnqPdF8Bel5B5vm189zmZmZrUTqTQoDIuKKiFiYX1cCfhammdlKpt6k8LKkT0jql1+fAF5uZGBmZtZ89SaFT5GekDYTmAEcAhzVoJjMzKxF6r0k9VvA6Ih4FUDShsD3SMnCzMxWEvUeKWxfSQgAEfEKsGNjQjIzs1apNymsImmDykg+Uqj3KMPMzHqJenfs5wK/l3RTHv8YcHZjQjIzs1ap60ghIsYCBwOz8uvgiLi6q2UkXZ5vdJtUKjtT0nRJE/Jrv9K00yRNkfSMJHe2Z2bWAnU3AUXEZGByD9Z9JXAhMLaq/LyI+F65QNJ2wGHAe0g3x90jaZuIWNSD7ZmZ2XLqcdfZ9YqI3wCv1Dn7KOD6iFgQEc8DU4CdGxWbmZnV1rCk0IUTJU3MzUuVk9eDgKmleablsqVIGiOpQ1LHnDlzGh2rmVmf0uykcBGwFTCcdBPcuT1dQURcEhEjImLEgAHuacPMbEVqalKIiFkRsSgi/g5cyuImounAkNKsg3OZmZk1UVOTgqSBpdGDgMqVSbcDh0laXdIWwDDgkWbGZmZmDbwBTdJ1wEhgY0nTgG8AIyUNB4L09LbjACLiSUk3kq5uWgic4CuPzMyar2FJISIOr1F8WRfzn41viDMza6lWXH1kZmZtyknBzMwKTgpmZlZwUjAzs4KTgpmZFZwUzMys4KRgZmYFJwUzMys4KZiZWcFJwczMCk4KZmZWcFIwM7OCk4KZmRWcFMzMrOCkYGZmBScFMzMrOCmYmVnBScHMzApOCmZmVnBSMDOzgpOCmZkVGpYUJF0uabakSaWyDSXdLemP+e8GuVySLpA0RdJESTs1Ki4zM+tcI48UrgT2qSo7Fbg3IoYB9+ZxgH2BYfk1BriogXGZmVknGpYUIuI3wCtVxaOAq/LwVcCBpfKxkTwErC9pYKNiMzOz2pp9TmGTiJiRh2cCm+ThQcDU0nzTctlSJI2R1CGpY86cOY2L1MysD2rZieaICCCWYblLImJERIwYMGBAAyIzM+u7mp0UZlWahfLf2bl8OjCkNN/gXGZmZk3U7KRwOzA6D48GbiuVH5mvQtoVeK3UzGRmZk3Sv1ErlnQdMBLYWNI04BvAd4AbJR0DvAgcmme/E9gPmAL8FTi6UXGZmVnnGpYUIuLwTibtWWPeAE5oVCxmZlYf39FsZmYFJwUzMys4KZiZWcFJwczMCg070byyOe/uZ5cY/+Le27QoEjOzxvGRgpmZFXykkPlIwMzMRwpmZlbiI4VOVB85mJn1BT5SMDOzgpOCmZkVnBTMzKzgpGBmZgUnBTMzKzgpmJlZwZekLiPf7GZmKyMfKZiZWcFJwczMCk4KZmZWcFIwM7OCTzQ3SPlEtE9Cm1lv0ZKkIOkFYB6wCFgYESMkbQjcAAwFXgAOjYhXWxGfmVlf1crmo3+JiOERMSKPnwrcGxHDgHvzuJmZNVE7nVMYBVyVh68CDmxhLGZmfVKrkkIAd0kaL2lMLtskImbk4ZnAJq0Jzcys72rViebdI2K6pHcCd0t6ujwxIkJS1FowJ5ExAJtttlnjIzUz60NacqQQEdPz39nArcDOwCxJAwHy39mdLHtJRIyIiBEDBgxoVshmZn1C05OCpLUkrVMZBj4MTAJuB0bn2UYDtzU7NjOzvq4VzUebALdKqmz/JxHxS0mPAjdKOgZ4ETi0BbGZmfVpTU8KEfEcsEON8peBPZsdT2/kHlrNrFHa6ZJUMzNrMXdzsYJU/3o3M+uNfKRgZmYFJwUzMyu4+agFetrU5BPJZtYsPlIwM7OCk4KZmRWcFMzMrOBzCrYE3xhn1rf5SMHMzAo+UmiCRt/Y1pPnQftIwMy64qTQC/QkqXinb2bLw0mhj3P3HGZW5qRgTdOTo5h2PuJp59jMlpdPNJuZWcFHCtaldjmJ7V/nZs3hpLCSa+Q5g75yPqKvvE8zcFIwW+F8VGO9mZOCtYx3nmbtx0nB2kYzm2m6Olfi5qLuNTOhL++2/OOjZ5wUzGy5dZdIvSPuuVYlMycFq1s7/4Lu7gvULrG38hd2d3rLjttHDo3VdklB0j7AD4B+wI8j4jstDsmWUW+58qmZfVMt77LtlET6op7++OiNCaetkoKkfsCPgL2BacCjkm6PiMmtjcx6m1adn7Dm60074p7c99MqiohWx1CQ9D7gzIj4SB4/DSAivl1r/hEjRkRHR8cybctfZLP20crmvvK2e9N+YXmSiqTxETGi1rS2OlIABgFTS+PTgF3KM0gaA4zJo/MlPbOM29oYeGkZl220do2tXeMCx7Ys2iauLy1d1LTYamy7KytLnW3e2YR2SwrdiohLgEuWdz2SOjrLlK3WrrG1a1zg2JZFu8YF7Rtbu8YFKy62dusQbzowpDQ+OJeZmVkTtFtSeBQYJmkLSasBhwG3tzgmM7M+o62ajyJioaQTgV+RLkm9PCKebNDmlrsJqoHaNbZ2jQsc27Jo17igfWNr17hgBcXWVlcfmZlZa7Vb85GZmbWQk4KZmRX6ZFKQtI+kZyRNkXRqi2N5QdITkiZI6shlG0q6W9If898NmhTL5ZJmS5pUKqsZi5ILch1OlLRTC2I7U9L0XHcTJO1XmnZajlrzb4AAAAg+SURBVO0ZSR9pYFxDJN0vabKkJyWdlMtbWm9dxNUOdbaGpEckPZ5j+2Yu30LSwzmGG/LFJkhaPY9PydOHNjmuKyU9X6qz4bm8qd+BvM1+kv4g6Y48vuLrLCL61It0AvtPwJbAasDjwHYtjOcFYOOqsv8GTs3DpwLnNCmWDwI7AZO6iwXYD/gFIGBX4OEWxHYm8OUa826XP9fVgS3y592vQXENBHbKw+sAz+btt7TeuoirHepMwNp5eFXg4VwXNwKH5fKLgePz8GeBi/PwYcANTY7rSuCQGvM39TuQt/kl4CfAHXl8hddZXzxS2BmYEhHPRcTfgOuBUS2Oqdoo4Ko8fBVwYDM2GhG/AV6pM5ZRwNhIHgLWlzSwybF1ZhRwfUQsiIjngSmkz70Rcc2IiMfy8DzgKdKd+S2tty7i6kwz6ywiYn4eXTW/AtgDuDmXV9dZpS5vBvaUpCbG1ZmmfgckDQb2B36cx0UD6qwvJoVaXWl09WVptADukjReqQsPgE0iYkYengls0prQuoylXerxxHzofnmpma0lseVD9B1JvzDbpt6q4oI2qLPcDDIBmA3cTToymRsRC2tsv4gtT38N2KgZcUVEpc7OznV2nqTVq+OqEXMjnA98Bfh7Ht+IBtRZX0wK7Wb3iNgJ2Bc4QdIHyxMjHf+1xXXD7RRLdhGwFTAcmAGc26pAJK0N/BT4QkS8Xp7WynqrEVdb1FlELIqI4aReC3YGtm1FHNWq45L0XuA0Unz/DGwIfLXZcUn6KDA7IsY3elt9MSm0VVcaETE9/50N3Er6gsyqHIbmv7NbFV8XsbS8HiNiVv4S/x24lMXNHU2NTdKqpB3vtRFxSy5ueb3Viqtd6qwiIuYC9wPvIzW/VG6oLW+/iC1PXw94uUlx7ZOb4iIiFgBX0Jo6ez/wr5JeIDV570F67swKr7O+mBTapisNSWtJWqcyDHwYmJTjGZ1nGw3c1or4ss5iuR04Ml+BsSvwWqm5pCmq2m8PItVdJbbD8hUYWwDDgEcaFIOAy4CnIuL7pUktrbfO4mqTOhsgaf08vCbp+SlPkXbCh+TZquusUpeHAPflo69mxPV0KbmL1GZfrrOmfAci4rSIGBwRQ0n7rPsi4ggaUWeNOkvezi/SVQPPktoxz2hhHFuSrvh4HHiyEgup7e9e4I/APcCGTYrnOlKTwtuk9sljOouFdMXFj3IdPgGMaEFsV+dtT8xfgoGl+c/IsT0D7NvAuHYnNQ1NBCbk136trrcu4mqHOtse+EOOYRLw9dL34RHSSe6bgNVz+Rp5fEqevmWT47ov19kk4BoWX6HU1O9AKc6RLL76aIXXmbu5MDOzQl9sPjIzs044KZiZWcFJwczMCk4KZmZWcFIwM7OCk4K1jKSNSj1PztSSvXeu1sK4TpB0RDfz7JGvTV9R2/xV5Z6VZpC0de7OwWwJbfU4TutbIuJlUncLSDoTmB8R3yvPk28YUqQ7cBtOUv+I+FEds+4BvAQ8tCK2GxEN66rarCd8pGBtJ/+KnSzpWtJNfQMlXSKpQ6mf+6+X5p2m9IyAP+QOy7bJ5Xso9Ys/QdJj+Y5xJJ2u9PyKxyWdncvG5Y7OOkidxZ0l6Qulaefn9TwhaYSkrYBjgVNy+W75Dvn7cwx3K/VoiaRrJP1IqU/7P0n6oKSrJD0t6bKq91G5m/bovJ7HJV2Ryw6TNCmX3V+jzm5W6RkIebsHStpK0m9z/YyXtEuNZY+VdH5p/JeSds/D+0r6fa7DGyr1aCsvHylYu9oWODIiKg8eOjUiXlHqx+V+STdHxOQ876yI2FHS50n9zX8GOAUYExEPK3UK95akA0gdD+4cEW9K2rC0vX4RMSJv66yqWFaPiOGS9gB+nId/DLwUEefnZX6Rp12r1Nvt+SzufmC9iNhF0r8BPyf18/M08Jik90ZE+cFBO5A6XNstv99KjN8ARkbErEryqHIDcCjwK0lrAB8i3fXdD9g7It6StC2pO+WlEkMtkt5JehbEnhHxV0lnACcB/1XP8tY7+UjB2tWfKgkhO1zSY8BjwD+QHgpTUemEbjwwNA//DviBpM8B60bEImAv4PKIeBMgIsrPZ7ihi1iuy/PfB7wzJ5lqu5A6KgMYC3ygNO3n+e8TwF8iYnJuDptcirdiD9IDUV6pivF3wFhJx1L7e/t/wN5KneDtT+rrZgHpoTmXKT2x7nqWrLfu7JbnfzCffziiRry2kvGRgrWrNyoDkoaRfqHuHBFzJV1D6tulYkH+u4j8Px0RZ0m6nbSDfEjSnvVur4bqvmB62jdMJb6/l4Yr4/V+Bz9NSjwfJR1h7BgRrxYBpV/y40iduH2c9LQwgJNJ/ep/gvTQmPksbSFLJppK3Qr4ZUR8ss4YbSXgIwXrDdYF5gGvK/VY2e1JWUlbRcTEiPg26eji3aSHuXxKqQdMqpqPuvLxPP9IUlPVGzme8tVCD5GabyDtgH9T57qr3Qd8vBJbKcYtIz3d62vAq9R+mMsNpCaj95HeK6Quk2dE6uRsNGlHX+0FYEclQ4F/yuUPAh+StGWOZa2coG0l5iMF6w0eIzW1PA28SGpK6c6XJX2A9Gt8InBXRPwtt9l3SHqb1KzztTrW9XZuPukHHJ3LbgNuknQwcEJ+XS7pNGBWab4eiYjHJf038BtJC0lNYscA5yl1aa38XibVWPyXpHMGN0XE27nsQuBmSZ8iNTEtqLHcr0n97z9FOrE/IccyS9IxQPFAeOB0Us+vtpJyL6lmXchNMidGhK/ptz7BzUdmZlbwkYKZmRV8pGBmZgUnBTMzKzgpmJlZwUnBzMwKTgpmZlb4fxIYQqwEmoTzAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>transcipt_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12DGR120tipp</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12DGR140tipp</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12DGR141tipp</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reaction_id  transcipt_level\n",
       "0  12DGR120tipp              inf\n",
       "1  12DGR140tipp              inf\n",
       "2  12DGR141tipp              inf"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Without constraints, flux balance analysis found that 8.402 grams of biomass could be made.\n",
      "\n",
      "{'EX_h2o_e': inf, 'EX_h_e': inf, 'EX_cl_e': inf, 'EX_pi_e': inf, 'EX_nh4_e': inf, 'EX_fe3_e': inf, 'EX_k_e': inf, 'EX_ca2_e': inf, 'EX_mg2_e': inf, 'EX_mn2_e': inf, 'EX_cobalt2_e': inf, 'EX_zn2_e': inf, 'EX_cu2_e': inf, 'EX_o2_e': inf, 'EX_fe2_e': inf, 'EX_mobd_e': inf, 'EX_so4_e': inf, 'EX_phenol_e': 10.0}\n",
      "\n",
      "Given the bounds from transcriptmics and a minimum absolute bound of 30, flux balance analysis found that 1576.369 grams of biomass could be made.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sol_WTPR1 = EFlux2test(model, trans_df, 'WT-P-R1', min_bound = 30)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20978510-4343-48aa-bd76-cae4e38a8f01",
   "metadata": {},
   "source": [
    "Unused code used for reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c70e3fb-7997-4935-b6af-7a18e443a7fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def rxn_to_GPR(model, rxn):\n",
    "#     parallel_gene_sets = set()\n",
    "#     if rxn.gene_reaction_rule:\n",
    "#         for parallel_gene_string in rxn.gene_reaction_rule.split(' or '):\n",
    "#             parallel_gene_string = parallel_gene_string.strip('() ')\n",
    "#             parallel_gene_list = []\n",
    "            \n",
    "#             for gene_name in parallel_gene_string.split(' and '):\n",
    "#                 gene_name = gene_name.strip('() ')\n",
    "#                 parallel_gene_list.append(gene_name)\n",
    "            \n",
    "#             parallel_gene_sets.add(frozenset(parallel_gene_list))\n",
    "            \n",
    "#     return parallel_gene_sets\n",
    "            \n",
    "# Orginal code\n",
    "#     if r.gene_reaction_rule:\n",
    "#         for x in [x.strip('() ') for x in r.gene_reaction_rule.split(' or ')]:\n",
    "#             temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "#         return temp\n",
    "#     else:\n",
    "#         return {}\n",
    "    \n",
    "# Christina's version\n",
    "#     if r.gene_reaction_rule:\n",
    "#         t2 = 0\n",
    "#         #If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "#         #If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "#         #If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "#         for p in gpr_dict[r.id]:\n",
    "#             xg = []\n",
    "#             for g in p:\n",
    "#                 if g in transcriptomics_dict.keys():\n",
    "#                     print(transcriptomics_dict.get(g))\n",
    "#                     xg.append(transcriptomics_dict.get(g))\n",
    "#                 else:\n",
    "#                     xg.append(np.Inf)\n",
    "#                 t1 = np.min(xg)\n",
    "#             t2 = t2 + t1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61a35c08-977d-4a61-acb5-4868b5e22b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# _ = [print(rxn, rxn.upper_bound) for rxn in model.reactions if rxn.upper_bound != 1000] \n",
    "# print()\n",
    "# _ = [print(rxn, rxn.lower_bound) for rxn in model.reactions if ( rxn.lower_bound != -1000 and rxn.lower_bound != 0)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3a84421-52c3-4620-95c9-dea6ed91b142",
   "metadata": {},
   "outputs": [],
   "source": [
    "eflux2sol_WTPR1_df = pd.DataFrame(sol_WTPR1.fluxes, columns=['fluxes'])\n",
    "eflux2sol_WTPR1_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b93e9bf6-7128-4a3d-bb1e-a45fd3ce49e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6d92333-13e8-43a4-bf5c-4fb31fd006c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "99fefd29-1e17-4bd1-92f3-b06e4d0b1d1c",
   "metadata": {},
   "source": [
    "##### Just linear scale up to solution to see how it compares:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a49529-2838-471e-9f6c-206ddaa49cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "testscaleeflux2 = eflux2sol_WTPR1_df*(100/1.76124)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "999610be-d11c-4525-9c73-c93059ae8622",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(testscaleeflux2.index,testscaleeflux2.fluxes)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c23b85-c646-4b05-8297-7b085fac11d9",
   "metadata": {},
   "source": [
    "Old efflux code saved as reference material"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "171abbc4-d08e-4695-81db-8f43c51d9562",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def EFlux2test(model, edd_df, line_name, min_bound = 10):\n",
    "    # covert transcriptomic df to dictionary of genes and transcript values\n",
    "#     transcriptomics_dict = {row['Measurement Type']:row['Value'] for _, row in trans_df.iterrows()}\n",
    "    \n",
    "#     # plot histogram of transcriptomic measurements\n",
    "#     plot_histogram(transcriptomics_dict.values(), 'Transcriptomic values')\n",
    "    \n",
    "#     # convert model into dictionary of genes and their associated GPRs\n",
    "#     gpr_dict = model_to_GPR_dictionary(model)\n",
    "    \n",
    "#     with model:        \n",
    "#         # Ensure that medium is correct, and the carbon source is the only limiting substrate\n",
    "#         medium = {key:np.Inf for (key,value) in model.medium.items()}\n",
    "#         medium[\"EX_glc__D_e\"] = 0.0\n",
    "#         medium[\"EX_phenol_e\"] = 1.76124#100.0\n",
    "#         model.medium = medium\n",
    "#         print(model.medium) # print the media to prevent inivisible errors\n",
    "        \n",
    "#         # calculate maximum biomass yield without constraints\n",
    "#         model.tolerance = 1e-9\n",
    "#         unbounded_sol = model.optimize()\n",
    "#         print(f'\\nWithout constraints, flux balance analysis found that {unbounded_sol.objective_value:.3f} grams of biomass could be made.\\n')\n",
    "\n",
    "#         # calculate maximum biomass yield without constraints\n",
    "#         bounds = []\n",
    "#         bounded_model = model.copy()\n",
    "#         with bounded_model:\n",
    "#             for rxn_id in list(gpr_dict.keys()):\n",
    "#                 #exclude phenol related rxns from transcript bounds for now:\n",
    "# #                 if bounded_model.reactions.get_by_id(rxn_id).upper_bound != 1000:\n",
    "# #                     display('Phenol related rxns excluded from transcript bounds',rxn_id)\n",
    "# #                     continue\n",
    "#                 bound = 0\n",
    "#                 for genes_in_series in gpr_dict[rxn_id]:\n",
    "#                     bound += min([transcriptomics_dict[gene] if gene in transcriptomics_dict.keys() else 0 for gene in genes_in_series ])\n",
    "\n",
    "#                 bound = max(bound, min_bound) # min_bound is a parameter of this function\n",
    "#                 bounds.append(bound)\n",
    "\n",
    "#                 if bounded_model.reactions.get_by_id(rxn_id).upper_bound > 0:\n",
    "#                     bounded_model.reactions.get_by_id(rxn_id).upper_bound = bound\n",
    "\n",
    "#                 elif bounded_model.reactions.get_by_id(rxn_id).lower_bound < 0:\n",
    "#                     bounded_model.reactions.get_by_id(rxn_id).lower_bound = -1*bound\n",
    "                    \n",
    "#         #Alternative to external function:\n",
    "#         # I think this is more readable than the nested loops. \n",
    "#         #         xg = []\n",
    "#         #         t2 = 0\n",
    "#         #         for r in model.reactions:\n",
    "#         #             if r.gene_reaction_rule:\n",
    "#         #                 t2 = 0\n",
    "#         #                 #If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "#         #                 #If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "#         #                 #If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "#         #                 for p in gpr_dict[r.id]:\n",
    "#         #                     xg = []\n",
    "#         #                     for g in p:\n",
    "#         #                         if g in transcriptomics_dict.keys():\n",
    "#         #                             print(transcriptomics_dict.get(g))\n",
    "#         #                             xg.append(transcriptomics_dict.get(g))\n",
    "#         #                         else:\n",
    "#         #                             xg.append(np.Inf)\n",
    "#         #                         t1 = np.min(xg)\n",
    "#         #                     t2 = t2 + t1\n",
    "\n",
    "#             # plot histogram of absolute value of bounds inforced\n",
    "#             plot_histogram(bounds, 'Absolute value of bounds')\n",
    "\n",
    "#             bounded_sol = bounded_model.optimize()\n",
    "#             print(f'\\nGiven the bounds from transcriptmics and a minimum absolute bound of {min_bound}, flux balance analysis found that {bounded_sol.objective_value:.3f} grams of biomass could be made.\\n')\n",
    "\n",
    "#         # Constrain the biomass to the optimal value\n",
    "#             print('FBA status', bounded_sol.status)\n",
    "#             print('FBA solution', bounded_sol.objective_value)\n",
    "#             display(bounded_model.objective)\n",
    "#             # Constrain the biomass to the optimal value\n",
    "#             for rxn_id in bounded_model.reactions:\n",
    "#                 if rxn_id.objective_coefficient:\n",
    "#                     rxn_id.lower_bound = bounded_sol.objective_value\n",
    "#             # minimize the sum of squared flux values\n",
    "#             bounded_model.objective = bounded_model.problem.Objective(add([rxn_id.flux_expression**2 for rxn_id in bounded_model.reactions]), direction='min')\n",
    "#             display(bounded_model.problem.Objective)\n",
    "#             eflux2_sol = bounded_model.optimize()\n",
    "#             print('EFlux2 status', eflux2_sol.status)\n",
    "#             print('EFlux2 solution', eflux2_sol.objective_value)\n",
    "#     return eflux2_sol"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4f678b3-8c1c-4444-9e30-97c812e5fe7c",
   "metadata": {},
   "source": [
    "#### **Plot solutions: Comparison of SPOT and EFLUX2 solution**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4f7d0ac-a223-4e43-a666-f75507694c95",
   "metadata": {},
   "source": [
    "##### **Load 13cdata**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bbd32f4-f440-47d7-bf5f-f669e93204ba",
   "metadata": {},
   "source": [
    "####  **Get 13C MFA measured fluxes for phenol**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae1c076f-74cd-44e5-82a9-adeb0f9c0726",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenol_fluxes = pd.read_csv('../13C_flux_data/13C_phenol_flux_data.csv')\n",
    "phenol_fluxes.dropna(inplace=True)\n",
    "print(f'There are {len(phenol_fluxes)} fluxes that can be compared between the MFA and FBA')\n",
    "phenol_fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9408c4ea-55a7-4859-838a-4ada1879d93f",
   "metadata": {},
   "source": [
    "### Add Phenol EFLUX2 and SPOT flux values to phenol fluxes dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5864b3e7-3e48-4d32-b096-4d88a74f7c4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_flux_value(reaction_id, solution):\n",
    "    if reaction_id.startswith('reverse_'):\n",
    "        reaction_id = reaction_id.split('reverse_')[1]\n",
    "        return -1*solution.fluxes[reaction_id]\n",
    "    else:\n",
    "        return solution.fluxes[reaction_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0f84d6-4bfe-400b-8071-da8337ac9b95",
   "metadata": {},
   "outputs": [],
   "source": [
    "EFLUX2_fluxes = []\n",
    "#SPOT_fluxes = []\n",
    "for _, row in phenol_fluxes.iterrows():\n",
    "    reactions = row['Forward Reactions']\n",
    "    flux_value_eflux2 = 0\n",
    "    #flux_value_spot = 0\n",
    "    for x in [x.strip('() ') for x in reactions.split(' or ')]:\n",
    "        and_split = [y.strip('() ') for y in x.split(' and ')]\n",
    "        flux_value_eflux2 += min([get_flux_value(v, testscaleeflux2) for v in and_split])\n",
    "        #flux_value_spot += min([get_flux_value(v,spotsoldf) for v in and_split])\n",
    "    EFLUX2_fluxes.append(flux_value_eflux2)\n",
    "    #SPOT_fluxes.append(flux_value_spot)\n",
    "    \n",
    "phenol_fluxes['EFLUX2 Value'] = EFLUX2_fluxes\n",
    "#phenol_fluxes['SPOT Value'] = SPOT_fluxes\n",
    "phenol_fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c48b766-9871-40fd-8b05-afba0f5877b9",
   "metadata": {},
   "source": [
    "### Plot EFLUX2 vs 13C MFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bb88ab1-a3c2-4d9c-a132-238fef787cba",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.plot(ax.get_xlim(), ax.get_ylim(), ls=\"--\", c=\".3\")\n",
    "\n",
    "for pathway in list(set(phenol_fluxes['Pathway'])):\n",
    "    pathway_df = phenol_fluxes[phenol_fluxes['Pathway'] == pathway]\n",
    "    \n",
    "    measured_flux_list = list(pathway_df['Flux'])\n",
    "    simulated_flux_list = list(pathway_df['EFLUX2 Value'])\n",
    "    \n",
    "    ax.scatter(measured_flux_list, simulated_flux_list, label=pathway)\n",
    "\n",
    "\n",
    "# Dashed line\n",
    "x = np.linspace(*ax.get_xlim())\n",
    "ax.plot(x, x, ls=\"--\", c=\".3\")\n",
    "\n",
    "plt.title(r'Phenol 13C MFA vs. EFLUX2 Fluxes', fontSize=20)\n",
    "plt.xlabel(r'13C MFA flux (per 100 mmol of phenol uptake)')\n",
    "plt.ylabel(r'EFLUX2 flux (per 100 mmol of phenol uptake)')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1970487-c86c-450c-94c9-07d133446f92",
   "metadata": {},
   "source": [
    "-- Stopping here --"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e2eebd-f0a6-4fd9-afb9-d0ead616d55c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "565aadd8-ae1c-48fc-a175-d4476368b31e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17a82a57-e872-4726-88e9-b52500318e0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3510a539-0186-4bb5-8cf7-f94d9b9cb6e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f5b67a1-3478-4b65-b7bd-ec3a9e8b4c97",
   "metadata": {},
   "outputs": [],
   "source": [
    "from optlang.symbolics import add\n",
    "\"\"\"\n",
    "    Provides EFLUX2 predictions as explained in Machado et. al (2014) \n",
    "    \n",
    "        Parameters\n",
    "        ----------\n",
    "        model : cobrapy model.\n",
    "        Transcriptomics : pandas dataframe with transcriptomics data.\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        eflux2_sol as output from eflux2_model.optimize().\n",
    "        \n",
    "\"\"\"\n",
    "#Code only works for GPRs written in disjunctive normal form (DNF). Majority of models have them in DNF but there are some exceptions. \n",
    "\n",
    "def EFlux2test(model, df_trans):\n",
    "    # Convert EDD style dataframe to dictionary\n",
    "    transcriptomics = {row['Measurement Type']:row['Value'] for _, row in df_trans.iterrows()}\n",
    "    print(list(transcriptomics.keys())[:5])\n",
    "    with model:\n",
    "        \n",
    "        # Parse GPR into a dict containing isozymes (separated by 'or')\n",
    "        # Each isozyme has a set of subunits (separated by 'and')\n",
    "        #'and' and 'or' can occur at the same time, or can occur by itself.\n",
    "        gpr_dict = dict()\n",
    "        for r in model.reactions:\n",
    "            if r.gene_reaction_rule:\n",
    "                temp = set()\n",
    "                for x in [x.strip('() ') for x in r.gene_reaction_rule.split(' or ')]:\n",
    "                    temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "                gpr_dict[r.id] = temp\n",
    "                \n",
    "        # define a variable to hold all bounds\n",
    "        all_bounds = []\n",
    "                \n",
    "        # Set the bounds using the transcriptomics data\n",
    "        for r in model.reactions:\n",
    "            if r.gene_reaction_rule:\n",
    "                \n",
    "                #If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "                #If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "                #If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "                \n",
    "#                 t = np.sum([np.min([Transcriptomics.loc[g] if g in Transcriptomics.index\n",
    "#                                     else np.array([np.Inf]) for g in p])\n",
    "#                             for p in gpr_dict[r.id]])\n",
    "                t = np.sum([np.min([transcriptomics[g] if g in transcriptomics.keys()\n",
    "                                    else np.array([np.Inf]) for g in p])\n",
    "                            for p in gpr_dict[r.id]])\n",
    "                \n",
    "                # add bound to the list\n",
    "                if t != np.Inf:\n",
    "                    all_bounds.append(t)\n",
    "                \n",
    "                # Enforce transcriptomic bound\n",
    "                if r.lower_bound < 0.0:\n",
    "                    r.lower_bound = -t\n",
    "                if r.upper_bound > 0.0:\n",
    "                    r.upper_bound = t\n",
    "            else:\n",
    "                #When there is no GPR, the arbitrary bounds are removed. \n",
    "                #Common arbitrary bound value of 1000 for E.coli, might be different depending on the model, e.g., 99999.0 for iMM904 yeast model in BiGG\n",
    "                if r.lower_bound <= -1000.0:\n",
    "                    r.lower_bound = -np.Inf\n",
    "                if r.upper_bound >= 1000.0:\n",
    "                    r.upper_bound = np.Inf\n",
    "              \n",
    "        # These statements can help with scaling bounds to reasonable limits\n",
    "        print(all_bounds[:20])\n",
    "        print(len(all_bounds))\n",
    "        print(f'The average bound was {sum(all_bounds) / len(all_bounds)}')\n",
    "        print(f'The max bound was {max(all_bounds)}')\n",
    "        \n",
    "        # print the media to prevent inivisible errors\n",
    "        display(model.medium)\n",
    "        \n",
    "        # solve FBA with transcriptomic enforced bounds to calculate the maximum biomass\n",
    "        model.tolerance = 1e-9\n",
    "        fba_sol = model.optimize()\n",
    "        \n",
    "        print(f'The FBA solution was {fba_sol.status}.' )\n",
    "        print(f'Given the bounds from transcriptomics, flux balance analysis found that{fba_sol.objective_value:.3f} grams of biomass could be made.\\n')\n",
    "        \n",
    "        # Constrain the biomass to the optimal value\n",
    "        for r in model.reactions:\n",
    "            if r.objective_coefficient:\n",
    "                r.lower_bound = fba_sol.objective_value\n",
    "                \n",
    "        # minimize the sum of squared flux values\n",
    "        model.objective = model.problem.Objective(add([r.flux_expression**2 for r in model.reactions]), direction='min')\n",
    "        eflux2_sol = model.optimize()\n",
    "        \n",
    "        print(f'The EFLUX2 solution was {eflux2_sol.status}.' )\n",
    "        print(f'The minimum sum of squared fluxes to make {fba_sol.objective_value:.3f} grams of biomass given the bounds from transcriptomics is {eflux2_sol.objective_value:.1f}.\\n')\n",
    "\n",
    "    return eflux2_sol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dbde7cf-10d7-4adb-ba13-230142c780ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91390bb8-99d6-4895-a78a-2f79c9806f10",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcc4dc6c-620b-4c1b-89bc-b657aa0ed3aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eflux2sol_WTPR1, spotsol_WTPR1 = eflux2spotpred(model, df_Trans2, 'WT-P-R1', 'phenol', 100)\n",
    "# eflux2sol_WTPR2, spotsol_WTPR2 = eflux2spotpred(model, df_Trans2, 'WT-P-R2', 'phenol', 100)\n",
    "# eflux2sol_WTPR3, spotsol_WTPR3 = eflux2spotpred(model, df_Trans2, 'WT-P-R3', 'phenol', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edab2d08-2018-4ae7-a59f-f09441feba72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eflux2sol_WTPR1.fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d8d0611-3d4c-4027-9c3d-d527dc6c001d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eflux2sol_WTPR1_df = pd.DataFrame(eflux2sol_WTPR1.fluxes, columns=['fluxes'])\n",
    "# eflux2sol_WTPR2_df = pd.DataFrame(eflux2sol_WTPR2.fluxes, columns=['fluxes'])\n",
    "# eflux2sol_WTPR3_df = pd.DataFrame(eflux2sol_WTPR3.fluxes, columns=['fluxes'])\n",
    "# eflux2sol_WTPR = pd.concat([eflux2sol_WTPR1_df, eflux2sol_WTPR2_df, eflux2sol_WTPR3_df], axis=1)\n",
    "# eflux2sol_WTPR.rename(columns = {'fluxes':'R1', 'fluxes':'R2','fluxes':'R3'}, inplace=False)\n",
    "# eflux2sol_WTPR.mean(axis=1)\n",
    "# eflux2sol_WTPR.std(axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb8788d4-faa3-4075-9926-4559bfe9e5cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eflux2sol_WTPR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abb2c28f-530e-415e-ab92-03fdc08091f5",
   "metadata": {},
   "source": [
    "##### **Modularized way**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e44e04e9-5d89-4c5d-a80c-ca5f60b6ffa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function for EFLUX2 and SPOT Predictions:\n",
    "def eflux2spotpred(model, transcript_df, line_name, substrate, sub_uptake_rate):    \n",
    "    with model:\n",
    "        \n",
    "        # This is to prevent medium components other than glucose from being growth limitations\n",
    "        medium = model.medium\n",
    "        medium = {key:1000 for (key,value) in model.medium.items()}\n",
    "    \n",
    "        if substrate == 'phenol':\n",
    "            model.objective = 'Growth_Phenol'\n",
    "            model.reactions.get_by_id('Growth_Glucose').upper_bound = 0\n",
    "            model.reactions.get_by_id('Growth_Glucose').lower_bound = 0\n",
    "            medium[\"EX_glc__D_e\"] = 0.0\n",
    "            medium[\"EX_phenol_e\"] = sub_uptake_rate\n",
    "            \n",
    "        elif substrate == 'glucose':\n",
    "            model.objective = 'Growth_Glucose'\n",
    "            model.reactions.get_by_id('Growth_Phenol').upper_bound = 0\n",
    "            model.reactions.get_by_id('Growth_Phenol').lower_bound = 0\n",
    "            medium[\"EX_glc__D_e\"] = sub_uptake_rate\n",
    "            medium[\"EX_phenol_e\"] = 0.0\n",
    "            \n",
    "        else:\n",
    "            print('Unknown substrate: Please choose among phenol and glucose')\n",
    "            \n",
    "        model.medium = medium\n",
    "        \n",
    "        eflux2sol = EFlux2test(model, transcript_df[transcript_df['Line Name'] == line_name])\n",
    "        spotsol = SPOT(model, transcript_df[transcript_df['Line Name'] == line_name])\n",
    "    return eflux2sol, spotsol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cfa2911-aa1d-4f50-9da0-3ac46efa1883",
   "metadata": {},
   "outputs": [],
   "source": [
    "eflux2sol_WTPR1, spotsol_WTPR1 = eflux2spotpred(model, df_trans, 'WT-P-R1', 'phenol', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "391ad0d3-c426-4b20-97ac-13d3108ceada",
   "metadata": {},
   "outputs": [],
   "source": [
    "#display(eflux2sol_WTPR1.fluxes['Growth_Phenol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c6c4753-aba5-460e-a3f9-a4491b686e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#spotsol_WTPR1df =pd.DataFrame(spotsol_WTPR1, columns=['fluxes'])\n",
    "#spotsol_WTPR1df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "212e000a-362e-4ae7-87da-ca33b132d891",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function for predictions for three replicates and averaging the solutions and calculating the standard deviation:\n",
    "def eflux2spotpredthreereps(model, transcript_df, line_name1, line_name2, line_name3, substrate, sub_uptake_rate):\n",
    "    \n",
    "    #call prediction functions for individual EFLUX2 predictions for all 3 replicates:\n",
    "    eflux2sol_R1, spotsol_R1 = eflux2spotpred(model, transcript_df, line_name1, substrate, sub_uptake_rate)\n",
    "    eflux2sol_R2, spotsol_R2 = eflux2spotpred(model, transcript_df, line_name2, substrate, sub_uptake_rate)\n",
    "    eflux2sol_R3, spotsol_R3 = eflux2spotpred(model, transcript_df, line_name3, substrate, sub_uptake_rate)\n",
    "    \n",
    "    #EFLUX2 calculations:\n",
    "    eflux2sol_R1_df = pd.DataFrame(eflux2sol_R1.fluxes, columns=['fluxes'])\n",
    "    eflux2sol_R2_df = pd.DataFrame(eflux2sol_R2.fluxes, columns=['fluxes'])\n",
    "    eflux2sol_R3_df = pd.DataFrame(eflux2sol_R3.fluxes, columns=['fluxes'])\n",
    "    eflux2sol_all = pd.concat([eflux2sol_R1_df, eflux2sol_R2_df, eflux2sol_R3_df], axis=1)\n",
    "\n",
    "    eflux2sol = pd.DataFrame(eflux2sol_all.mean(axis=1), columns=['fluxes'])\n",
    "    eflux2sol_std = eflux2sol_all.std(axis=1)\n",
    "    \n",
    "    #SPOT calculations:\n",
    "    spotsol_R1_df = pd.DataFrame(spotsol_R1, columns=['fluxes'])\n",
    "    spotsol_R2_df = pd.DataFrame(spotsol_R2, columns=['fluxes'])\n",
    "    spotsol_R3_df = pd.DataFrame(spotsol_R3, columns=['fluxes'])\n",
    "    spotsol_all = pd.concat([spotsol_R1_df, spotsol_R2_df, spotsol_R3_df], axis=1)\n",
    "    \n",
    "    spotsol = spotsol_all.mean(axis=1)\n",
    "    spotsol_std = spotsol_all.std(axis=1)\n",
    "    \n",
    "    return eflux2sol, eflux2sol_std, spotsol, spotsol_std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fbb0754-d6a2-4a38-b1da-bc449a52816d",
   "metadata": {},
   "outputs": [],
   "source": [
    "eflux2sol, eflux2sol_std, spotsol, spotsol_std = eflux2spotpredthreereps(model, df_trans, 'WT-P-R1', 'WT-P-R2','WT-P-R3', 'phenol', 100)\n",
    "#eflux2sol, eflux2sol_std = eflux2spotpredthreereps(model, df_Trans2, 'WT-P-R1', 'WT-P-R2','WT-P-R3', 'phenol', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "267b9c37-7c79-475a-98dc-f529bfe19658",
   "metadata": {},
   "outputs": [],
   "source": [
    "eflux2soldf = pd.DataFrame(eflux2sol, columns=['fluxes'])\n",
    "eflux2solstddf = pd.DataFrame(eflux2sol_std, columns=['fluxes'])\n",
    "spotsoldf = pd.DataFrame(spotsol, columns=['fluxes'])\n",
    "spotsolstddf = pd.DataFrame(spotsol_std, columns=['fluxes'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9759f638-53ff-4067-a6ff-09cbf08b90c6",
   "metadata": {},
   "source": [
    "#### **Plot solutions: Comparison of SPOT and EFLUX2 solution**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0782d2e4-632a-4d5b-ad87-0643d4611df7",
   "metadata": {},
   "source": [
    "##### **Load 13cdata**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9c82e93-eb7d-4b72-91c6-4a27927b2fcd",
   "metadata": {},
   "source": [
    "####  **Get 13C MFA measured fluxes for phenol**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7807b73-d258-4fec-9c26-652535e22262",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenol_fluxes = pd.read_csv('../13C_flux_data/13C_phenol_flux_data.csv')\n",
    "phenol_fluxes.dropna(inplace=True)\n",
    "print(f'There are {len(phenol_fluxes)} fluxes that can be compared between the MFA and FBA')\n",
    "phenol_fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7992430a-5aee-4f12-9727-236620de1f0f",
   "metadata": {},
   "source": [
    "### Add Phenol EFLUX2 and SPOT flux values to phenol fluxes dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95187395-9c83-44d3-b7b8-b14ae4e31142",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_flux_value(reaction_id, solution):\n",
    "    if reaction_id.startswith('reverse_'):\n",
    "        reaction_id = reaction_id.split('reverse_')[1]\n",
    "        return -1*solution.fluxes[reaction_id]\n",
    "    else:\n",
    "        return solution.fluxes[reaction_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c49f728-3eb0-4cd9-aa59-84a593543913",
   "metadata": {},
   "outputs": [],
   "source": [
    "EFLUX2_fluxes = []\n",
    "SPOT_fluxes = []\n",
    "for _, row in phenol_fluxes.iterrows():\n",
    "    reactions = row['Forward Reactions']\n",
    "    flux_value_eflux2 = 0\n",
    "    flux_value_spot = 0\n",
    "    for x in [x.strip('() ') for x in reactions.split(' or ')]:\n",
    "        and_split = [y.strip('() ') for y in x.split(' and ')]\n",
    "        flux_value_eflux2 += min([get_flux_value(v, eflux2soldf) for v in and_split])\n",
    "        flux_value_spot += min([get_flux_value(v,spotsoldf) for v in and_split])\n",
    "    EFLUX2_fluxes.append(flux_value_eflux2)\n",
    "    SPOT_fluxes.append(flux_value_spot)\n",
    "    \n",
    "phenol_fluxes['EFLUX2 Value'] = EFLUX2_fluxes\n",
    "phenol_fluxes['SPOT Value'] = SPOT_fluxes\n",
    "phenol_fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e59f7309-6a75-43af-9490-dd865315e675",
   "metadata": {},
   "source": [
    "### Plot EFLUX2 vs 13C MFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "259d24b2-2bfd-46d3-ac7f-5461ee13a559",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.plot(ax.get_xlim(), ax.get_ylim(), ls=\"--\", c=\".3\")\n",
    "\n",
    "for pathway in list(set(phenol_fluxes['Pathway'])):\n",
    "    pathway_df = phenol_fluxes[phenol_fluxes['Pathway'] == pathway]\n",
    "    \n",
    "    measured_flux_list = list(pathway_df['Flux'])\n",
    "    simulated_flux_list = list(pathway_df['EFLUX2 Value'])\n",
    "    \n",
    "    ax.scatter(measured_flux_list, simulated_flux_list, label=pathway)\n",
    "\n",
    "\n",
    "# Dashed line\n",
    "x = np.linspace(*ax.get_xlim())\n",
    "ax.plot(x, x, ls=\"--\", c=\".3\")\n",
    "\n",
    "plt.title(r'Phenol 13C MFA vs. EFLUX2 Fluxes', fontSize=20)\n",
    "plt.xlabel(r'13C MFA flux (per 100 mmol of phenol uptake)')\n",
    "plt.ylabel(r'EFLUX2 flux (per 100 mmol of phenol uptake)')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01734cb6-69c9-4952-81e3-13c754d2bd38",
   "metadata": {},
   "source": [
    "### Plot SPOT vs 13C MFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a761d23a-ae2e-4566-88db-63f0bc558078",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.plot(ax.get_xlim(), ax.get_ylim(), ls=\"--\", c=\".3\")\n",
    "\n",
    "for pathway in list(set(phenol_fluxes['Pathway'])):\n",
    "    pathway_df = phenol_fluxes[phenol_fluxes['Pathway'] == pathway]\n",
    "    \n",
    "    measured_flux_list = list(pathway_df['Flux'])\n",
    "    simulatedspot_flux_list = list(pathway_df['SPOT Value'])\n",
    "    \n",
    "    ax.scatter(measured_flux_list, simulatedspot_flux_list, label=pathway)\n",
    "\n",
    "\n",
    "# Dashed line\n",
    "x = np.linspace(*ax.get_xlim())\n",
    "ax.plot(x, x, ls=\"--\", c=\".3\")\n",
    "\n",
    "plt.title(r'Phenol 13C MFA vs. SPOT Fluxes', fontSize=20)\n",
    "plt.xlabel(r'13C MFA flux (per 100 mmol of phenol uptake)')\n",
    "plt.ylabel(r'SPOT flux (per 100 mmol of phenol uptake)')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e89a67cb-b755-40a8-b3cf-9611298e9ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "rxns = [r.id for r in model.reactions]\n",
    "# #EFlux2_fluxdict = dict(zip(rxns,EFlux2_min.solution.get_values()))\n",
    "# EFlux2_fluxdict = dict(zip(rxns,eflux2solfluxes.to_numpy()))#replace previous .get_values()\n",
    "# SPOT_fluxdict = dict(zip(rxns,spotsolWTPR1.to_numpy()))#replace previous .get_values()\n",
    "# #display(SPOT_fluxdict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97cabe14-623a-45fa-a4d7-913e86973ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scatterplot of solutions\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "\n",
    "def maefunc(y_predict, y_actual):\n",
    "    \"\"\"Mean Absolute Error.\n",
    "    Multioutput case included.\"\"\"\n",
    "\n",
    "    if y_actual.ndim == 1:\n",
    "        return np.mean(np.abs([y_a - y_p for y_a, y_p in zip(y_actual, y_predict)]))\n",
    "    else:\n",
    "        return [\n",
    "            np.mean(\n",
    "                np.abs([y_a - y_p for y_a, y_p in zip(y_actual[:, i], y_predict[:, i])])\n",
    "            )\n",
    "            for i in range(y_actual.shape[1])\n",
    "        ]\n",
    "    \n",
    "def scatterplotcompare(observed, predicted, labels, strtitle):\n",
    "    fig, ax = plt.subplots(figsize=(8, 8))\n",
    "    #res = stats.linregress(observed.to_numpy().astype(float), predicted.to_numpy())\n",
    "    #print(res.rvalue**2)\n",
    "    lims = [\n",
    "                np.min([observed, predicted]),  # min of both axes\n",
    "                np.max([observed, predicted]),  # max of both axes\n",
    "            ]\n",
    "    ax.set_xlim(lims)\n",
    "    #ax.plot([0.0, 0.2], [0.0, 0.2], 'k',  c='0.55')\n",
    "    # Plot Diagonal Dashed Line\n",
    "    ax.plot(lims, lims, ls=\"--\", color=\".8\", zorder=0)\n",
    "    for i in range(0, len(observed)):\n",
    "        ax.scatter(observed[i], predicted[i])\n",
    "        #ax.annotate(str(labels[i]),(observed[i],predicted[i]))\n",
    "    #ax.annotate(\"r-squared = {:.3f}\".format(round(r2_score(observed, predicted))), (0, 1))\n",
    "    #plt.plot(observed, res.intercept + res.slope*observed, 'k', label='fitted line')\n",
    "    rss = np.sum((predicted - observed) ** 2)\n",
    "    tss = np.sum((predicted - predicted.mean()) ** 2)\n",
    "    r2_1 = np.round(1 - rss / tss,2)\n",
    "    #display(r2_1)\n",
    "    #r2 = round(r2_score(observed, predicted),2)\n",
    "    mse = np.round(1/len(observed) * rss,2) \n",
    "    rmse = np.round(np.sqrt(1/len(observed) * rss),2)\n",
    "    #display(rmse)\n",
    "    #display(r2_score(observed, predicted))\n",
    "    mae_score = np.round(maefunc(observed, predicted),2)\n",
    "    plt.xlabel(r'EFLUX2')#Observed growth rates [$mmol/gDW/hr$]')\n",
    "    plt.ylabel(r'SPOT')#Predicted growth rates [$mmol/gDW/hr$]')\n",
    "    #plt.title(r'Growth rates: Observed vs. Predicted ('+strtitle+'), \\n' + f\"$R^2$={r2_1}, MAE={mae_score}, MSE = {mse}, RMSE={rmse}\", fontsize=18)\n",
    "    plt.title(r'EFLUX2 vs. SPOT ('+strtitle+'), \\n' + f\"$R^2$={r2_1}, MAE={mae_score}, MSE = {mse}, RMSE={rmse}\", fontsize=18)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0537a8eb-489b-4d5a-a409-f94185068f80",
   "metadata": {},
   "outputs": [],
   "source": [
    "scatterplotcompare(eflux2soldf.to_numpy(), spotsoldf.to_numpy(), labels='', strtitle='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cdf2f9a-363e-4b17-acbe-b9f9db7f4e86",
   "metadata": {},
   "outputs": [],
   "source": [
    "scatterplotcompare(eflux2soldf.loc['Growth_Phenol',:], spotsoldf.loc['Growth_Phenol',:], labels='Growth', strtitle='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a15513-f561-4e16-825c-2caf0b75e264",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c2a0376-c9c9-45ae-b02b-3d0c51f0cd7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from optlang.symbolics import add\n",
    "\"\"\"\n",
    "    Provides EFLUX2 predictions as explained in Machado et. al (2014) \n",
    "    \n",
    "        Parameters\n",
    "        ----------\n",
    "        model : cobrapy model.\n",
    "        Transcriptomics : pandas dataframe with transcriptomics data.\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        eflux2_sol as output from eflux2_model.optimize().\n",
    "        \n",
    "\"\"\"\n",
    "#Code only works for GPRs written in disjunctive normal form (DNF). Majority of models have them in DNF but there are some exceptions. \n",
    "\n",
    "def EFlux2test(model, Transcriptomics):\n",
    "    #with model as eflux2_model:\n",
    "    eflux2_model = model.copy()\n",
    "    # Parse GPR into a dict containing isozymes (separated by 'or')\n",
    "    # Each isozyme has a set of subunits (separated by 'and')\n",
    "    #'and' and 'or' can occur at the same time, or can occur by itself.\n",
    "    gpr_dict = dict()\n",
    "    for r in eflux2_model.reactions:\n",
    "        if r.gene_reaction_rule:\n",
    "            temp = set()\n",
    "            for x in [x.strip('() ') for x in r.gene_reaction_rule.split(' or ')]:\n",
    "                temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "            gpr_dict[r.id] = temp\n",
    "    #display(gpr_dict)\n",
    "    # Set the bounds using the transcriptomics data\n",
    "    for r in eflux2_model.reactions:\n",
    "        t2=0\n",
    "        if r.gene_reaction_rule:\n",
    "            #If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "            #If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "            #If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "            t = np.sum([np.min([Transcriptomics.loc[g].values if g in Transcriptomics.index\n",
    "                                else np.array([np.Inf]) for g in p])\n",
    "                        for p in gpr_dict[r.id]])\n",
    "            for p in gpr_dict[r.id]:\n",
    "                xg=[]\n",
    "                for g in p:\n",
    "                    if g in Transcriptomics.index:\n",
    "                        xg.append(Transcriptomics.loc[g].values)\n",
    "                    else:\n",
    "                        xg.append(np.array([np.Inf]))\n",
    "                    t2=np.min(xg)\n",
    "                t2+=t2\n",
    "            if t2!=t:\n",
    "                display('difference',t2, t)\n",
    "            if t != np.inf:\n",
    "                display(t)\n",
    "            if r.lower_bound < 0.0:\n",
    "                r.lower_bound = -t\n",
    "            else:\n",
    "                pass\n",
    "            if r.upper_bound > 0.0:\n",
    "                r.upper_bound = t\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "            #When there is no GPR, the arbitrary bounds are removed. \n",
    "            #Common arbitrary bound value of 1000 for E.coli, might be different depending on the model, e.g., 99999.0 for iMM904 yeast model in BiGG\n",
    "            if r.lower_bound < -99999:\n",
    "                r.lower_bound = -np.Inf\n",
    "            if r.upper_bound > 99999:\n",
    "                r.upper_bound = np.Inf \n",
    "            if r.lower_bound >= 0:\n",
    "                r.lower_bound = 0\n",
    "            if r.upper_bound <= 0:\n",
    "                r.upper_bound = 0\n",
    "\n",
    "#             if r.lower_bound < 0.0 and r.lower_bound > -1000.0:\n",
    "#                 print(r.id, r.lower_bound, r.upper_bound)\n",
    "#                 r.lower_bound = -1000.0\n",
    "#             elif r.lower_bound > 0.0:\n",
    "#                 print(r.id, r.lower_bound, r.upper_bound)\n",
    "#                 r.lower_bound = 0.0\n",
    "#             if r.upper_bound > 0.0 and r.upper_bound < 1000.0:\n",
    "#                 print(r.id, r.lower_bound, r.upper_bound)\n",
    "#                 r.upper_bound = 1000.0\n",
    "#             elif r.upper_bound < 0.0:\n",
    "#                 print(r.id, r.lower_bound, r.upper_bound)\n",
    "#                 r.upper_bound = 0.0\n",
    "#             if r.lower_bound == -1000.0:\n",
    "#                 r.lower_bound = -np.Inf\n",
    "#             if r.upper_bound == 1000.0:\n",
    "#                 r.upper_bound = np.Inf\n",
    "\n",
    "#             if r.lower_bound < 0.0:\n",
    "#                 r.lower_bound = -np.Inf\n",
    "#             elif r.lower_bound >= 0.0:\n",
    "#                 r.lower_bound = 0.0\n",
    "#             elif r.upper_bound > 0.0:\n",
    "#                 r.upper_bound = np.Inf\n",
    "#             elif r.upper_bound <= 0.0:\n",
    "                r.upper_bound = 0.0\n",
    "#                 if r=='Phenol_e':\n",
    "#                     display(r)\n",
    "#                     r.lower_bound = -100\n",
    "#                     r.upper_bound = 0\n",
    "    display(eflux2_model.medium)\n",
    "    # solve FBA to calculate the maximum biomass\n",
    "    eflux2_model.tolerance = 1e-9\n",
    "    fba_sol = eflux2_model.optimize()\n",
    "    print('FBA status', fba_sol.status)\n",
    "    print('FBA solution', fba_sol.objective_value)\n",
    "    display(eflux2_model.objective)\n",
    "    # Constrain the biomass to the optimal value\n",
    "    for r in eflux2_model.reactions:\n",
    "        if r.objective_coefficient:\n",
    "            r.lower_bound = fba_sol.objective_value\n",
    "    # minimize the sum of squared flux values\n",
    "    eflux2_model.objective = eflux2_model.problem.Objective(add([r.flux_expression**2 for r in eflux2_model.reactions]), direction='min')\n",
    "    display(eflux2_model.problem.Objective)\n",
    "    eflux2_sol = eflux2_model.optimize()\n",
    "    print('EFlux2 status', eflux2_sol.status)\n",
    "    print('EFlux2 solution', eflux2_sol.objective_value)\n",
    "    # return eflux2 solution\n",
    "    return eflux2_sol"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biodesign_3.7",
   "language": "python",
   "name": "biod_3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
