{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7563604a-2ba5-4e5d-b9e0-309615df0b5b",
   "metadata": {},
   "source": [
    "### **Notebook to create a dataframe that contains mapping of gsm reaction to transcript measurements for all reactions** "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a6edbe7-d3d0-49ac-9d13-8f43882945f0",
   "metadata": {},
   "source": [
    "##### **Import python packages**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5e37f811-1a95-45e0-932f-962edb144307",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cobra\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "from edd_utils import login, export_study, export_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "244d4ba7-9d90-4f21-8de8-18a77f1a87dd",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### **Load Yondeda 2016 data and Henson 2018 data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8232d35-b2bf-4b51-9625-3a12a76661f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Studies to Download\n",
    "glucose_study_slug = 'biodesign_yoneda_set3_reprocessed'\n",
    "phenol_study_slug = 'biodesign_henson2018_reprocessed'\n",
    "# EDD server\n",
    "edd_server = 'public-edd.jbei.org'\n",
    "user       = 'garrettroell'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b2b03dbf-95d3-49f6-b272-3e35cd07705a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Password for garrettroell:  ················\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "678d36af5ff54bc282465a2cb676c4c9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=192297.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "08484b377077463fbe5a539989319ffd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=440493.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "session = login(edd_server=edd_server, user=user)\n",
    "\n",
    "glucose_df = export_study(session, glucose_study_slug, edd_server=edd_server)\n",
    "phenol_df = export_study(session, phenol_study_slug, edd_server=edd_server)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f51c4cc-b55f-44ab-8b2c-374350327f29",
   "metadata": {},
   "source": [
    "##### **Filter transcriptomics data from all EDD data into different dataframes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9b399ad9-6558-4eaa-aac6-f57a2c3690dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study ID</th>\n",
       "      <th>Study Name</th>\n",
       "      <th>Line ID</th>\n",
       "      <th>Line Name</th>\n",
       "      <th>Line Description</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Assay ID</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Formal Type</th>\n",
       "      <th>Measurement Type</th>\n",
       "      <th>Compartment</th>\n",
       "      <th>Units</th>\n",
       "      <th>Value</th>\n",
       "      <th>Hours</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14400</td>\n",
       "      <td>Biodesign_Yoneda_set3_reprocessed</td>\n",
       "      <td>15006</td>\n",
       "      <td>EVOL33-LN-G-R1</td>\n",
       "      <td>R. Opacus adaptively evolved strain1 with 1.0 ...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>15030</td>\n",
       "      <td>EVOL33-LN-G-R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14400</td>\n",
       "      <td>Biodesign_Yoneda_set3_reprocessed</td>\n",
       "      <td>15007</td>\n",
       "      <td>EVOL33-LN-G-R2</td>\n",
       "      <td>R. Opacus adaptively evolved strain1 with 1.0 ...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>15031</td>\n",
       "      <td>EVOL33-LN-G-R2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14400</td>\n",
       "      <td>Biodesign_Yoneda_set3_reprocessed</td>\n",
       "      <td>15008</td>\n",
       "      <td>EVOL33-LN-G-R3</td>\n",
       "      <td>R. Opacus adaptively evolved strain1 with 1.0 ...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>15032</td>\n",
       "      <td>EVOL33-LN-G-R3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14400</td>\n",
       "      <td>Biodesign_Yoneda_set3_reprocessed</td>\n",
       "      <td>14991</td>\n",
       "      <td>EVOL33-LN-LP-R1</td>\n",
       "      <td>R. Opacus adaptively evolved strain1 with 0.75...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>15015</td>\n",
       "      <td>EVOL33-LN-LP-R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14400</td>\n",
       "      <td>Biodesign_Yoneda_set3_reprocessed</td>\n",
       "      <td>14992</td>\n",
       "      <td>EVOL33-LN-LP-R2</td>\n",
       "      <td>R. Opacus adaptively evolved strain1 with 0.75...</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>15016</td>\n",
       "      <td>EVOL33-LN-LP-R2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000104864_1</td>\n",
       "      <td>0</td>\n",
       "      <td>FPKM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Study ID                         Study Name  Line ID        Line Name  \\\n",
       "0     14400  Biodesign_Yoneda_set3_reprocessed    15006   EVOL33-LN-G-R1   \n",
       "1     14400  Biodesign_Yoneda_set3_reprocessed    15007   EVOL33-LN-G-R2   \n",
       "2     14400  Biodesign_Yoneda_set3_reprocessed    15008   EVOL33-LN-G-R3   \n",
       "3     14400  Biodesign_Yoneda_set3_reprocessed    14991  EVOL33-LN-LP-R1   \n",
       "4     14400  Biodesign_Yoneda_set3_reprocessed    14992  EVOL33-LN-LP-R2   \n",
       "\n",
       "                                    Line Description         Protocol  \\\n",
       "0  R. Opacus adaptively evolved strain1 with 1.0 ...  Transcriptomics   \n",
       "1  R. Opacus adaptively evolved strain1 with 1.0 ...  Transcriptomics   \n",
       "2  R. Opacus adaptively evolved strain1 with 1.0 ...  Transcriptomics   \n",
       "3  R. Opacus adaptively evolved strain1 with 0.75...  Transcriptomics   \n",
       "4  R. Opacus adaptively evolved strain1 with 0.75...  Transcriptomics   \n",
       "\n",
       "   Assay ID       Assay Name Formal Type Measurement Type  Compartment Units  \\\n",
       "0     15030   EVOL33-LN-G-R1         NaN   WP_000104864_1            0  FPKM   \n",
       "1     15031   EVOL33-LN-G-R2         NaN   WP_000104864_1            0  FPKM   \n",
       "2     15032   EVOL33-LN-G-R3         NaN   WP_000104864_1            0  FPKM   \n",
       "3     15015  EVOL33-LN-LP-R1         NaN   WP_000104864_1            0  FPKM   \n",
       "4     15016  EVOL33-LN-LP-R2         NaN   WP_000104864_1            0  FPKM   \n",
       "\n",
       "   Value  Hours  \n",
       "0    0.0   14.0  \n",
       "1    0.0   14.0  \n",
       "2    0.0   14.0  \n",
       "3    0.0   24.0  \n",
       "4    0.0   24.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glucose_trans_df = glucose_df[glucose_df['Protocol'].str.contains('Transcriptomics')]\n",
    "phenol_trans_df = phenol_df[phenol_df['Protocol'].str.contains('Transcriptomics')]\n",
    "glucose_trans_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7252a93-7cea-4b7c-8375-c03a194a9524",
   "metadata": {},
   "source": [
    "##### **Load Genome Scale Model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6f6b96da-3aaf-49fd-9dec-b674a678d243",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name =  '../GSMs/Ropacus_annotated_curated_with_phenol_custom_biomass.xml'\n",
    "model = cobra.io.read_sbml_model(file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c83f05dd-50a8-4cd4-af11-525367f046f1",
   "metadata": {},
   "source": [
    "#### Define functions to find transcript levels for a reaction given a transcriptomic dataframe ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e20b6edc-71c4-4205-90d7-0deb3ea70552",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gpr_dict(model):   \n",
    "    gpr_dict = dict()\n",
    "    for rxn in model.reactions:\n",
    "        if rxn.gene_reaction_rule:\n",
    "            temp = set()\n",
    "            for x in [x.strip('() ') for x in rxn.gene_reaction_rule.split(' or ')]:\n",
    "                temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "            gpr_dict[rxn.id] = temp\n",
    "    return gpr_dict\n",
    "\n",
    "\"\"\"\n",
    "    Calculates bound value based on transcriptomics data for reactions in gene reaction rule\n",
    "    \n",
    "    NOTE: \n",
    "    If a reaction R1 has the GPR of 'A and B', it would be parsed to { {A, B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B) ] ) = min(A, B).\n",
    "    If a reaction R1 has the GPR of 'A or B', it would be parsed to { {A}, {B} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A), min(B) ] ) = sum( [A, B] ).\n",
    "    If a reaction R1 has the GPR of '(A and B) or (C and D)', it would be parsed to { {A, B}, {C, D} } in gpr_dict['R1']. Then t for R1 would be sum( [ min(A, B), min(C, D) ] ).\n",
    "    \n",
    "        Parameters\n",
    "        ----------\n",
    "        model : cobrapy model.\n",
    "        Transcriptomics : pandas dataframe with transcriptomics data.Data frame has gene identifiers as index and just one column with transcript values.  \n",
    "        rxn : cobrapy model reaction\n",
    "        \n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        transscript bound value: float.\n",
    "\"\"\"\n",
    "\n",
    "def find_trans_bound_from_gpr(model, Transcriptomics, rxn, gpr_dict, newinf=np.inf):\n",
    "    finaltransval = 0\n",
    "    listids = []\n",
    "    for parallel_gene in gpr_dict[rxn.id]:\n",
    "        transvals = []\n",
    "        for gene in parallel_gene:\n",
    "            if gene in Transcriptomics.index:\n",
    "                transvals.append(Transcriptomics.loc[gene].values)\n",
    "            else:\n",
    "                transvals.append(np.inf)\n",
    "            mintransval=np.min(transvals)\n",
    "            if mintransval == np.inf:\n",
    "                mintransval= newinf\n",
    "        finaltransval = finaltransval + mintransval\n",
    "#         if finaltransval==newinfbound:\n",
    "#             display(rxn.id)\n",
    "#             listids.append(rxn.id)\n",
    "    return finaltransval\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd652bbb-02ff-4bb9-8a6c-daf465456c69",
   "metadata": {},
   "source": [
    "#### Define a function to get the input transcriptomic dataframe for find_trans_bound_from_gpr ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b09cf5c7-a3f0-4f31-9cab-24b032c1840a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_trans_df(trans_df, line_name):\n",
    "    line_name_df = trans_df[trans_df['Line Name']==line_name]\n",
    "    filtered_df = line_name_df.filter(['Value', 'Measurement Type'])\n",
    "    filtered_df.set_index('Measurement Type', inplace=True)\n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51fdfc33-c92f-4173-9657-854ff6029097",
   "metadata": {},
   "source": [
    "#### Create the basic structure of the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7d7720c2-8c5f-400a-a01a-41ece366f492",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_name</th>\n",
       "      <th>reaction_string</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12DGR120tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr120_p --&gt; 12dgr120_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR140tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr140_p --&gt; 12dgr140_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR141tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr141_p --&gt; 12dgr141_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR161tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr161_p --&gt; 12dgr161_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR180tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr180_p --&gt; 12dgr180_c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  reaction_name  \\\n",
       "12DGR120tipp  1,2 diacylglycerol transport via flipping (per...   \n",
       "12DGR140tipp  1,2 diacylglycerol transport via flipping (per...   \n",
       "12DGR141tipp  1,2 diacylglycerol transport via flipping (per...   \n",
       "12DGR161tipp  1,2 diacylglycerol transport via flipping (per...   \n",
       "12DGR180tipp  1,2 diacylglycerol transport via flipping (per...   \n",
       "\n",
       "                        reaction_string  \n",
       "12DGR120tipp  12dgr120_p --> 12dgr120_c  \n",
       "12DGR140tipp  12dgr140_p --> 12dgr140_c  \n",
       "12DGR141tipp  12dgr141_p --> 12dgr141_c  \n",
       "12DGR161tipp  12dgr161_p --> 12dgr161_c  \n",
       "12DGR180tipp  12dgr180_p --> 12dgr180_c  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction_df = pd.DataFrame({\n",
    "        'reaction_name': [rxn.name for rxn in model.reactions],\n",
    "        'reaction_string': [rxn.reaction for rxn in model.reactions],\n",
    "    },\n",
    "    columns = ['reaction_name', 'reaction_string'],\n",
    "    index=[rxn.id for rxn in model.reactions]\n",
    ")\n",
    "reaction_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a5abb9a-1fec-4255-b323-9ee5c388d648",
   "metadata": {},
   "source": [
    "#### Fill dataframe with data from glucose conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "742f0f5a-55d9-4553-9a9d-dc4e8629bd4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WT-LN-G-R1\n",
      "WT-LN-G-R2\n",
      "WT-LN-G-R3\n",
      "EVOL33-LN-G-R1\n",
      "EVOL33-LN-G-R2\n",
      "EVOL33-LN-G-R3\n",
      "EVOL40-LN-G-R1\n",
      "EVOL40-LN-G-R2\n",
      "EVOL40-LN-G-R3\n",
      "WT-P-R1\n",
      "WT-P-R2\n",
      "WT-P-R3\n",
      "PVHG-P-R1\n",
      "PVHG-P-R2\n",
      "PVHG-P-R3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_name</th>\n",
       "      <th>reaction_string</th>\n",
       "      <th>WT-LN-G-R1 trans</th>\n",
       "      <th>WT-LN-G-R2 trans</th>\n",
       "      <th>WT-LN-G-R3 trans</th>\n",
       "      <th>EVOL33-LN-G-R1 trans</th>\n",
       "      <th>EVOL33-LN-G-R2 trans</th>\n",
       "      <th>EVOL33-LN-G-R3 trans</th>\n",
       "      <th>EVOL40-LN-G-R1 trans</th>\n",
       "      <th>EVOL40-LN-G-R2 trans</th>\n",
       "      <th>EVOL40-LN-G-R3 trans</th>\n",
       "      <th>WT-P-R1 trans</th>\n",
       "      <th>WT-P-R2 trans</th>\n",
       "      <th>WT-P-R3 trans</th>\n",
       "      <th>PVHG-P-R1 trans</th>\n",
       "      <th>PVHG-P-R2 trans</th>\n",
       "      <th>PVHG-P-R3 trans</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12DGR120tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr120_p --&gt; 12dgr120_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR140tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr140_p --&gt; 12dgr140_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR141tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr141_p --&gt; 12dgr141_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR161tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr161_p --&gt; 12dgr161_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR180tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr180_p --&gt; 12dgr180_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR181tipp</th>\n",
       "      <td>1,2 diacylglycerol transport via flipping (per...</td>\n",
       "      <td>12dgr181_p --&gt; 12dgr181_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12PPDStpp</th>\n",
       "      <td>(S)-Propane-1,2-diol facilitated transport (pe...</td>\n",
       "      <td>12ppd__S_p &lt;=&gt; 12ppd__S_c</td>\n",
       "      <td>1.66651</td>\n",
       "      <td>1.74287</td>\n",
       "      <td>1.04483</td>\n",
       "      <td>1.59518</td>\n",
       "      <td>1.61617</td>\n",
       "      <td>1.21061</td>\n",
       "      <td>1.50650</td>\n",
       "      <td>1.20227</td>\n",
       "      <td>1.22692</td>\n",
       "      <td>2.37356</td>\n",
       "      <td>2.41686</td>\n",
       "      <td>2.82043</td>\n",
       "      <td>1.63712</td>\n",
       "      <td>1.43141</td>\n",
       "      <td>2.30775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12PPDt</th>\n",
       "      <td>S-Propane-1,2-diol facilitated transport</td>\n",
       "      <td>12ppd__S_e &lt;=&gt; 12ppd__S_c</td>\n",
       "      <td>1.66651</td>\n",
       "      <td>1.74287</td>\n",
       "      <td>1.04483</td>\n",
       "      <td>1.59518</td>\n",
       "      <td>1.61617</td>\n",
       "      <td>1.21061</td>\n",
       "      <td>1.50650</td>\n",
       "      <td>1.20227</td>\n",
       "      <td>1.22692</td>\n",
       "      <td>2.37356</td>\n",
       "      <td>2.41686</td>\n",
       "      <td>2.82043</td>\n",
       "      <td>1.63712</td>\n",
       "      <td>1.43141</td>\n",
       "      <td>2.30775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14GLUCANabcpp</th>\n",
       "      <td>1,4-alpha-D-glucan transport via ABC system (p...</td>\n",
       "      <td>14glucan_p + atp_c + h2o_c --&gt; 14glucan_c + ad...</td>\n",
       "      <td>7.94798</td>\n",
       "      <td>6.75360</td>\n",
       "      <td>7.66209</td>\n",
       "      <td>9.89014</td>\n",
       "      <td>9.62353</td>\n",
       "      <td>10.03080</td>\n",
       "      <td>5.96575</td>\n",
       "      <td>8.13841</td>\n",
       "      <td>9.33820</td>\n",
       "      <td>1.14283</td>\n",
       "      <td>3.18587</td>\n",
       "      <td>1.34889</td>\n",
       "      <td>2.12826</td>\n",
       "      <td>1.43141</td>\n",
       "      <td>1.23080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14GLUCANtexi</th>\n",
       "      <td>1,4-alpha-D-glucan transport via diffusion (ex...</td>\n",
       "      <td>14glucan_e --&gt; 14glucan_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPE120tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoethanolamine (n-C12...</td>\n",
       "      <td>2agpe120_p --&gt; 2agpe120_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPE140tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoethanolamine (n-C14...</td>\n",
       "      <td>2agpe140_p --&gt; 2agpe140_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPE141tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoethanolamine (n-C14...</td>\n",
       "      <td>2agpe141_p --&gt; 2agpe141_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPE161tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoethanolamine (n-C16...</td>\n",
       "      <td>2agpe161_p --&gt; 2agpe161_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPE180tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoethanolamine (n-C18...</td>\n",
       "      <td>2agpe180_p --&gt; 2agpe180_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPE181tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoethanolamine (n-C18...</td>\n",
       "      <td>2agpe181_p --&gt; 2agpe181_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPEAT120</th>\n",
       "      <td>2-acyl-glycerophospho-ethanolamine acyltransfe...</td>\n",
       "      <td>2agpe120_c + atp_c + ddca_c --&gt; amp_c + pe120_...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPEAT140</th>\n",
       "      <td>2-acyl-glycerophospho-ethanolamine acyltransfe...</td>\n",
       "      <td>2agpe140_c + atp_c + ttdca_c --&gt; amp_c + pe140...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPEAT141</th>\n",
       "      <td>2-acyl-glycerophospho-ethanolamine acyltransfe...</td>\n",
       "      <td>2agpe141_c + atp_c + ttdcea_c --&gt; amp_c + pe14...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPEAT160</th>\n",
       "      <td>2-acyl-glycerophospho-ethanolamine acyltransfe...</td>\n",
       "      <td>2agpe160_c + atp_c + hdca_c --&gt; amp_c + pe160_...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPEAT161</th>\n",
       "      <td>2-acyl-glycerophospho-ethanolamine acyltransfe...</td>\n",
       "      <td>2agpe161_c + atp_c + hdcea_c --&gt; amp_c + pe161...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPEAT180</th>\n",
       "      <td>2-acyl-glycerophospho-ethanolamine acyltransfe...</td>\n",
       "      <td>2agpe180_c + atp_c + ocdca_c --&gt; amp_c + pe180...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPEAT181</th>\n",
       "      <td>2-acyl-glycerophospho-ethanolamine acyltransfe...</td>\n",
       "      <td>2agpe181_c + atp_c + ocdcea_c --&gt; amp_c + pe18...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPG120tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoglycerol (n-C12:0) ...</td>\n",
       "      <td>2agpg120_p --&gt; 2agpg120_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPG140tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoglycerol (n-C14:0) ...</td>\n",
       "      <td>2agpg140_p --&gt; 2agpg140_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPG141tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoglycerol (n-C14:1) ...</td>\n",
       "      <td>2agpg141_p --&gt; 2agpg141_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPG160tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoglycerol (n-C16:0) ...</td>\n",
       "      <td>2agpg160_p --&gt; 2agpg160_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPG161tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoglycerol (n-C16:1) ...</td>\n",
       "      <td>2agpg161_p --&gt; 2agpg161_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPG180tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoglycerol (n-C18:0) ...</td>\n",
       "      <td>2agpg180_p --&gt; 2agpg180_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2AGPG181tipp</th>\n",
       "      <td>2-Acyl-sn-glycero-3-phosphoglycerol (n-C18:1) ...</td>\n",
       "      <td>2agpg181_p --&gt; 2agpg181_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_s_e</th>\n",
       "      <td>R_EX_s_e</td>\n",
       "      <td>s_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_sucr_e</th>\n",
       "      <td>R_EX_sucr_e</td>\n",
       "      <td>sucr_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_sula_e</th>\n",
       "      <td>R_EX_sula_e</td>\n",
       "      <td>sula_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_sulfac_e</th>\n",
       "      <td>R_EX_sulfac_e</td>\n",
       "      <td>sulfac_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_tartr__D_e</th>\n",
       "      <td>R_EX_tartr__D_e</td>\n",
       "      <td>tartr__D_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_tartr__L_e</th>\n",
       "      <td>R_EX_tartr__L_e</td>\n",
       "      <td>tartr__L_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_thymd_e</th>\n",
       "      <td>R_EX_thymd_e</td>\n",
       "      <td>thymd_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_tre_e</th>\n",
       "      <td>R_EX_tre_e</td>\n",
       "      <td>tre_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_tsul_e</th>\n",
       "      <td>R_EX_tsul_e</td>\n",
       "      <td>tsul_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_ura_e</th>\n",
       "      <td>R_EX_ura_e</td>\n",
       "      <td>ura_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_xan_e</th>\n",
       "      <td>R_EX_xan_e</td>\n",
       "      <td>xan_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_xmp_e</th>\n",
       "      <td>R_EX_xmp_e</td>\n",
       "      <td>xmp_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_xtsn_e</th>\n",
       "      <td>R_EX_xtsn_e</td>\n",
       "      <td>xtsn_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_xylan4_e</th>\n",
       "      <td>R_EX_xylan4_e</td>\n",
       "      <td>xylan4_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_xyl3_e</th>\n",
       "      <td>R_EX_xyl3_e</td>\n",
       "      <td>xyl3_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_xyl__D_e</th>\n",
       "      <td>R_EX_xyl__D_e</td>\n",
       "      <td>xyl__D_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_xylb_e</th>\n",
       "      <td>R_EX_xylb_e</td>\n",
       "      <td>xylb_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sink_2ohph_c</th>\n",
       "      <td>R_sink_2ohph_c</td>\n",
       "      <td>2ohph_c --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sink_4hba_c</th>\n",
       "      <td>R_sink_4hba_c</td>\n",
       "      <td>4hba_c --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sink_5drib_c</th>\n",
       "      <td>R_sink_5drib_c</td>\n",
       "      <td>5drib_c --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sink_bmocogdp_c</th>\n",
       "      <td>R_sink_bmocogdp_c</td>\n",
       "      <td>bmocogdp_c --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sink_lipopb_c</th>\n",
       "      <td>R_sink_lipopb_c</td>\n",
       "      <td>lipopb_c --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sink_sheme_c</th>\n",
       "      <td>R_sink_sheme_c</td>\n",
       "      <td>sheme_c --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth</th>\n",
       "      <td>Biomass reaction</td>\n",
       "      <td>0.000216499790822049 10fthf_c + 0.498715520392...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATPM</th>\n",
       "      <td>R_ATPM</td>\n",
       "      <td>atp_c + h2o_c --&gt; adp_c + h_c + pi_c</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_phenol_e</th>\n",
       "      <td>Phenol exchange</td>\n",
       "      <td>phenol_e &lt;=&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phenol_transport</th>\n",
       "      <td>Phenol exchange</td>\n",
       "      <td>phenol_e &lt;=&gt; phenol_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHEMOX</th>\n",
       "      <td>Phenol exchange</td>\n",
       "      <td>h_c + nadh_c + o2_c + phenol_c --&gt; catechol_c ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Phenol</th>\n",
       "      <td>Phenol biomass reaction</td>\n",
       "      <td>0.000216 10fthf_c + 0.10703270532983 ala__L_c ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth_Glucose</th>\n",
       "      <td>glucose biomass reaction</td>\n",
       "      <td>0.000216 10fthf_c + 0.306046641802482 ala__L_c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2385 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      reaction_name  \\\n",
       "12DGR120tipp      1,2 diacylglycerol transport via flipping (per...   \n",
       "12DGR140tipp      1,2 diacylglycerol transport via flipping (per...   \n",
       "12DGR141tipp      1,2 diacylglycerol transport via flipping (per...   \n",
       "12DGR161tipp      1,2 diacylglycerol transport via flipping (per...   \n",
       "12DGR180tipp      1,2 diacylglycerol transport via flipping (per...   \n",
       "...                                                             ...   \n",
       "EX_phenol_e                                         Phenol exchange   \n",
       "phenol_transport                                    Phenol exchange   \n",
       "PHEMOX                                              Phenol exchange   \n",
       "Growth_Phenol                               Phenol biomass reaction   \n",
       "Growth_Glucose                             glucose biomass reaction   \n",
       "\n",
       "                                                    reaction_string  \\\n",
       "12DGR120tipp                              12dgr120_p --> 12dgr120_c   \n",
       "12DGR140tipp                              12dgr140_p --> 12dgr140_c   \n",
       "12DGR141tipp                              12dgr141_p --> 12dgr141_c   \n",
       "12DGR161tipp                              12dgr161_p --> 12dgr161_c   \n",
       "12DGR180tipp                              12dgr180_p --> 12dgr180_c   \n",
       "...                                                             ...   \n",
       "EX_phenol_e                                           phenol_e <=>    \n",
       "phenol_transport                              phenol_e <=> phenol_c   \n",
       "PHEMOX            h_c + nadh_c + o2_c + phenol_c --> catechol_c ...   \n",
       "Growth_Phenol     0.000216 10fthf_c + 0.10703270532983 ala__L_c ...   \n",
       "Growth_Glucose    0.000216 10fthf_c + 0.306046641802482 ala__L_c...   \n",
       "\n",
       "                  WT-LN-G-R1 trans  WT-LN-G-R2 trans  WT-LN-G-R3 trans  \\\n",
       "12DGR120tipp                   NaN               NaN               NaN   \n",
       "12DGR140tipp                   NaN               NaN               NaN   \n",
       "12DGR141tipp                   NaN               NaN               NaN   \n",
       "12DGR161tipp                   NaN               NaN               NaN   \n",
       "12DGR180tipp                   NaN               NaN               NaN   \n",
       "...                            ...               ...               ...   \n",
       "EX_phenol_e                    NaN               NaN               NaN   \n",
       "phenol_transport               NaN               NaN               NaN   \n",
       "PHEMOX                         NaN               NaN               NaN   \n",
       "Growth_Phenol                  NaN               NaN               NaN   \n",
       "Growth_Glucose                 NaN               NaN               NaN   \n",
       "\n",
       "                  EVOL33-LN-G-R1 trans  EVOL33-LN-G-R2 trans  \\\n",
       "12DGR120tipp                       NaN                   NaN   \n",
       "12DGR140tipp                       NaN                   NaN   \n",
       "12DGR141tipp                       NaN                   NaN   \n",
       "12DGR161tipp                       NaN                   NaN   \n",
       "12DGR180tipp                       NaN                   NaN   \n",
       "...                                ...                   ...   \n",
       "EX_phenol_e                        NaN                   NaN   \n",
       "phenol_transport                   NaN                   NaN   \n",
       "PHEMOX                             NaN                   NaN   \n",
       "Growth_Phenol                      NaN                   NaN   \n",
       "Growth_Glucose                     NaN                   NaN   \n",
       "\n",
       "                  EVOL33-LN-G-R3 trans  EVOL40-LN-G-R1 trans  \\\n",
       "12DGR120tipp                       NaN                   NaN   \n",
       "12DGR140tipp                       NaN                   NaN   \n",
       "12DGR141tipp                       NaN                   NaN   \n",
       "12DGR161tipp                       NaN                   NaN   \n",
       "12DGR180tipp                       NaN                   NaN   \n",
       "...                                ...                   ...   \n",
       "EX_phenol_e                        NaN                   NaN   \n",
       "phenol_transport                   NaN                   NaN   \n",
       "PHEMOX                             NaN                   NaN   \n",
       "Growth_Phenol                      NaN                   NaN   \n",
       "Growth_Glucose                     NaN                   NaN   \n",
       "\n",
       "                  EVOL40-LN-G-R2 trans  EVOL40-LN-G-R3 trans  WT-P-R1 trans  \\\n",
       "12DGR120tipp                       NaN                   NaN            NaN   \n",
       "12DGR140tipp                       NaN                   NaN            NaN   \n",
       "12DGR141tipp                       NaN                   NaN            NaN   \n",
       "12DGR161tipp                       NaN                   NaN            NaN   \n",
       "12DGR180tipp                       NaN                   NaN            NaN   \n",
       "...                                ...                   ...            ...   \n",
       "EX_phenol_e                        NaN                   NaN            NaN   \n",
       "phenol_transport                   NaN                   NaN            NaN   \n",
       "PHEMOX                             NaN                   NaN            NaN   \n",
       "Growth_Phenol                      NaN                   NaN            NaN   \n",
       "Growth_Glucose                     NaN                   NaN            NaN   \n",
       "\n",
       "                  WT-P-R2 trans  WT-P-R3 trans  PVHG-P-R1 trans  \\\n",
       "12DGR120tipp                NaN            NaN              NaN   \n",
       "12DGR140tipp                NaN            NaN              NaN   \n",
       "12DGR141tipp                NaN            NaN              NaN   \n",
       "12DGR161tipp                NaN            NaN              NaN   \n",
       "12DGR180tipp                NaN            NaN              NaN   \n",
       "...                         ...            ...              ...   \n",
       "EX_phenol_e                 NaN            NaN              NaN   \n",
       "phenol_transport            NaN            NaN              NaN   \n",
       "PHEMOX                      NaN            NaN              NaN   \n",
       "Growth_Phenol               NaN            NaN              NaN   \n",
       "Growth_Glucose              NaN            NaN              NaN   \n",
       "\n",
       "                  PVHG-P-R2 trans  PVHG-P-R3 trans  \n",
       "12DGR120tipp                  NaN              NaN  \n",
       "12DGR140tipp                  NaN              NaN  \n",
       "12DGR141tipp                  NaN              NaN  \n",
       "12DGR161tipp                  NaN              NaN  \n",
       "12DGR180tipp                  NaN              NaN  \n",
       "...                           ...              ...  \n",
       "EX_phenol_e                   NaN              NaN  \n",
       "phenol_transport              NaN              NaN  \n",
       "PHEMOX                        NaN              NaN  \n",
       "Growth_Phenol                 NaN              NaN  \n",
       "Growth_Glucose                NaN              NaN  \n",
       "\n",
       "[2385 rows x 17 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gpr_dict = create_gpr_dict(model)\n",
    "\n",
    "glucose_conditions = [\n",
    "    'WT-LN-G-R1',\n",
    "    'WT-LN-G-R2',\n",
    "    'WT-LN-G-R3',\n",
    "    'EVOL33-LN-G-R1',\n",
    "    'EVOL33-LN-G-R2',\n",
    "    'EVOL33-LN-G-R3',\n",
    "    'EVOL40-LN-G-R1',\n",
    "    'EVOL40-LN-G-R2',\n",
    "    'EVOL40-LN-G-R3',\n",
    "]\n",
    "\n",
    "phenol_conditions = [\n",
    "    'WT-P-R1',\n",
    "    'WT-P-R2',\n",
    "    'WT-P-R3',\n",
    "    'PVHG-P-R1',\n",
    "    'PVHG-P-R2',\n",
    "    'PVHG-P-R3',\n",
    "]\n",
    "\n",
    "for condition in glucose_conditions + phenol_conditions:\n",
    "    print(condition)\n",
    "    \n",
    "    # use the transcript dataframe that matches the condition\n",
    "    trans_df = glucose_trans_df if '-G-' in condition else phenol_trans_df\n",
    "    \n",
    "    # get the formatted transcriptomic df for the condition\n",
    "    transcriptomics_df = construct_trans_df(trans_df, condition)\n",
    "\n",
    "    # define a list to hold all transcript measurements for a condtion\n",
    "    transcript_measurements = []\n",
    "\n",
    "    for index, _ in reaction_df.iterrows():\n",
    "        rxn = model.reactions.get_by_id(index)\n",
    "        if rxn.id in gpr_dict.keys():\n",
    "            trans_bound = find_trans_bound_from_gpr(model, transcriptomics_df, rxn, gpr_dict, newinf=np.inf)\n",
    "            transcript_measurements.append(trans_bound)\n",
    "        else:\n",
    "            transcript_measurements.append(np.nan)\n",
    "\n",
    "    # transcript_measurements \n",
    "\n",
    "    reaction_df[f'{condition} trans'] = transcript_measurements\n",
    "\n",
    "reaction_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd628efb-7b15-4cb6-8829-734adbba52a4",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Save dataframe to as csv to be used in other notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "87cc4180-7320-4c0f-8f11-fd3931a9d695",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>Forward Reactions</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Location on map</th>\n",
       "      <th>Flux</th>\n",
       "      <th>90% Confidence Lower Bound</th>\n",
       "      <th>90% Confidence Upper Bound</th>\n",
       "      <th>EVOL33-LN-G-R1 transcripts</th>\n",
       "      <th>EVOL33-LN-G-R2 transcripts</th>\n",
       "      <th>EVOL33-LN-G-R3 transcripts</th>\n",
       "      <th>EVOL40-LN-G-R1 transcripts</th>\n",
       "      <th>EVOL40-LN-G-R2 transcripts</th>\n",
       "      <th>EVOL40-LN-G-R3 transcripts</th>\n",
       "      <th>WT-LN-G-R1 transcripts</th>\n",
       "      <th>WT-LN-G-R2 transcripts</th>\n",
       "      <th>WT-LN-G-R3 transcripts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Glucose Uptake</td>\n",
       "      <td>reverse_EX_glc__D_e</td>\n",
       "      <td>Gluc.ext + ATP -&gt; G6P</td>\n",
       "      <td>(50, 460)</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EMP Pathway</td>\n",
       "      <td>PGI</td>\n",
       "      <td>G6P &lt;-&gt; F6P</td>\n",
       "      <td>(-150, 430)</td>\n",
       "      <td>-1.61</td>\n",
       "      <td>-2.09</td>\n",
       "      <td>1.42</td>\n",
       "      <td>339.77434</td>\n",
       "      <td>333.37109</td>\n",
       "      <td>325.13618</td>\n",
       "      <td>282.25838</td>\n",
       "      <td>282.34734</td>\n",
       "      <td>271.35304</td>\n",
       "      <td>270.55161</td>\n",
       "      <td>242.91212</td>\n",
       "      <td>285.06464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EMP Pathway</td>\n",
       "      <td>PFK or reverse_FBP</td>\n",
       "      <td>F6P + ATP -&gt; FBP</td>\n",
       "      <td>(-220, 195)</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.91</td>\n",
       "      <td>642.00867</td>\n",
       "      <td>599.81818</td>\n",
       "      <td>617.06697</td>\n",
       "      <td>667.32051</td>\n",
       "      <td>648.11338</td>\n",
       "      <td>625.11398</td>\n",
       "      <td>704.42129</td>\n",
       "      <td>648.56446</td>\n",
       "      <td>741.30736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EMP Pathway</td>\n",
       "      <td>FBA</td>\n",
       "      <td>FBP &lt;-&gt; DHAP + GAP</td>\n",
       "      <td>(-140, 115)</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.91</td>\n",
       "      <td>113.47081</td>\n",
       "      <td>103.43466</td>\n",
       "      <td>116.04595</td>\n",
       "      <td>88.28107</td>\n",
       "      <td>102.10006</td>\n",
       "      <td>85.06621</td>\n",
       "      <td>100.31113</td>\n",
       "      <td>98.25414</td>\n",
       "      <td>109.88136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EMP Pathway</td>\n",
       "      <td>TPI</td>\n",
       "      <td>DHAP &lt;-&gt; GAP</td>\n",
       "      <td>(-270, 150)</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.91</td>\n",
       "      <td>100.92202</td>\n",
       "      <td>101.01041</td>\n",
       "      <td>118.81306</td>\n",
       "      <td>79.72413</td>\n",
       "      <td>76.11263</td>\n",
       "      <td>86.29313</td>\n",
       "      <td>79.60794</td>\n",
       "      <td>63.17894</td>\n",
       "      <td>72.44159</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Pathway    Forward Reactions               Reaction Location on map  \\\n",
       "0  Glucose Uptake  reverse_EX_glc__D_e  Gluc.ext + ATP -> G6P       (50, 460)   \n",
       "1     EMP Pathway                  PGI            G6P <-> F6P     (-150, 430)   \n",
       "2     EMP Pathway   PFK or reverse_FBP       F6P + ATP -> FBP     (-220, 195)   \n",
       "3     EMP Pathway                  FBA     FBP <-> DHAP + GAP     (-140, 115)   \n",
       "4     EMP Pathway                  TPI           DHAP <-> GAP     (-270, 150)   \n",
       "\n",
       "     Flux  90% Confidence Lower Bound  90% Confidence Upper Bound  \\\n",
       "0  100.00                      100.00                      100.00   \n",
       "1   -1.61                       -2.09                        1.42   \n",
       "2    0.00                        0.00                        1.91   \n",
       "3    0.00                        0.00                        1.91   \n",
       "4    0.00                        0.00                        1.91   \n",
       "\n",
       "   EVOL33-LN-G-R1 transcripts  EVOL33-LN-G-R2 transcripts  \\\n",
       "0                     0.00000                     0.00000   \n",
       "1                   339.77434                   333.37109   \n",
       "2                   642.00867                   599.81818   \n",
       "3                   113.47081                   103.43466   \n",
       "4                   100.92202                   101.01041   \n",
       "\n",
       "   EVOL33-LN-G-R3 transcripts  EVOL40-LN-G-R1 transcripts  \\\n",
       "0                     0.00000                     0.00000   \n",
       "1                   325.13618                   282.25838   \n",
       "2                   617.06697                   667.32051   \n",
       "3                   116.04595                    88.28107   \n",
       "4                   118.81306                    79.72413   \n",
       "\n",
       "   EVOL40-LN-G-R2 transcripts  EVOL40-LN-G-R3 transcripts  \\\n",
       "0                     0.00000                     0.00000   \n",
       "1                   282.34734                   271.35304   \n",
       "2                   648.11338                   625.11398   \n",
       "3                   102.10006                    85.06621   \n",
       "4                    76.11263                    86.29313   \n",
       "\n",
       "   WT-LN-G-R1 transcripts  WT-LN-G-R2 transcripts  WT-LN-G-R3 transcripts  \n",
       "0                 0.00000                 0.00000                 0.00000  \n",
       "1               270.55161               242.91212               285.06464  \n",
       "2               704.42129               648.56446               741.30736  \n",
       "3               100.31113                98.25414               109.88136  \n",
       "4                79.60794                63.17894                72.44159  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glucose_conditions = [\n",
    "    'EVOL33-LN-G-R1',\n",
    "    'EVOL33-LN-G-R2',\n",
    "    'EVOL33-LN-G-R3',\n",
    "    'EVOL40-LN-G-R1',\n",
    "    'EVOL40-LN-G-R2',\n",
    "    'EVOL40-LN-G-R3',\n",
    "    'WT-LN-G-R1',\n",
    "    'WT-LN-G-R2',\n",
    "    'WT-LN-G-R3'\n",
    "]\n",
    "\n",
    "for condition in glucose_conditions:\n",
    "\n",
    "    # get the transcriptomic df for the condition\n",
    "    transcriptomics_df = constructTransdffrompred(trans_df, condition)\n",
    "\n",
    "    # define a list to hold all transcript measurements for a condtion\n",
    "    transcript_measurements = [] \n",
    "\n",
    "    # loop over the 13C MFA reactions\n",
    "    for _, row in glucose_fluxes.iterrows():\n",
    "\n",
    "        # get the reaction ids associated with the flux for that row\n",
    "        reaction_id_string = row['Forward Reactions']\n",
    "        mfa_reaction = row['Reaction']\n",
    "\n",
    "        # remove directionality\n",
    "        reaction_id_string = reaction_id_string.replace('reverse_', '')\n",
    "\n",
    "        # convert to an array of parallel sets of reactions\n",
    "        reaction_ids = reaction_id_string.split(' or ')\n",
    "\n",
    "        # define a value to act as a counter\n",
    "        transcript_level = 0\n",
    "\n",
    "        for reaction_id in reaction_ids:\n",
    "            # remove spaces and parentheses\n",
    "            reaction_id = reaction_id.strip('() ')\n",
    "\n",
    "            if ' and ' in reaction_id:\n",
    "                series_reactions = reaction_id.split(' and ')\n",
    "                # loop over reactions that have and relationships and add min value to the transcript total\n",
    "                series_transcripts = []\n",
    "                for series_reaction in series_reactions:\n",
    "                    # get transcript level for a given reaction in the genome scale model\n",
    "                    rxn = model.reactions.get_by_id(series_reaction)\n",
    "                    transcript_level = findtransboundval_forgprrxns(model, transcriptomics_df, rxn, newinf=np.inf)\n",
    "                    series_transcripts.append(transcript_level)\n",
    "                # only add the minimum transcript level of reactions connected by 'and' relations\n",
    "                transcript_level += min(series_transcripts)\n",
    "            else:\n",
    "                rxn = model.reactions.get_by_id(reaction_id)\n",
    "                if rxn.id in create_gprdict(model).keys():\n",
    "                    reaction_transcripts =  findtransboundval_forgprrxns(model, transcriptomics_df, rxn)\n",
    "                    transcript_level += reaction_transcripts\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "        transcript_measurements.append(transcript_level)\n",
    "\n",
    "    glucose_fluxes[condition + ' transcripts'] = transcript_measurements\n",
    "\n",
    "glucose_fluxes.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biodesign_3.7",
   "language": "python",
   "name": "biod_3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
